--- intel/intel_bufmgr_gem.c	Wed Feb  9 09:36:09 2011
+++ intel/intel_bufmgr_gem.c	Wed Feb  9 09:38:11 2011
@@ -116,7 +116,7 @@
 struct _drm_intel_bo_gem {
 	drm_intel_bo bo;
 
-	atomic_t refcount;
+	int refcount;
 	uint32_t gem_handle;
 	const char *name;
 
@@ -351,7 +351,7 @@
 {
 	drm_intel_bo_gem *bo_gem = (drm_intel_bo_gem *) bo;
 
-	atomic_inc(&bo_gem->refcount);
+        bo_gem->refcount++;
 }
 
 /**
@@ -693,7 +693,7 @@
 	}
 
 	bo_gem->name = name;
-	atomic_set(&bo_gem->refcount, 1);
+	bo_gem->refcount = 1;
 	bo_gem->validate_index = -1;
 	bo_gem->reloc_tree_fences = 0;
 	bo_gem->used_as_reloc_target = 0;
@@ -813,7 +813,7 @@
 	bo_gem->bo.virtual = NULL;
 	bo_gem->bo.bufmgr = bufmgr;
 	bo_gem->name = name;
-	atomic_set(&bo_gem->refcount, 1);
+	bo_gem->refcount = 1;
 	bo_gem->validate_index = -1;
 	bo_gem->gem_handle = open_arg.handle;
 	bo_gem->global_name = handle;
@@ -945,8 +945,8 @@
 {
 	drm_intel_bo_gem *bo_gem = (drm_intel_bo_gem *) bo;
 
-	assert(atomic_read(&bo_gem->refcount) > 0);
-	if (atomic_dec_and_test(&bo_gem->refcount))
+	assert(bo_gem->refcount > 0);
+	if (--bo_gem->refcount == 0)
 		drm_intel_gem_bo_unreference_final(bo, time);
 }
 
@@ -954,8 +954,8 @@
 {
 	drm_intel_bo_gem *bo_gem = (drm_intel_bo_gem *) bo;
 
-	assert(atomic_read(&bo_gem->refcount) > 0);
-	if (atomic_dec_and_test(&bo_gem->refcount)) {
+	assert(bo_gem->refcount > 0);
+	if (--bo_gem->refcount == 0) {
 		drm_intel_bufmgr_gem *bufmgr_gem =
 		    (drm_intel_bufmgr_gem *) bo->bufmgr;
 		struct timespec time;
