--- src/bitmap/pcfread.c.orig   Tue Aug 29 19:15:50 2006
+++ src/bitmap/pcfread.c        Tue Dec 18 17:22:26 2007
@@ -588,7 +588,10 @@
     pFont->info.lastRow = pcfGetINT16(file, format);
     pFont->info.defaultCh = pcfGetINT16(file, format);
     if (IS_EOF(file)) goto Bail;
-
+    if (pFont->info.firstCol > pFont->info.lastCol ||
+       pFont->info.firstRow > pFont->info.lastRow ||
+       pFont->info.lastCol-pFont->info.firstCol > 255) goto Bail;
+
     nencoding = (pFont->info.lastCol - pFont->info.firstCol + 1) *
        (pFont->info.lastRow - pFont->info.firstRow + 1);

@@ -726,7 +729,10 @@
     pFontInfo->lastRow = pcfGetINT16(file, format);
     pFontInfo->defaultCh = pcfGetINT16(file, format);
     if (IS_EOF(file)) goto Bail;
-
+    if (pFontInfo->firstCol > pFontInfo->lastCol ||
+       pFontInfo->firstRow > pFontInfo->lastRow ||
+       pFontInfo->lastCol-pFontInfo->firstCol > 255) goto Bail;
+
     nencoding = (pFontInfo->lastCol - pFontInfo->firstCol + 1) *
        (pFontInfo->lastRow - pFontInfo->firstRow + 1);
