--- a/include/input.h	Tue Feb 24 12:58:09 2015
+++ b/include/input.h	Tue Feb 24 12:58:19 2015
@@ -605,7 +605,7 @@
                           xEvent *xE,
                           WindowPtr pWin, Window child, Bool calcChild);
 extern Bool PointInBorderSize(WindowPtr pWin, int x, int y);
-extern WindowPtr XYToWindow(SpritePtr pSprite, int x, int y);
+extern _X_EXPORT WindowPtr XYToWindow(SpritePtr pSprite, int x, int y);
 extern int EventIsDeliverable(DeviceIntPtr dev, int evtype, WindowPtr win);
 extern Bool ActivatePassiveGrab(DeviceIntPtr dev, GrabPtr grab,
                                 InternalEvent *ev, InternalEvent *real_event);
--- a/glx/glxdriswrast.c	Wed Feb 25 09:30:45 2015
+++ b/glx/glxdriswrast.c	Wed Feb 25 09:29:27 2015
@@ -479,7 +479,7 @@
     return NULL;
 }
 
-_X_EXPORT __GLXprovider __glXDRISWRastProvider = {
+__GLXprovider __glXDRISWRastProvider = {
     __glXDRIscreenProbe,
     "DRISWRAST",
     NULL
--- a/glx/glxscreens.c	Wed Feb 25 09:33:52 2015
+++ b/glx/glxscreens.c	Wed Feb 25 09:34:06 2015
@@ -193,7 +193,7 @@
     return dixLookupPrivate(&pScreen->devPrivates, glxScreenPrivateKey);
 }
 
-_X_EXPORT void
+void
 GlxSetVisualConfigs(int nconfigs, void *configs, void **privates)
 {
     /* We keep this stub around for the DDX drivers that still

--- a/hw/xfree86/Makefile.am	Tue Apr 21 12:36:19 2015
+++ b/hw/xfree86/Makefile.am	Tue Apr 21 12:38:59 2015
@@ -135,7 +135,7 @@
 EXTRA_DIST += sdksyms.sh
 
 sdksyms.dep sdksyms.c: sdksyms.sh
-	$(AM_V_GEN)CPP='$(CPP)' AWK='$(AWK)' $(SHELL) $(srcdir)/sdksyms.sh $(top_srcdir) $(CFLAGS) $(AM_CFLAGS) $(AM_CPPFLAGS) -I/usr/include -I/usr/include/xorg -I/usr/include/pixman-1 -I/usr/include/X11/dri
+	$(AM_V_GEN)CPP='$(CPP)' AWK='$(AWK)' SED='$(SED)' $(SHELL) $(srcdir)/sdksyms.sh $(top_srcdir) $(CFLAGS) $(AM_CFLAGS) $(AM_CPPFLAGS) -I/usr/include -I/usr/include/xorg -I/usr/include/pixman-1 -I/usr/include/X11/dri
 
 SDKSYMS_DEP = sdksyms.dep
 include $(SDKSYMS_DEP)
--- a/hw/xfree86/sdksyms.sh	Mon Apr 20 14:13:28 2015
+++ b/hw/xfree86/sdksyms.sh	Tue Apr 21 12:41:07 2015
@@ -314,7 +314,13 @@
 LC_ALL=C
 export LC_ALL
 ${CPP:-cpp} "$@" sdksyms.c > /dev/null || exit $?
-${CPP:-cpp} "$@" sdksyms.c | ${AWK:-awk} -v topdir=$topdir '
+# Some cpps add spaces in the middle of __attribute__((visibility(...)))
+# so we need to remove them for the following awk script to work.
+# Changes by alanc upstream
+${CPP:-cpp} "$@" sdksyms.c | \
+    ${SED:-sed} -e 's/( /(/g' -e 's/_ (/_(/g' -e 's/ )/)/g' \
+    -e 's/visibility (/visibility(/' | \
+    ${AWK:-awk} -v topdir=$topdir '
 BEGIN {
     sdk = 0;
     print("/*");

--- a/hw/xfree86/drivers/modesetting/Makefile.am	Thu Feb 26 13:47:39 2015
+++ b/hw/xfree86/drivers/modesetting/Makefile.am	Thu Feb 26 13:48:23 2015
@@ -39,7 +39,9 @@
 
 modesetting_drv_la_LTLIBRARIES = modesetting_drv.la
 modesetting_drv_la_LDFLAGS = -module -avoid-version
-modesetting_drv_la_LIBADD = $(UDEV_LIBS) $(DRM_LIBS)
+modesetting_drv_la_LDFLAGS += $(XORG_EXTERNS_FLAG)
+modesetting_drv_la_LIBADD = $(UDEV_LIBS) $(DRM_LIBS) $(LIBDRM_LIBS) $(top_builddir)/fb/libfb.la $(top_builddir)/miext/shadow/libshadow.la
+
 modesetting_drv_ladir = @moduledir@/drivers
 
 modesetting_drv_la_SOURCES = \
--- a/hw/xfree86/utils/cvt/cvt.c	Sun May  3 12:41:51 2015
+++ b/hw/xfree86/utils/cvt/cvt.c	Mon May  4 09:13:02 2015
@@ -39,6 +39,13 @@
 {
     abort();
 }
+
+/* added to quieten LogMessage undefined symbol error while compiling with Studio 12.4 */
+#include "os.h"
+_X_HIDDEN void
+LogMessage(MessageType type, const char *format, ...) {
+	abort();
+}
 #endif
 
 #include "xf86.h"
--- a/hw/dmx/config/xdmxconfig.c	Mon May  4 10:01:10 2015
+++ b/hw/dmx/config/xdmxconfig.c	Mon May  4 10:01:43 2015
@@ -63,6 +63,13 @@
 {
     abort();
 }
+
+/* added to quieten LogMessage undefined symbol error while compiling with Studio 12.4 */
+#include "os.h"
+_X_HIDDEN void
+LogMessage(MessageType type, const char *format, ...) {
+       abort();
+}
 #endif
 
 #define DMX_INFO "xdmxconfig v0.9\nCopyright 2002 Red Hat Inc.\n"
