--- hw/xfree86/os-support/solaris/sun_kbd.c     2007-07-24 16:45:36.000000000 -0700
+++ hw/xfree86/os-support/solaris/sun_kbd.c     2007-07-25 10:53:36.000000000 -0700
@@ -443,6 +443,10 @@
 sunKbdSoundBell(sunKbdPrivPtr priv, int loudness, int pitch, int duration)
 {
     int	kbdCmd, i;
+#ifdef KIOCMKTONE
+    int cycles;
+    int mktonevalue;
+#endif
 
     if (loudness && pitch)
     {
@@ -454,6 +458,32 @@
 	}
 #endif
 	
+#ifdef KIOCMKTONE
+	if (pitch == 0)
+	    cycles = UINT16_MAX;
+	else if (pitch >= UINT16_MAX)
+	    cycles = 0;
+	else {
+	    cycles = (PIT_HZ + pitch / 2) / pitch;
+	    if (cycles > UINT16_MAX)
+		cycles = UINT16_MAX;
+	}
+
+	mktonevalue = cycles | (((duration * loudness * 20) / 1000) << 16);
+
+	errno = 0;
+	SYSCALL(i = ioctl (priv->kbdFD, KIOCMKTONE, mktonevalue));
+	if (i == 0)
+	    return;
+
+	if (errno != EINVAL) {
+	    if (errno != EAGAIN)
+		xf86Msg(X_ERROR, "%s: Failed to activate bell: %s\n",
+		    priv->devName, strerror(errno));
+	    return;
+	}
+#endif
+
  	kbdCmd = KBD_CMD_BELL;
 		
 	SYSCALL(i = ioctl (priv->kbdFD, KIOCCMD, &kbdCmd));
