diff -urp -x '*~' -x '*.orig' hw/xfree86/common/xf86RandR.c hw/xfree86/common/xf86RandR.c
--- hw/xfree86/common/xf86RandR.c	2008-09-24 20:24:55.563464000 -0700
+++ hw/xfree86/common/xf86RandR.c	2008-09-24 20:25:14.259832000 -0700
@@ -72,7 +72,7 @@ xf86RandRGetInfo (ScreenPtr pScreen, Rot
 
     *rotations = RR_Rotate_0;
 
-    for (mode = scrp->modes; ; mode = mode->next)
+    for (mode = scrp->modes; mode!=NULL ; mode = mode->next)
     {
 	int refresh = xf86RandRModeRefresh (mode);
 
