--- GL/symlink-mesa.sh	Mon Jul  7 11:37:13 2008
+++ GL/symlink-mesa.sh	Mon Jul  7 12:56:34 2008
@@ -85,15 +85,15 @@
     done
 }
 
-symlink_mesa_ac() {
-    src_dir src/mesa/array_cache
-    dst_dir mesa/array_cache
+#symlink_mesa_ac() {
+#    src_dir src/mesa/array_cache
+#    dst_dir mesa/array_cache
+#
+#    for src in $REAL_SRC_DIR/*.c $REAL_SRC_DIR/*.h; do
+#        action `basename $src`
+#    done
+#}
 
-    for src in $REAL_SRC_DIR/*.c $REAL_SRC_DIR/*.h; do
-        action `basename $src`
-    done
-}
-
 symlink_mesa_swrast() {
     src_dir src/mesa/swrast
     dst_dir mesa/swrast
@@ -121,6 +121,15 @@
     done
 }
 
+symlink_mesa_vbo() {
+    src_dir src/mesa/vbo
+    dst_dir mesa/vbo
+
+    for src in $REAL_SRC_DIR/*.c $REAL_SRC_DIR/*.h; do
+        action `basename $src`
+    done
+}
+
 symlink_mesa_shader() {
     src_dir src/mesa/shader
     dst_dir mesa/shader
@@ -172,6 +181,10 @@
     action xm_span.c
     action xm_tri.c
     action xmesaP.h
+    action xm_image.c
+    action xm_image.h
+    action xm_glide.c
+    action xm_glide.h
 
     # another hack
     src_dir src/mesa/drivers/common
@@ -203,10 +216,11 @@
 symlink_mesa() {
     symlink_mesa_main
     symlink_mesa_math
-    symlink_mesa_ac
+#symlink_mesa_ac
     symlink_mesa_swrast
     symlink_mesa_ss
     symlink_mesa_tnl
+    symlink_mesa_vbo
     symlink_mesa_shader
     symlink_mesa_shader_grammar
     symlink_mesa_shader_slang
@@ -226,10 +240,10 @@
 
     action indirect_size.h
 
-    src_dir src/mesa/drivers/dri/common
+    src_dir src/glx/x11
 
-    action glcontextmodes.c
-    action glcontextmodes.h
+    action glcontextmodes.c
+    action glcontextmodes.h
 
     src_dir src/mesa/glapi
 

--- GL/mesa/X/Makefile.am	Mon Jul  7 13:14:00 2008
+++ GL/mesa/X/Makefile.am	Mon Jul  7 13:18:30 2008
@@ -2,7 +2,6 @@
 
 INCLUDES = -I@MESA_SOURCE@/include \
            -I../X \
-           -I../array_cache \
            -I../glapi \
            -I../main \
            -I../math \
@@ -10,6 +9,7 @@
            -I../swrast \
            -I../swrast_setup \
            -I../tnl \
+           -I../vbo \
            -I.. \
            -I../../glx \
            -I$(top_srcdir)/GL/glx \
@@ -36,4 +36,6 @@
                   xm_line.c \
                   xm_span.c \
                   xm_tri.c \
+                  xm_image.c \
+                  xm_glide.c \
 		  drivers/common/driverfuncs.c


--- GL/mesa/Makefile.am	Mon Jul  7 13:10:15 2008
+++ GL/mesa/Makefile.am	Mon Jul  7 13:13:10 2008
@@ -1,4 +1,4 @@
-SUBDIRS = main math array_cache swrast swrast_setup tnl shader X glapi
+SUBDIRS = main math swrast swrast_setup tnl vbo shader X glapi
 
 noinst_LTLIBRARIES = libGLcore.la
 
@@ -5,10 +5,10 @@
 libGLcore_la_SOURCES = dummy.c
 libGLcore_la_LIBADD = main/libmain.la \
                       math/libmath.la \
-                      array_cache/libac.la \
                       swrast/libswrast.la \
                       swrast_setup/libss.la \
                       tnl/libtnl.la \
+                      vbo/libvbo.la \
                       shader/libshader.la \
 		      shader/grammar/libgrammar.la \
 		      shader/slang/libslang.la \


--- GL/mesa/glapi/Makefile.am	Mon Jul  7 13:18:29 2008
+++ GL/mesa/glapi/Makefile.am	Mon Jul  7 13:21:01 2008
@@ -7,7 +7,6 @@
 
 INCLUDES = -I@MESA_SOURCE@/include \
            -I../X \
-           -I../array_cache \
            -I../glapi \
            -I../main \
            -I../math \
@@ -15,6 +14,7 @@
            -I../swrast \
            -I../swrast_setup \
            -I../tnl \
+           -I../vbo \
            -I.. \
            -I$(top_srcdir)/hw/xfree86/os-support
 


--- GL/mesa/math/Makefile.am	Mon Jul  7 13:21:51 2008
+++ GL/mesa/math/Makefile.am	Mon Jul  7 13:24:36 2008
@@ -7,7 +7,6 @@
 
 INCLUDES = -I@MESA_SOURCE@/include \
            -I../X \
-           -I../array_cache \
            -I../glapi \
            -I../main \
            -I../math \
@@ -15,6 +14,7 @@
            -I../swrast \
            -I../swrast_setup \
            -I../tnl \
+           -I../vbo \
            -I.. \
            -I$(top_srcdir)/hw/xfree86/os-support
 

--- GL/mesa/shader/Makefile.am	Mon Jul  7 13:23:13 2008
+++ GL/mesa/shader/Makefile.am	Mon Jul  7 13:28:35 2008
@@ -9,7 +9,6 @@
 
 INCLUDES = -I@MESA_SOURCE@/include \
            -I../X \
-           -I../array_cache \
            -I../glapi \
            -I../main \
            -I../math \
@@ -19,6 +18,7 @@
            -I../swrast \
            -I../swrast_setup \
            -I../tnl \
+           -I../vbo \
            -I.. \
            -I$(top_srcdir)/hw/xfree86/os-support
 
@@ -28,8 +28,13 @@
                       atifragshader.c \
                       nvfragparse.c \
                       nvprogram.c \
-                      nvvertexec.c \
 		      nvvertparse.c \
                       program.c \
-                      shaderobjects.c \
-		      shaderobjects_3dlabs.c
+                      prog_debug.c \
+                      prog_execute.c \
+                      prog_instruction.c \
+                      prog_parameter.c \
+                      prog_print.c \
+                      prog_statevars.c \
+                      programopt.c \
+		      shader_api.c


--- GL/mesa/tnl/Makefile.am	Mon Jul  7 13:28:59 2008
+++ GL/mesa/tnl/Makefile.am	Mon Jul  7 13:34:02 2008
@@ -7,7 +7,6 @@
 
 INCLUDES = -I@MESA_SOURCE@/include \
            -I../X \
-           -I../array_cache \
            -I../glapi \
            -I../main \
            -I../math \
@@ -16,19 +15,13 @@
            -I../swrast \
            -I../swrast_setup \
            -I../tnl \
+           -I../vbo \
            -I.. \
            -I$(top_srcdir)/hw/xfree86/os-support
 
-nodist_libtnl_la_SOURCES = t_array_api.c \
-                      t_array_import.c \
-                      t_context.c \
+nodist_libtnl_la_SOURCES = t_context.c \
                       t_pipeline.c \
-                      t_save_api.c \
-                      t_save_loopback.c \
-                      t_save_playback.c \
-                      t_vb_arbprogram.c \
-                      t_vb_arbprogram_sse.c \
-                      t_vb_arbshader.c \
+                      t_draw.c \
                       t_vb_cull.c \
                       t_vb_fog.c \
                       t_vb_light.c \
@@ -42,9 +35,4 @@
                       t_vertex.c \
                       t_vertex_generic.c \
                       t_vertex_sse.c \
-                      t_vp_build.c \
-                      t_vtx_api.c \
-                      t_vtx_eval.c \
-                      t_vtx_exec.c \
-                      t_vtx_generic.c \
-                      t_vtx_x86.c
+                      t_vp_build.c


--- GL/mesa/shader/slang/Makefile.am	Mon Jul  7 13:47:04 2008
+++ GL/mesa/shader/slang/Makefile.am	Mon Jul  7 13:51:53 2008
@@ -8,7 +8,6 @@
 INCLUDES = -I@MESA_SOURCE@/include \
            -I../grammar \
            -I../../X \
-           -I../../array_cache \
            -I../../glapi \
            -I../../main \
            -I../../math \
@@ -16,26 +15,28 @@
            -I../../swrast \
            -I../../swrast_setup \
            -I../../tnl \
+           -I../../vbo \
            -I../.. \
            -I$(top_srcdir)/hw/xfree86/os-support
 
-nodist_libslang_la_SOURCES = slang_analyse.c \
-		      slang_assemble_assignment.c \
-		      slang_assemble.c \
-		      slang_assemble_conditional.c \
-		      slang_assemble_constructor.c \
-		      slang_assemble_typeinfo.c \
+nodist_libslang_la_SOURCES = slang_builtin.c \
+		      slang_codegen.c \
 		      slang_compile.c \
 		      slang_compile_function.c \
 		      slang_compile_operation.c \
 		      slang_compile_struct.c \
 		      slang_compile_variable.c \
-		      slang_execute.c \
-		      slang_execute_x86.c \
-		      slang_export.c \
-		      slang_library_texsample.c \
+		      slang_emit.c \
+		      slang_ir.c \
+		      slang_label.c \
 		      slang_library_noise.c \
 		      slang_link.c \
+		      slang_log.c \
+		      slang_mem.c \
 		      slang_preprocess.c \
+		      slang_print.c \
+		      slang_simplify.c \
 		      slang_storage.c \
+		      slang_typeinfo.c \
+		      slang_vartable.c \
 		      slang_utility.c


--- GL/mesa/shader/grammar/Makefile.am	Mon Jul  7 13:50:10 2008
+++ GL/mesa/shader/grammar/Makefile.am	Mon Jul  7 13:52:39 2008
@@ -7,7 +7,6 @@
 
 INCLUDES = -I@MESA_SOURCE@/include \
            -I../../X \
-           -I../../array_cache \
            -I../../glapi \
            -I../../main \
            -I../../math \
@@ -15,6 +14,7 @@
            -I../../swrast \
            -I../../swrast_setup \
            -I../../tnl \
+           -I../../vbo \
            -I../.. \
            -I$(top_srcdir)/hw/xfree86/os-support
 


--- GL/glx/Makefile.am	Mon Jul  7 13:58:38 2008
+++ GL/glx/Makefile.am	Mon Jul  7 14:02:38 2008
@@ -28,10 +28,14 @@
 	-I$(top_srcdir)/mi
 
 
+#nodist_libglx_la_SOURCES = indirect_size.h \
+#			   glapi.c \
+#			   glcontextmodes.c \
+#			   glcontextmode.h \
+#			   glthread.c
+
 nodist_libglx_la_SOURCES = indirect_size.h \
 			   glapi.c \
-			   glcontextmodes.c \
-			   glcontextmode.h \
 			   glthread.c
 
 libglxdri_la_SOURCES = \
@@ -53,6 +57,9 @@
         glxserver.h \
         glxutil.c \
         glxutil.h \
+        glcontextmodes.c \
+        glcontextmodes.h \
+        glxutil.h \
         indirect_dispatch.c \
         indirect_dispatch.h \
         indirect_dispatch_swap.c \


--- configure.ac	Mon Jul  7 17:03:19 2008
+++ configure.ac	Mon Jul  7 17:03:41 2008
@@ -1723,7 +1723,6 @@
 GL/Makefile
 GL/glx/Makefile
 GL/mesa/Makefile
-GL/mesa/array_cache/Makefile
 GL/mesa/glapi/Makefile
 GL/mesa/main/Makefile
 GL/mesa/math/Makefile
@@ -1733,6 +1732,7 @@
 GL/mesa/swrast/Makefile
 GL/mesa/swrast_setup/Makefile
 GL/mesa/tnl/Makefile
+GL/mesa/vbo/Makefile
 GL/mesa/X/Makefile
 include/Makefile
 afb/Makefile


--- GL/mesa/swrast/Makefile.am	Sun Sep 17 23:04:17 2006
+++ GL/mesa/swrast/Makefile.am	Tue Jul  8 10:30:31 2008
@@ -7,7 +7,6 @@
 
 INCLUDES = -I@MESA_SOURCE@/include \
            -I../X \
-           -I../array_cache \
            -I../glapi \
            -I../main \
            -I../math \
@@ -17,6 +16,7 @@
            -I../swrast \
            -I../swrast_setup \
            -I../tnl \
+           -I../vbo \
            -I.. \
            -I$(top_srcdir)/hw/xfree86/os-support
 
@@ -24,7 +24,6 @@
                       s_aatriangle.c \
                       s_accum.c \
                       s_alpha.c \
-                      s_arbshader.c \
                       s_atifragshader.c \
                       s_bitmap.c \
                       s_blend.c \
@@ -40,10 +39,10 @@
                       s_lines.c \
                       s_logic.c \
                       s_masking.c \
-                      s_nvfragprog.c \
                       s_points.c \
                       s_readpix.c \
                       s_span.c \
+                      s_fragprog.c \
                       s_stencil.c \
 		      s_texcombine.c \
 		      s_texfilter.c \


--- GL/mesa/X/xf86glx_util.c	Sun Sep 17 23:04:17 2006
+++ GL/mesa/X/xf86glx_util.c	Mon Jul  7 23:16:17 2008
@@ -47,7 +47,7 @@
 #endif
 
 #define ROUNDUP(nbytes, pad) ((((nbytes) + ((pad)-1)) / (pad)) * ((pad)>>3))
-
+#if 0
 XMesaImage *XMesaCreateImage(int bitsPerPixel, int width, int height, char *data)
 {
     XMesaImage *image;
@@ -102,6 +102,7 @@
     }
     return 0;
 }
+#endif
 
 #ifndef XMESA_USE_PUTPIXEL_MACRO
 void XMesaPutPixel(XMesaImage *image, int x, int y, unsigned long pixel)

--- GL/mesa/main/Makefile.am	Fri Nov 17 09:47:26 2006
+++ GL/mesa/main/Makefile.am	Tue Jul  8 10:32:41 2008
@@ -7,7 +7,6 @@
 
 INCLUDES = -I@MESA_SOURCE@/include \
            -I../X \
-           -I../array_cache \
            -I../glapi \
            -I../main \
            -I../math \
@@ -15,6 +14,7 @@
            -I../swrast \
            -I../swrast_setup \
            -I../tnl \
+           -I../vbo \
            -I.. \
            -I$(top_srcdir)/hw/xfree86/os-support
 
@@ -62,7 +62,7 @@
                     matrix.c \
                     mipmap.c \
                     mm.c \
-                    occlude.c \
+                    queryobj.c \
                     pixel.c \
                     points.c \
                     polygon.c \
@@ -69,6 +69,7 @@
                     rastpos.c \
                     rbadaptors.c \
                     renderbuffer.c \
+                    shaders.c \
                     state.c \
                     stencil.c \
                     texcompress.c \

--- GL/mesa/Makefile.in	Thu Apr 19 18:24:24 2007
+++ GL/mesa/Makefile.in	Mon Aug 11 11:32:50 2008
@@ -55,7 +55,7 @@
 CONFIG_CLEAN_FILES =
 LTLIBRARIES = $(noinst_LTLIBRARIES)
 libGLcore_la_DEPENDENCIES = main/libmain.la math/libmath.la \
-	array_cache/libac.la swrast/libswrast.la swrast_setup/libss.la \
+	vbo/libvbo.la swrast/libswrast.la swrast_setup/libss.la \
 	tnl/libtnl.la shader/libshader.la shader/grammar/libgrammar.la \
 	shader/slang/libslang.la X/libX.la
 am_libGLcore_la_OBJECTS = dummy.lo
@@ -540,12 +540,12 @@
 target_alias = @target_alias@
 xglmoduledir = @xglmoduledir@
 xpconfigdir = @xpconfigdir@
-SUBDIRS = main math array_cache swrast swrast_setup tnl shader X glapi
+SUBDIRS = main math swrast swrast_setup vbo tnl shader X glapi
 noinst_LTLIBRARIES = libGLcore.la
 libGLcore_la_SOURCES = dummy.c
 libGLcore_la_LIBADD = main/libmain.la \
                       math/libmath.la \
-                      array_cache/libac.la \
+                      vbo/libvbo.la \
                       swrast/libswrast.la \
                       swrast_setup/libss.la \
                       tnl/libtnl.la \
