#
# remove this when xorg 1.7 is integrated
#
--- ./Xext/shm.c.orig	Wed Sep 30 13:25:01 2009
+++ ./Xext/shm.c	Wed Sep 30 13:25:41 2009
@@ -500,7 +500,7 @@
 {
     PixmapPtr pPixmap;
   
-    pPixmap = GetScratchPixmapHeader(dst->pScreen, w, h, depth,
+    pPixmap = GetScratchPixmapHeader(dst->pScreen, sw, sh, depth,
 				     BitsPerPixel(depth),
 				     PixmapBytePad(w, depth),
 				     data);
