--- a/hw/xfree86/Makefile.am	Wed Sep 12 10:56:31 2012
+++ b/hw/xfree86/Makefile.am	Wed Sep 12 11:02:17 2012
@@ -25,10 +25,10 @@
 INT10_SUBDIR = int10
 endif
 
-SUBDIRS = common ddc x86emu $(INT10_SUBDIR) os-support parser \
-	  ramdac $(VGAHW_SUBDIR) loader modes $(DRI_SUBDIR) \
-	  $(DRI2_SUBDIR) . $(VBE_SUBDIR) i2c dixmods \
-	  fbdevhw shadowfb exa $(XF86UTILS_SUBDIR) doc man
+SUBDIRS = common ddc i2c x86emu $(INT10_SUBDIR) fbdevhw os-support parser \
+	  ramdac shadowfb $(VBE_SUBDIR) $(VGAHW_SUBDIR) loader \
+	  $(DRI_SUBDIR) $(DRI2_SUBDIR) . dixmods exa modes \
+	  $(XF86UTILS_SUBDIR) doc man
 
 DIST_SUBDIRS = common ddc i2c x86emu int10 fbdevhw os-support \
                parser ramdac shadowfb vbe vgahw \
@@ -38,7 +38,7 @@
 bin_PROGRAMS = Xorg
 nodist_Xorg_SOURCES = sdksyms.c
 
-AM_CFLAGS = $(DIX_CFLAGS) @XORG_CFLAGS@
+AM_CFLAGS = $(DIX_CFLAGS) $(PIXMAN_CFLAGS) @XORG_CFLAGS@
 INCLUDES = $(XORG_INCS) -I$(srcdir)/parser -I$(top_srcdir)/miext/cw \
 	-I$(srcdir)/ddc -I$(srcdir)/i2c -I$(srcdir)/modes -I$(srcdir)/ramdac \
 	-I$(srcdir)/dri -I$(srcdir)/dri2
--- a/hw/xfree86/i2c/Makefile.am	Fri Sep 14 10:37:31 2012
+++ b/hw/xfree86/i2c/Makefile.am	Fri Sep 14 10:38:21 2012
@@ -22,22 +22,29 @@
 # i2c drivers
 #
 bt829_drv_la_LDFLAGS = -module -avoid-version $(LD_NO_UNDEFINED_FLAG)
+bt829_drv_la_LDFLAGS += $(XORG_EXTERNS_FLAG)
 bt829_drv_la_SOURCES = bt829.c bt829.h bt829_module.c
 
 fi1236_drv_la_LDFLAGS = -module -avoid-version $(LD_NO_UNDEFINED_FLAG)
+fi1236_drv_la_LDFLAGS += $(XORG_EXTERNS_FLAG)
 fi1236_drv_la_SOURCES = fi1236.c fi1236.h fi1236_module.c tda9885.c
 
 msp3430_drv_la_LDFLAGS = -module -avoid-version $(LD_NO_UNDEFINED_FLAG)
+msp3430_drv_la_LDFLAGS += $(XORG_EXTERNS_FLAG)
 msp3430_drv_la_SOURCES = msp3430.c msp3430.h msp3430_module.c
 
 tda8425_drv_la_LDFLAGS = -module -avoid-version $(LD_NO_UNDEFINED_FLAG)
+tda8425_drv_la_LDFLAGS += $(XORG_EXTERNS_FLAG)
 tda8425_drv_la_SOURCES = tda8425.c tda8425.h tda8425_module.c
 
 tda9850_drv_la_LDFLAGS = -module -avoid-version $(LD_NO_UNDEFINED_FLAG)
+tda9850_drv_la_LDFLAGS += $(XORG_EXTERNS_FLAG)
 tda9850_drv_la_SOURCES = tda9850.c tda9850.h tda9850_module.c
 
 tda9885_drv_la_LDFLAGS = -module -avoid-version $(LD_NO_UNDEFINED_FLAG)
+tda9885_drv_la_LDFLAGS += $(XORG_EXTERNS_FLAG)
 tda9885_drv_la_SOURCES = tda9885.c tda9885.h tda9885_module.c
 
 uda1380_drv_la_LDFLAGS = -module -avoid-version $(LD_NO_UNDEFINED_FLAG)
+uda1380_drv_la_LDFLAGS += $(XORG_EXTERNS_FLAG)
 uda1380_drv_la_SOURCES = uda1380.c uda1380.h uda1380_module.c
--- a/hw/xfree86/dixmods/Makefile.am	Wed Sep 12 11:02:30 2012
+++ b/hw/xfree86/dixmods/Makefile.am	Wed Sep 12 11:04:21 2012
@@ -18,7 +18,9 @@
            -I$(top_srcdir)/glx
 
 libfb_la_LDFLAGS = -module -avoid-version $(LD_NO_UNDEFINED_FLAG)
+libfb_la_LDFLAGS += $(XORG_EXTERNS_FLAG)
 libfb_la_LIBADD = $(top_builddir)/fb/libfb.la
+libfb_la_LIBADD += -lpixman-1
 libfb_la_SOURCES = $(top_builddir)/fb/fbcmap_mi.c fbmodule.c
 libfb_la_CFLAGS = $(AM_CFLAGS)
 
@@ -28,9 +30,12 @@
 libwfb_la_CFLAGS = $(AM_CFLAGS) -DFB_ACCESS_WRAPPER
 
 libglx_la_LDFLAGS = -module -avoid-version $(LD_NO_UNDEFINED_FLAG)
+libglx_la_LDFLAGS += $(XORG_EXTERNS_FLAG)
 libglx_la_LIBADD = $(top_builddir)/glx/libglx.la $(GLX_SYS_LIBS)
 if AIGLX_DRI_LOADER
 libglx_la_LIBADD += $(top_builddir)/glx/libglxdri.la
+libglx_la_LDFLAGS += -R$(extsmoduledir)
+libglx_la_LDFLAGS += $(LIBDRM_LIBS)
 if NO_UNDEFINED
 libglx_la_LIBADD += ../dri/libdri.la ../dri2/libdri2.la
 endif
@@ -39,6 +44,9 @@
 
 libshadow_la_LDFLAGS = -module -avoid-version $(LD_NO_UNDEFINED_FLAG)
 libshadow_la_LIBADD = $(top_builddir)/miext/shadow/libshadow.la
+libshadow_la_DEPENDENCIES = libfb.la
+libshadow_la_LDFLAGS += 
+libshadow_la_LDFLAGS += $(XORG_EXTERNS_FLAG)
 if NO_UNDEFINED
 libshadow_la_LIBADD += libfb.la
 endif
--- a/hw/xfree86/shadowfb/Makefile.am	Fri Sep 14 14:22:33 2012
+++ b/hw/xfree86/shadowfb/Makefile.am	Fri Sep 14 14:23:16 2012
@@ -8,3 +8,7 @@
 INCLUDES = $(XORG_INCS)
 
 AM_CFLAGS = $(DIX_CFLAGS) $(XORG_CFLAGS)
+
+libshadowfb_la_LDFLAGS += -z ignore $(XORG_EXTERNS_FLAG)
+libshadowfb_la_LIBADD = $(XSERVER_SYS_LIBS) -lXfont
+
--- a/hw/xfree86/vbe/Makefile.am	Fri Sep 14 14:24:05 2012
+++ b/hw/xfree86/vbe/Makefile.am	Fri Sep 14 14:24:16 2012
@@ -11,3 +11,7 @@
 
 INCLUDES = $(XORG_INCS) -I$(srcdir)/../ddc -I$(srcdir)/../i2c \
            -I$(srcdir)/../int10
+
+libvbe_la_LDFLAGS += $(XORG_EXTERNS_FLAG)
+libvbe_la_LIBADD = ../int10/libint10.la
+
--- a/hw/xfree86/Makefile.am	Fri Sep 14 14:24:31 2012
+++ b/hw/xfree86/Makefile.am	Fri Sep 14 14:25:47 2012
@@ -123,6 +123,9 @@
 i2c/libi2c.la:
 	$(AM_V_at)cd i2c && $(MAKE) libi2c.la
 
+modes/libxf86modes.la:
+	$(AM_V_at)cd modes && $(MAKE) libxf86modes.la
+
 dixmods/libdixmods.la:
 	$(AM_V_at)cd dixmods && $(MAKE) libdixmods.la
 
