--- hw/xfree86/dixmods/Makefile.am	2008-09-26 21:45:31.528930000 -0700
+++ hw/xfree86/dixmods/Makefile.am	2008-09-27 09:30:57.338529000 -0700
@@ -70,7 +70,7 @@
 libdbe_la_SOURCES = dbemodule.c
 
 libfb_la_LDFLAGS = -avoid-version
-libfb_la_LIBADD = $(top_builddir)/fb/libfb.la
+libfb_la_LIBADD = $(top_builddir)/fb/libfb.la -lpixman-1
 libfb_la_SOURCES = $(top_builddir)/fb/fbcmap_mi.c fbmodule.c
 libfb_la_CFLAGS = $(AM_CFLAGS)
 
--- hw/xfree86/int10/Makefile.am	2007-10-17 07:01:23.000000000 -0700
+++ hw/xfree86/int10/Makefile.am	2008-09-30 14:20:57.961539000 -0700
@@ -42,3 +42,5 @@
 endif
 
 EXTRA_DIST = xf86x86emu.h INT10.HOWTO
+
+libint10_la_LIBADD = $(PCIACCESS_LIBS)
