--- hw/xfree86/common/xf86Events.c	Sat Oct 20 10:14:20 2007
+++ hw/xfree86/common/xf86Events.c	Sat Oct 20 10:15:59 2007
@@ -1456,8 +1456,32 @@
 #endif
 
   xorg_backtrace();
-    
   FatalError("Caught signal %d.  Server aborting\n", signo);
+
+/* CR 6545884 If server crashes keyboard left in un-encoded mode*/
+#ifdef sun
+        {
+            int kbd_fd,console_fd;
+            int translate=TR_ASCII;
+
+            Error ("Resetting keyboard\n");
+            /* Check if on console, since kbd_mode works on console only */
+            if (((console_fd = open("/dev/console", O_RDWR | O_NDELAY, 0)) != -1)) { 
+                if ((kbd_fd = open("/dev/kbd", O_RDONLY, 0)) < 0) {
+                    Error("/dev/kbd open failed\n");
+                } else {
+                    if (ioctl(kbd_fd, KIOCTRANS, (caddr_t) &translate)) {
+                        Error("Could not set /dev/kbd back to ASCII mode\n");
+                    }
+                    close(kbd_fd);
+                }
+            close (console_fd);
+            }
+            else
+                Error ("Not on Console not Resetting keyboard\n");
+        }
+#endif
+    
 }
 
 #ifdef MEMDEBUG
