diff --git a/record/record.c b/record/record.c
index facaebb..8256738 100644
--- a/record/record.c
+++ b/record/record.c
@@ -1470,6 +1470,9 @@ typedef struct
     short first, last;	/* if for extension, major opcode interval */
 } SetInfoRec, *SetInfoPtr;
 
+#undef ERR /* protect against #define ERR in <sys/regset.h> on Solaris x86,
+	      which gets included indirectly via #include <stdlib.h> */
+
 /* These constant are used to index into an array of SetInfoRec. */
 enum {REQ,	/* set info for requests */
       REP,	/* set info for replies */
