diff -urp -x '*~' -x '*.orig' hw/xfree86/dri/dri.c hw/xfree86/dri/dri.c
--- hw/xfree86/dri/dri.c	2008-10-17 15:49:03.916893000 -0700
+++ hw/xfree86/dri/dri.c	2008-10-17 15:49:21.389521000 -0700
@@ -2466,7 +2466,11 @@ int drmInstallSIGIOHandler(int fd, void 
     entry     = drmGetEntry(fd);
     entry->f  = f;
 
+#if defined(__sun) && defined(__SVR4)
+    return 1;
+#else
     return xf86InstallSIGIOHandler(fd, drmSIGIOHandler, 0);
+#endif
 }
 
 int drmRemoveSIGIOHandler(int fd)
@@ -2475,5 +2479,9 @@ int drmRemoveSIGIOHandler(int fd)
 
     entry->f = NULL;
 
+#if defined(__sun) && defined(__SVR4)
+    return 1;
+#else
     return xf86RemoveSIGIOHandler(fd);
+#endif
 }
