diff -urp -x '*~' -x '*.orig' hw/xfree86/dri/dri.c hw/xfree86/dri/dri.c
--- hw/xfree86/dri/dri.c	2009-05-08 17:45:21.396957000 -0700
+++ hw/xfree86/dri/dri.c	2009-05-08 17:45:31.288131000 -0700
@@ -2491,7 +2491,11 @@ int drmInstallSIGIOHandler(int fd, void 
     entry     = drmGetEntry(fd);
     entry->f  = f;
 
+#if defined(__sun) && defined(__SVR4)
+    return 1;
+#else
     return xf86InstallSIGIOHandler(fd, drmSIGIOHandler, 0);
+#endif
 }
 
 int drmRemoveSIGIOHandler(int fd)
@@ -2500,5 +2504,9 @@ int drmRemoveSIGIOHandler(int fd)
 
     entry->f = NULL;
 
+#if defined(__sun) && defined(__SVR4)
+    return 1;
+#else
     return xf86RemoveSIGIOHandler(fd);
+#endif
 }
