--- hw/xfree86/os-support/solaris/Makefile.am	2006-10-11 13:46:15.000000000 -0700
+++ hw/xfree86/os-support/solaris/Makefile.am	2006-10-18 09:59:34.248813000 -0700
@@ -23,9 +23,9 @@
 
 noinst_LTLIBRARIES = libsolaris.la
 libsolaris_la_SOURCES = sun_bios.c sun_init.c $(IO_SRC) $(KBD_SRCS) \
-	sun_mouse.c sun_vid.c $(AGP_SRC) \
+	sun_mouse.c sun_vid.c $(AGP_SRC) sun_apm.c \
 	$(srcdir)/../shared/libc_wrapper.c \
-	$(srcdir)/../shared/kmod_noop.c $(srcdir)/../shared/pm_noop.c \
+	$(srcdir)/../shared/kmod_noop.c \
 	$(srcdir)/../shared/posix_tty.c $(srcdir)/../shared/sigiostubs.c \
 	$(srcdir)/../shared/stdPci.c $(srcdir)/../shared/stdResource.c \
 	$(VTSW_SRC)

--- configure.ac
+++ configure.ac
@@ -1220,6 +1220,8 @@ #               ])
 			fi
 			if test "${OS_MINOR}" -lt 8 ; then
 				solaris_usl_console="yes"
+			else
+				XORG_CFLAGS="$XORG_CFLAGS -DXF86PM"
 			fi
 			;;
 		  *)

