--- hw/xfree86/common/extramodes	2006-10-11 13:46:14.000000000 -0700
+++ hw/xfree86/common/extramodes	2006-10-11 19:16:28.059134000 -0700
@@ -31,3 +31,50 @@
 # 2048x1536 @ 85Hz (VESA GTF) hsync: 137.0kHz
 Modeline "2048x1536" 388.04  2048 2216 2440 2832  1536 1537 1540 1612 -hsync +vsync
 
+//-------------------------------------------------------------------------
+// Copyright 2004 Sun Microsystems, Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, and/or sell copies of the Software, and to permit persons
+// to whom the Software is furnished to do so, provided that the above
+// copyright notice(s) and this permission notice appear in all copies of
+// the Software and that both the above copyright notice(s) and this
+// permission notice appear in supporting documentation.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT
+// OF THIRD PARTY RIGHTS. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+// HOLDERS INCLUDED IN THIS NOTICE BE LIABLE FOR ANY CLAIM, OR ANY SPECIAL
+// INDIRECT OR CONSEQUENTIAL DAMAGES, OR ANY DAMAGES WHATSOEVER RESULTING
+// FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
+// NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION
+// WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+//
+// Except as contained in this notice, the name of a copyright holder
+// shall not be used in advertising or otherwise to promote the sale, use
+// or other dealings in this Software without prior written authorization
+// of the copyright holder.
+//
+//-------------------------------------------------------------------------
+//
+// ident "@(#)sun_extramodes 1.4     05/11/08 SMI"
+//
+// Additional timing specs for Sun monitors
+
+// Skip this one for now as it breaks some other common 1600x1200@75Hz displays
+// # 1600x1200 @ 60Hz (Sun) hsync: 74.493kHz
+// Modeline "1600x1200" 132.3 1600 1608 1672 1776 1200 1205 1215 1239 +hsync +vsync
+
+# 1920x1080 @ 60Hz (Sun) hsync: 66.975kHz
+Modeline "1920x1080" 137.7 1920 1928 1992 2056 1080 1082 1086 1112 +hsync +vsync
+
+# 1920x1200 @ 60Hz (Sun 24.1" LCD) hsync: 73.54 kHz
+ModeLine "1920x1200" 151.200 1920 1928 1992 2056 1200 1202 1206 1234 +HSync +VSync
+
+# 1920x1200 @ 70.316Hz (Sun 24" CRT) hsync: 87.192 kHz
+ModeLine "1920x1200" 219.375 1920 1944 2172 2516 1200 1201 1204 1240 -HSync -VSync
+
