diff -urp -x '*~' -x '*.orig' hw/xfree86/common/xf86Config.c hw/xfree86/common/xf86Config.c
--- hw/xfree86/common/xf86Config.c	2010-05-03 17:47:58.000000000 -0700
+++ hw/xfree86/common/xf86Config.c	2010-10-20 20:27:00.062592706 -0700
@@ -77,9 +77,9 @@ extern DeviceAssocRec mouse_assoc;
  */
 #ifndef ROOT_CONFIGPATH
 #define ROOT_CONFIGPATH	"%A," "%R," \
-			"/etc/X11/%R," "%P/etc/X11/%R," \
+			"/etc/X11/%R," "%P/lib/X11/%R," \
 			"%E," "%F," \
-			"/etc/X11/%F," "%P/etc/X11/%F," \
+			"/etc/X11/%F," "%P/lib/X11/%F," \
 			"/etc/X11/%X-%M," "/etc/X11/%X," "/etc/%X," \
 			"%P/etc/X11/%X.%H," "%P/etc/X11/%X-%M," \
 			"%P/etc/X11/%X," \
@@ -87,8 +87,8 @@ extern DeviceAssocRec mouse_assoc;
 			"%P/lib/X11/%X"
 #endif
 #ifndef USER_CONFIGPATH
-#define USER_CONFIGPATH	"/etc/X11/%S," "%P/etc/X11/%S," \
-			"/etc/X11/%G," "%P/etc/X11/%G," \
+#define USER_CONFIGPATH	"/etc/X11/%S," "%P/lib/X11/%S," \
+			"/etc/X11/%G," "%P/lib/X11/%G," \
 			"/etc/X11/%X-%M," "/etc/X11/%X," "/etc/%X," \
 			"%P/etc/X11/%X.%H," "%P/etc/X11/%X-%M," \
 			"%P/etc/X11/%X," \
diff -urp -x '*~' -x '*.orig' hw/xfree86/common/xf86Globals.c hw/xfree86/common/xf86Globals.c
--- hw/xfree86/common/xf86Globals.c	2009-11-22 16:35:01.000000000 -0800
+++ hw/xfree86/common/xf86Globals.c	2009-12-02 18:57:45.196668578 -0800
@@ -143,7 +143,8 @@ xf86InfoRec xf86Info = {
 #endif
 };
 const char *xf86ConfigFile = NULL;
-const char *xf86ModulePath = DEFAULT_MODULE_PATH;
+/* Sun: add old module path for compatibility with out-of-tree driver packages */
+const char *xf86ModulePath = DEFAULT_MODULE_PATH ",/usr/X11/lib/modules/";
 MessageType xf86ModPathFrom = X_DEFAULT;
 const char *xf86LogFile = DEFAULT_LOGPREFIX;
 MessageType xf86LogFileFrom = X_DEFAULT;
diff -urp -x '*~' -x '*.orig' hw/xfree86/doc/man/xorg.conf.man.pre hw/xfree86/doc/man/xorg.conf.man.pre
--- hw/xfree86/doc/man/xorg.conf.man.pre	2010-05-03 17:47:59.000000000 -0700
+++ hw/xfree86/doc/man/xorg.conf.man.pre	2010-10-20 20:27:00.063252935 -0700
@@ -28,9 +28,9 @@ server is started as a normal user:
 .RS 4
 .nf
 .IR /etc/X11/ <cmdline>
-.IR __projectroot__/etc/X11/ <cmdline>
+.IR __projectroot__/lib/X11/ <cmdline>
 .IB /etc/X11/ $XORGCONFIG
-.IB __projectroot__/etc/X11/ $XORGCONFIG
+.IB __projectroot__/lib/X11/ $XORGCONFIG
 .I /etc/X11/__xconfigfile__\-4
 .I /etc/X11/__xconfigfile__
 .I /etc/__xconfigfile__
@@ -62,10 +62,10 @@ search locations are as follows:
 .nf
 <cmdline>
 .IR /etc/X11/ <cmdline>
-.IR __projectroot__/etc/X11/ <cmdline>
+.IR __projectroot__/lib/X11/ <cmdline>
 .B $XORGCONFIG
 .IB /etc/X11/ $XORGCONFIG
-.IB __projectroot__/etc/X11/ $XORGCONFIG
+.IB __projectroot__/lib/X11/ $XORGCONFIG
 .I /etc/X11/__xconfigfile__\-4
 .I /etc/X11/__xconfigfile__
 .I /etc/__xconfigfile__
