diff -urp -x '*~' -x '*.orig' hw/xfree86/common/xf86Configure.c hw/xfree86/common/xf86Configure.c
--- hw/xfree86/common/xf86Configure.c	2009-01-09 09:47:03.000000000 -0800
+++ hw/xfree86/common/xf86Configure.c	2009-01-22 20:33:12.636229000 -0800
@@ -518,7 +518,8 @@ configureModuleSection (void)
             /* 'strstr(dFP,"/dir")' is meant as 'dFP =~ m(/dir\W)' */
     	    if (defaultFontPath && (
 		(strcmp(*el, "freetype")  == 0 &&
-		 strstr(defaultFontPath, "/TTF")) ||
+ 		 (strstr(defaultFontPath, "/TrueType") ||
+ 		  strstr(defaultFontPath, "/TTF"))) ||
     	        (strcmp(*el, "type1")  == 0 &&
 		 strstr(defaultFontPath, "/Type1")))) 
 	    	ptr->mod_load_lst = (XF86LoadPtr)xf86addListItem(
diff -urp -x '*~' -x '*.orig' xkb/xkbInit.c xkb/xkbInit.c
--- xkb/xkbInit.c	2009-01-22 20:33:05.826163000 -0800
+++ xkb/xkbInit.c	2009-01-22 20:33:12.637635000 -0800
@@ -94,7 +94,7 @@ typedef struct	_SrvXkmInfo {
 #endif
 
 char	*		XkbBaseDirectory=	XKB_BASE_DIRECTORY;
-char	*		XkbBinDirectory=	XKB_BIN_DIRECTORY;
+char	*		XkbBinDirectory=	XKB_BASE_DIRECTORY;
 static int	 	XkbWantAccessX=		0;	
 
 static Bool		rulesDefined=		False;
