diff -urp -x '*~' -x '*.orig' cpprules.in cpprules.in
--- cpprules.in	2007-05-29 12:36:50.000000000 -0700
+++ cpprules.in	2009-03-10 22:46:35.671590000 -0700
@@ -25,7 +25,7 @@ XORGRELSTRING = @PACKAGE_STRING@
   XORGMANNAME = X Version 11
   XSERVERNAME = Xorg
 
-MANDEFS = \
+MANDEFS = -D__SUNSOFT__ \
         -D__vendorversion__="\"$(XORGRELSTRING)\" \"$(XORGMANNAME)\"" \
         -D__xorgversion__="\"$(XORGRELSTRING)\" \"$(XORGMANNAME)\"" \
 	-D__appmansuffix__=$(APP_MAN_SUFFIX) \
diff -urp -x '*~' -x '*.orig' doc/Xserver.man.pre doc/Xserver.man.pre
--- doc/Xserver.man.pre	2009-03-10 22:45:58.609756000 -0700
+++ doc/Xserver.man.pre	2009-03-10 22:46:35.714098000 -0700
@@ -1,3 +1,32 @@
+.\" ident	"@(#)sun-manpage.patch	1.7	09/03/11 SMI"
+.\" 
+.\" Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
+.\" 
+.\" Permission is hereby granted, free of charge, to any person obtaining a
+.\" copy of this software and associated documentation files (the
+.\" "Software"), to deal in the Software without restriction, including
+.\" without limitation the rights to use, copy, modify, merge, publish,
+.\" distribute, and/or sell copies of the Software, and to permit persons
+.\" to whom the Software is furnished to do so, provided that the above
+.\" copyright notice(s) and this permission notice appear in all copies of
+.\" the Software and that both the above copyright notice(s) and this
+.\" permission notice appear in supporting documentation.
+.\"
+.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+.\" OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+.\" MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT
+.\" OF THIRD PARTY RIGHTS. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+.\" HOLDERS INCLUDED IN THIS NOTICE BE LIABLE FOR ANY CLAIM, OR ANY SPECIAL
+.\" INDIRECT OR CONSEQUENTIAL DAMAGES, OR ANY DAMAGES WHATSOEVER RESULTING
+.\" FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
+.\" NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION
+.\" WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+.\"
+.\" Except as contained in this notice, the name of a copyright holder
+.\" shall not be used in advertising or otherwise to promote the sale, use
+.\" or other dealings in this Software without prior written authorization
+.\" of the copyright holder.
+.\" 
 .\" $Xorg: Xserver.man,v 1.4 2001/02/09 02:04:07 xorgcvs Exp $
 .\" $XdotOrg: xserver/xorg/doc/Xserver.man.pre,v 1.4 2005/12/23 20:11:12 alanc Exp $
 .\" Copyright 1984 - 1991, 1993, 1994, 1998  The Open Group
@@ -32,14 +61,32 @@ Xserver \- X Window System display serve
 .SH SYNOPSIS
 .B X
 [option ...]
+#ifdef __SUNSOFT__ 
+\fB/usr/X11/bin/Xserver\fP [\fB-c\fP \fIclass\fP] \fB:\fP\fIdisplaynumber\fP [\fIX server arguments\fP]
+#endif
+.PP
 .SH DESCRIPTION
 .I X
 is the generic name for the X Window System display server.  It is
 frequently a link or a copy of the appropriate server binary for
 driving the most frequently used server on a given machine.
+#ifdef __SUNSOFT__
+In this release of Solaris, /usr/X11/bin/X is a link to the Xserver program
+which starts the appropriate X server based on properties set for the service
+management facility, under the service identifier
+\fIsvc:/application/x11/x11-server\fP.  The properties are described in the
+SMF PROPERTIES section below.
+#endif
 .SH "STARTING THE SERVER"
+#ifdef __SUNSOFT__
+On OpenSolaris, the X server is usually started from the GNOME Display 
+Manager program \fIgdm\fP(1).  On Solaris, the CDE Display Manager program 
+\fIdtlogin\fP(1) is used by default instead.  This role may also be performed
+by the X Display Manager program \fIxdm\fP(1), or another display manager. 
+#else
 The X server is usually started from the X Display Manager program 
 \fIxdm\fP(1) or a similar display manager program.
+#endif
 This utility is run from the system boot files and takes care of keeping
 the server running, prompting for usernames and passwords, and starting up
 the user sessions.
@@ -100,6 +147,14 @@ specifies a file which contains a collec
 to authenticate access.  See also the \fIxdm\fP(1) and 
 \fIXsecurity\fP(__miscmansuffix__) manual pages.
 .TP 8
+.B bc
+disables certain kinds of error checking, for bug compatibility with
+previous releases (e.g., to work around bugs in R2 and R3 xterms and toolkits).
+Deprecated.
+#ifdef __SUNSOFT__
+[Only available in Xsun, not in Xorg.]
+#endif
+.TP 8
 .B \-bs
 disables backing store support on all screens.
 .TP 8
@@ -117,6 +172,13 @@ sets key-click volume (allowable range: 
 sets the visual class for the root window of color screens.
 The class numbers are as specified in the X protocol.
 Not obeyed by all servers.
+#ifdef __SUNSOFT__
+.TP 8
+.B \-co \fIfilename\fP
+sets name of RGB color database.  The default is
+.IR __projectroot__/lib/X11/rgb .
+[Only available in Xsun, not in Xorg.]
+#endif
 .ig
 .TP 8
 .B \-config \fIfilename\fP
@@ -153,10 +215,16 @@ is platform and configuration specific.
 .BI \-extension extensionName
 disables named extension.   If an unknown extension name is specified,
 a list of accepted extension names is printed.
+#ifdef __SUNSOFT__
+[Only available in Xorg, not in Xsun.]
+#endif
 .TP 8
 .BI \+extension extensionName
 enables named extension.   If an unknown extension name is specified,
 a list of accepted extension names is printed.
+#ifdef __SUNSOFT__
+[Only available in Xorg, not in Xsun.]
+#endif
 .TP 8
 .B \-f \fIvolume\fP
 sets feep (bell) volume (allowable range: 0-100).
@@ -183,6 +251,38 @@ causes all remaining command line argume
 sets the maximum big request to
 .I size
 MB.
+#ifdef __SUNSOFT__
+[Only available in Xorg, not in Xsun.]
+#endif
+#ifdef __SUNSOFT__
+.TP 8
+.B \-nolisten \fItransport\fP
+disables listening for client connections on the named transport. The 
+current transport names supported are as follows: 
+.sp
+.nf 
+tcp     TCP over IPv4 or IPv6
+unix    UNIX Domain Sockets (/tmp/.x11-unix)
+pipe    Named pipes (/tmp/.x11-pipe/) and shared memory
+.fi
+.PP
+.in +8
+By default each of the above servers opens listeners on each of the 
+above transports. The -nolisten command line argument disables listening 
+on the named transports. You can disable multiple transports by listing 
+this option multiple times, for example: 
+.sp
+.nf 
+-nolisten pipe -nolisten tcp
+.fi
+.PP
+.in -8
+.in +13
+If all available transports are disabled, then the commands issue a fatal 
+error message and exit. A server not listening for any clients is of little 
+value.
+.in -13
+#else 
 .TP 8
 .B \-nolisten \fItrans-type\fP
 disables a transport type.  For example, TCP/IP connections can be disabled
@@ -190,6 +290,7 @@ with
 .BR "\-nolisten tcp" .
 This option may be issued multiple times to disable listening to different
 transport types.
+#endif
 .TP 8
 .B \-noreset
 prevents a server reset when the last client connection is closed.  This
@@ -203,11 +304,19 @@ sets screen-saver pattern cycle time in 
 .B \-pn
 permits the server to continue running if it fails to establish all of
 its well-known sockets (connection points for clients), but
-establishes at least one.  This option is set by default.
+establishes at least one.
+#ifdef __SUNSOFT__
+This option is set by default in Xorg, but is off by default in Xsun.
+#else
+This option is set by default.
+#endif
 .TP 8
 .B \-nopn
 causes the server to exit if it fails to establish all of its well-known
 sockets (connection points for clients).
+#ifdef __SUNSOFT__
+[Only available in Xorg, not in Xsun.]
+#endif
 .TP 8
 .B \-r
 turns off auto-repeat.
@@ -268,6 +377,10 @@ This is a no-op for most implementations
 .B [+-]xinerama
 enables(+) or disables(-) the XINERAMA extension.  The default state is
 platform and configuration specific.
+#ifdef __SUNSOFT__
+See the \fBXINERAMA\fP section of the \fIXsun\fP(1) manual page for more 
+details.
+#endif
 .SH SERVER DEPENDENT OPTIONS
 Some X servers accept the following options:
 .TP 8
@@ -311,15 +424,24 @@ use a gray map of 13 color cells for the
 .TP 8
 .I color
 use a color cube of at most 4*4*4 colors (that is 64 color cells).
+#ifdef __SUNSOFT__
+[This option is only available in Xorg, not in Xsun.]
+#endif
 .RE
 .TP 8
 .B \-dumbSched
 disables smart scheduling on platforms that support the smart scheduler.
+#ifdef __SUNSOFT__
+[Only available in Xorg, not in Xsun.]
+#endif
 .TP
 .B \-schedInterval \fIinterval\fP
 sets the smart scheduler's scheduling interval to
 .I interval
 milliseconds.
+#ifdef __SUNSOFT__
+[Only available in Xorg, not in Xsun.]
+#endif
 .SH XDMCP OPTIONS
 X servers that support XDMCP have the following options.
 See the \fIX Display Manager Control Protocol\fP specification for more
@@ -357,6 +479,9 @@ specifies the local address to connect f
 has multiple network interfaces).  The \fIlocal-address\fP may be expressed
 in any form acceptable to the host platform's \fIgethostbyname\fP(3)
 implementation.
+#ifdef __SUNSOFT__
+[Only available in Xorg, not in Xsun.]
+#endif
 .TP 8
 .B \-once
 causes the server to terminate (rather than reset) when the XDMCP session
@@ -384,22 +509,45 @@ relative path from the XKB base director
 .TP 8
 .B [+-]kb
 enables(+) or disables(-) the XKEYBOARD extension.
+#ifdef __SUNSOFT__
+The default XKEYBOARD state in Xsun in this release of Solaris is disabled 
+unless +kb is specified.  This may change in a future release.   XKEYBOARD
+is enabled by default in the Xorg server.
+#endif
 .TP 8
-.BR [+-]accessx " [ \fItimeout\fP [ \fItimeout_mask\fP [ \fIfeedback\fP [ \fIoptions_mask\fP ] ] ] ]"
+.BR [+-]accessx " [\fItimeout\fP[\fItimeout_mask\fP[\fIfeedback\fP[\fIoptions_mask\fP]]]]"
 enables(+) or disables(-) AccessX key sequences.
+#ifndef __SUNSOFT__
 .TP 8
 .B \-xkbdir \fIdirectory\fP
 base directory for keyboard layout files.  This option is not available
 for setuid X servers (i.e., when the X server's real and effective uids
 are different).
+#endif
 .TP 8
 .B \-ardelay \fImilliseconds\fP
 sets the autorepeat delay (length of time in milliseconds that a key must
 be depressed before autorepeat starts).
+#ifdef __SUNSOFT__
+[Only available in Xorg, not in Xsun.]
+#endif
 .TP 8
 .B \-arinterval \fImilliseconds\fP
 sets the autorepeat interval (length of time in milliseconds that should
 elapse between autorepeat-generated keystrokes).
+#ifdef __SUNSOFT__
+[Only available in Xorg, not in Xsun.]
+#endif
+#ifdef __SUNSOFT__
+.TP 8
+.B \-ar1 \fImilliseconds\fP
+sets the autorepeat delay (length of time in milliseconds that a key must
+be depressed before autorepeat starts).
+.TP 8
+.B \-ar2 \fImilliseconds\fP
+sets the autorepeat interval (length of time in milliseconds that should
+elapse between autorepeat-generated keystrokes).
+#endif
 .TP 8
 .B \-xkbmap \fIfilename\fP
 loads keyboard description in \fIfilename\fP on server startup.
@@ -541,6 +689,96 @@ the following font path:
     /usr/share/fonts/default/Type1,
     /usr/share/fonts/default/ghostscript
 .fi
+#ifdef __SUNSOFT__
+[Font path catalogues are only available in Xorg, not in Xsun.]
+ 
+.SH SMF PROPERTIES
+.PP
+Configuration properties for the \fB/usr/X11/bin/Xserver\fP program are
+managed by the service management facility, \fBsmf\fR(5), under the service 
+identifier:
+.PP
+.nf
+svc:/application/x11/x11-server
+.fi
+Use \fBsvccfg\fR(1M) to make configuration changes and to view configuration 
+information for this service\&. 
+.PP
+\fB/usr/X11/bin/Xserver\fP will look for configuration options in the 
+instances in this order, stopping at the first match for the given
+property:
+.TP
+1.
+\fB:display\fP\fIN\fP (where \fIN\fP is the display number of the X server, 
+i.e. \fB:display0\fP for :0)
+.TP
+2.
+Any class names provided via the \fB-c\fP options to
+\fB/usr/X11/bin/Xserver\fP.  Multiple \fB-c\fP options may be provided, if so,
+they are searched in the order found on the command line.
+.TP
+3.
+If any XDMCP options are provided to open a remote display connection,
+\fB:remote\fP, otherwise \fB:local\fP
+.TP
+4.
+The default properties for the service
+.PP
+Class names may be any valid FMRI instance name.   Sun reserves all class names
+beginning with lowercase letters for its own use - customers wishing to create
+their own class names may use names beginning with an uppercase letter.
+
+.PP
+These properties may be set in any instance of x11-server or for the service
+itself:
+.TP
+.B options/server 
+(Type astring) The absolute path to the X server binary to run, currently
+must be one of:
+.sp
+.nf
+\fB/usr/openwin/bin/Xsun\fP
+\fB/usr/openwin/bin/Xvfb\fP
+\fB/usr/X11/bin/Xorg\fP
+\fB/usr/X11/bin/i386/Xorg\fP
+\fB/usr/X11/bin/amd64/Xorg\fP
+\fB/usr/X11/bin/Xvfb\fP
+\fB/usr/X11/bin/i386/Xvfb\fP
+\fB/usr/X11/bin/amd64/Xvfb\fP
+\fB/usr/X11/bin/Xvnc\fP
+.fi
+.TP
+.B options/tcp_listen
+(Type boolean) Setting this to false disables listening for incoming TCP/IP 
+connections (see \fB-nolisten tcp\fP option above).
+.TP
+.B options/default_depth
+(Type int) Sets the depth of the default visual of the X server started.
+.TP
+.B options/server_args
+(Type astring) Sets additional arguments to pass to the X server started.  Care
+must be used when using options specific to a particular X server as then it
+may cause failure to start the X server if the server to be started is later
+changed to one that does not support those options.
+.SS "Example: Allow remote access"
+.PP
+The following command allows access from remote clients.
+.sp
+.nf
+svccfg -s svc:/application/x11/x11-server setprop options/tcp_listen = true
+.fi
+.sp
+Restart the Xserver by either logging out and logging back in, or manually 
+if the Xserver was started from the command line.
+.sp
+.SS "Example: Enables the XKEYBOARD extension"
+.PP
+The following command enables the XKEYBOARD extension.
+.sp
+.nf
+svccfg -s svc:/application/x11/x11-server setprop options/server_args=+kb
+.fi
+#endif
 
 .SH FILES
 .TP 30
@@ -549,12 +787,22 @@ Initial access control list for display 
 .TP 30
 .IR __projectroot__/lib/X11/fonts/misc , __projectroot__/lib/X11/fonts/75dpi , __projectroot__/lib/X11/fonts/100dpi
 Bitmap font directories
+#ifdef __SUNSOFT__
+.IR __projectroot__/lib/X11/fonts/TrueType , __projectroot__/lib/X11/fonts/Type1
+Outline font directories
+#else
 .TP 30
 .IR __projectroot__/lib/X11/fonts/TTF , __projectroot__/lib/X11/fonts/Type1
 Outline font directories
+#endif
 .TP 30
 .I /tmp/.X11-unix/X\fBn\fP
 Unix domain socket for display number \fBn\fP
+#ifdef __SUNSOFT__
+.TP 30
+.I /tmp/.X11-pipe/X\fBn\fP
+Named pipe (FIFO) for display number \fBn\fP
+#endif
 .TP 30
 .IR /tmp/rcX\fBn\fP
 Kerberos 5 replay cache for display number \fBn\fP
@@ -564,6 +812,15 @@ Error log file for display number \fBn\f
 .TP 30
 .I __projectroot__/lib/X11/xdm/xdm-errors
 Default error log file if the server is run from \fIxdm\fP(1)
+#ifdef __SUNSOFT__
+.TP 30
+.I /var/log/gdm/:\fBn\fP.log
+Default error log file for display number \fBn\fP if the server is run 
+from \fIgdm\fP(1)
+.TP 30
+.I /var/dt/Xerrors
+Default error log file if the server is run from \fIdtlogin\fP(1)
+#endif
 .SH "SEE ALSO"
 General information: \fIX\fP(__miscmansuffix__)
 .PP
@@ -580,14 +837,28 @@ Security: \fIXsecurity\fP(__miscmansuffi
 \fIxdm\fP(1), \fIxhost\fP(1), \fIxfwp\fP(1),
 .I "Security Extension Specification"
 .PP
+#ifdef __SUNSOFT__
+Starting the server: \fIgdm\fP(1), \fIdtlogin\fP(1), \fIxdm\fP(1), \fIxinit\fP(1)
+#else
 Starting the server: \fIxdm\fP(1), \fIxinit\fP(1)
+#endif
 .PP
 Controlling the server once started: \fIxset\fP(1), \fIxsetroot\fP(1),
 \fIxhost\fP(1)
+#ifdef __SUNSOFT__ 
+.\" Only list the servers shipped on Solaris
+.PP
+Server-specific man pages:
+\fIXsun\fP(1), \fIXorg\fP(1), \fIXephyr\fP(1), \fIXnest\fP(1), \fIXvfb\fP(1), \fIXprt\fP(1)
+.PP
+Service management framework:
+\fIsmf\fP(5), \fIsvcadm\fP(1M), \fIsvccfg\fP(1M)
+#else
 .PP
 Server-specific man pages:
 \fIXorg\fP(1), \fIXdmx\fP(1), \fIXnest\fP(1),
 \fIXvfb\fP(1), \fIXDarwin\fP(1), \fIXWin\fP(1).
+#endif
 .PP
 Server internal documentation:
 .I "Definition of the Porting Layer for the X v11 Sample Server"
