--- hw/xfree86/common/xf86Config.c	2009-03-31 11:15:39.134549000 -0700
+++ hw/xfree86/common/xf86Config.c	2009-03-31 11:14:16.834280000 -0700
@@ -908,7 +908,7 @@ configServerFlags(XF86ConfFlagsPtr flags
     MessageType from;
     const char *s;
 #ifdef XKB
-    char *rules = "base";
+    char *rules = XKB_DFLT_RULES;
 #endif
 
     /*
diff -urp -x '*~' -x '*.orig' hw/xfree86/common/xf86Configure.c hw/xfree86/common/xf86Configure.c
--- hw/xfree86/common/xf86Configure.c	2008-09-23 11:24:58.000000000 -0700
+++ hw/xfree86/common/xf86Configure.c	2008-09-24 20:18:22.067988000 -0700
@@ -555,7 +555,8 @@ configureModuleSection (void)
             /* 'strstr(dFP,"/dir")' is meant as 'dFP =~ m(/dir\W)' */
     	    if (defaultFontPath && (
 		(strcmp(*el, "freetype")  == 0 &&
-		 strstr(defaultFontPath, "/TTF")) ||
+ 		 (strstr(defaultFontPath, "/TrueType") ||
+ 		  strstr(defaultFontPath, "/TTF"))) ||
     	        (strcmp(*el, "type1")  == 0 &&
 		 strstr(defaultFontPath, "/Type1")))) 
 	    	ptr->mod_load_lst = (XF86LoadPtr)xf86addListItem(
diff -urp -x '*~' -x '*.orig' xkb/xkbInit.c xkb/xkbInit.c
--- xkb/xkbInit.c	2008-09-23 11:25:01.000000000 -0700
+++ xkb/xkbInit.c	2008-09-24 20:18:22.068856000 -0700
@@ -123,7 +123,7 @@ typedef struct	_SrvXkmInfo {
 #endif
 
 char	*		XkbBaseDirectory=	XKB_BASE_DIRECTORY;
-char	*		XkbBinDirectory=	XKB_BIN_DIRECTORY;
+char	*		XkbBinDirectory=	XKB_BASE_DIRECTORY;
 static int	 	XkbWantAccessX=		0;	
 
 static Bool		rulesDefined=		False;
