--- a/glx/glapi_gentable.c.orig  Tue Mar 27 09:06:48 2012
+++ b/glx/glapi_gentable.c       Tue Mar 27 09:07:56 2012
@@ -57,6 +57,18 @@
 #include "os.h"
 #endif
 
+#if defined(__sparc__) || defined(__sparc)
+#define WORKAROUND_7154410
+#endif
+
+#ifdef WORKAROUND_7154410
+static void *
+workaround_dlsym(void *handle, const char *name) {
+    return dlsym(handle, name);
+}
+#define dlsym workaround_dlsym
+#endif
+
 static void
 __glapi_gentable_NoOp(void) {
     const char *fstr = "Unknown";
