From 9ed9d98dec58d1f7468d02c4b951811b8e1ca7c1 Mon Sep 17 00:00:00 2001
From: Julien Cristau <jcristau@debian.org>
Date: Tue, 24 Jun 2008 22:36:32 +0200
Subject: [PATCH] x86emu: fix build, fallout from 9719354
---
 hw/xfree86/x86emu/x86emu/types.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/hw/xfree86/x86emu/x86emu/types.h b/hw/xfree86/x86emu/x86emu/types.h
index ff28885..c0c09c1 100644
--- a/hw/xfree86/x86emu/x86emu/types.h
+++ b/hw/xfree86/x86emu/x86emu/types.h
@@ -75,7 +75,7 @@ #if defined (_LP64) || \
     defined(__sparc64__) || \
     defined(__s390x__) || \
     (defined(__hppa__) && defined(__LP64)) || \
-    defined(__amd64__) || defined(amd64) \
+    defined(__amd64__) || defined(amd64) || \
     (defined(__sgi) && (_MIPS_SZLONG == 64))
 #define NUM32 int
 #else
-- 
1.4.1

