--- hw/xfree86/common/xf86pciBus.c	2009-10-26 14:55:28.776379000 +0800
+++ hw/xfree86/common/xf86pciBus.c	2009-10-26 14:54:32.631838000 +0800
@@ -371,6 +371,12 @@ xf86PciProbe(void)
 	    xf86PciVideoInfo[num - 1] = info;
 
 	    pci_device_probe(info);
+
+	    if (pci_device_is_boot_vga(info)) {
+		primaryBus.type = BUS_PCI;
+	        primaryBus.id.pci = info;
+	    }	
+
 	    info->user_data = 0;
 	}
     }
