diff -urp -x '*~' -x '*.orig' hw/xfree86/Makefile.am hw/xfree86/Makefile.am
--- hw/xfree86/Makefile.am	2007-07-30 18:35:09.238323000 -0700
+++ hw/xfree86/Makefile.am	2007-07-30 18:35:26.620143000 -0700
@@ -60,7 +60,7 @@ Xorg_LDADD = $(XORG_LIBS) \
              @XSERVER_LIBS@ \
 	     dixmods/libxorgxkb.la
 
-Xorg_LDFLAGS = $(LD_EXPORT_SYMBOLS_FLAG)
+Xorg_LDFLAGS = $(LD_EXPORT_SYMBOLS_FLAG) -Wl,-M,mapfile.scope
 
 BUILT_SOURCES = xorg.conf.example
 EXTRA_DIST = xorgconf.cpp
