diff -urp -x '*~' -x '*.orig' hw/xfree86/Makefile.am hw/xfree86/Makefile.am
--- hw/xfree86/Makefile.am	2009-01-09 09:47:03.000000000 -0800
+++ hw/xfree86/Makefile.am	2009-01-22 20:34:34.486569000 -0800
@@ -63,7 +63,7 @@ XORG_LIBS = \
 Xorg_DEPENDENCIES = $(XORG_LIBS)
 Xorg_LDADD = $(XORG_LIBS) $(XORG_SYS_LIBS) $(XSERVER_SYS_LIBS)
 
-Xorg_LDFLAGS = $(LD_EXPORT_SYMBOLS_FLAG)
+Xorg_LDFLAGS = $(LD_EXPORT_SYMBOLS_FLAG) -Wl,-M,mapfile.scope
 
 BUILT_SOURCES = xorg.conf.example
 DISTCLEANFILES += xorg.conf.example xorg.conf.example.pre
