diff -urp -x '*~' -x '*.orig' hw/xfree86/Makefile.am hw/xfree86/Makefile.am
--- hw/xfree86/Makefile.am	2008-09-23 11:24:58.000000000 -0700
+++ hw/xfree86/Makefile.am	2008-09-23 16:32:54.777197000 -0700
@@ -72,7 +72,7 @@ XORG_LIBS = \
 Xorg_DEPENDENCIES = $(XORG_LIBS)
 Xorg_LDADD = $(XORG_LIBS) $(XORG_SYS_LIBS) $(XSERVER_SYS_LIBS)
 
-Xorg_LDFLAGS = $(LD_EXPORT_SYMBOLS_FLAG)
+Xorg_LDFLAGS = $(LD_EXPORT_SYMBOLS_FLAG) -Wl,-M,mapfile.scope
 
 BUILT_SOURCES = xorg.conf.example
 DISTCLEANFILES += xorg.conf.example xorg.conf.example.pre
