diff -urp -x '*~' -x '*.orig' dix/Makefile.am dix/Makefile.am
--- dix/Makefile.am	2008-11-25 22:19:08.000000000 -0800
+++ dix/Makefile.am	2009-01-22 20:36:23.765565000 -0800
@@ -54,12 +54,12 @@ Xserver-dtrace.h: $(srcdir)/Xserver.d
 
 # Generate dtrace object code for probes in libdix
 dtrace-dix.o: $(top_srcdir)/dix/Xserver.d $(am_libdix_la_OBJECTS)
-	$(DTRACE) -G -C -o $@ -s $(top_srcdir)/dix/Xserver.d .libs/*.o
+	$(DTRACE) -G -C -o $@ -s $(top_srcdir)/dix/Xserver.d $(am_libdix_la_OBJECTS:%.lo=.libs/%.o)
 
 noinst_PROGRAMS = dix.O
 
 dix.O: dtrace-dix.o $(am_libdix_la_OBJECTS)
-	ld -r -o $@ .libs/*.o
+	ld -r -o $@ $(am_libdix_la_OBJECTS:%.lo=.libs/%.o)
 endif
 
 dix.c:
