diff -urp -x '*~' -x '*.orig' dix/Makefile.am dix/Makefile.am
--- dix/Makefile.am	2008-09-23 11:24:56.000000000 -0700
+++ dix/Makefile.am	2008-09-24 20:26:33.498245000 -0700
@@ -68,12 +68,12 @@ Xserver-dtrace.h: $(srcdir)/Xserver.d
 
 # Generate dtrace object code for probes in libdix
 dtrace-dix.o: $(top_srcdir)/dix/Xserver.d $(am_libdix_la_OBJECTS)
-	$(DTRACE) -G -C -o $@ -s $(top_srcdir)/dix/Xserver.d .libs/*.o
+	$(DTRACE) -G -C -o $@ -s $(top_srcdir)/dix/Xserver.d $(am_libdix_la_OBJECTS:%.lo=.libs/%.o)
 
 noinst_PROGRAMS = dix.O
 
 dix.O: dtrace-dix.o $(am_libdix_la_OBJECTS)
-	ld -r -o $@ .libs/*.o
+	ld -r -o $@ $(am_libdix_la_OBJECTS:%.lo=.libs/%.o)
 endif
 
 dix.c:
