--- xftfreetype.c.orig	Fri Mar 17 15:52:00 2006
+++ xftfreetype.c	Mon Jun 12 15:59:48 2006
@@ -607,7 +607,20 @@
     {
 	fi->load_flags |= FT_LOAD_NO_HINTING;
     }
-
+#ifdef FC_HINT_STYLE
+#ifdef FT_CONFIG_ADD_FULL_HINTING
+    /*
+     * So long as we are not in NO_HINTING, then check to see if the user has
+     * selected Full hinting mode. If they have, then set the new FT flag to
+     * use full hinting.
+     */
+    else
+    {
+	if ( hint_style == FC_HINT_FULL )
+	    fi->load_flags |= FT_LOAD_USE_FULL_HINTING;
+    }
+#endif /* FT_CONFIG_ADD_FULL_HINTING */
+#endif /* FC_HINT_STYLE */
     /* Figure out the load target, which modifies the hinting
      * behavior of FreeType based on the intended use of the glyphs.
      */

--- Makefile.in.orig	Mon Oct  2 16:09:55 2006
+++ Makefile.in	Mon Oct  2 16:08:46 2006
@@ -221,7 +221,7 @@
 SCRIPTS = $(bin_SCRIPTS)
 
 
-DEFAULT_INCLUDES =  -I. -I$(srcdir) -I.
+DEFAULT_INCLUDES =  -I. -I$(srcdir) -I. -DFT_CONFIG_ADD_FULL_HINTING -I../../../freetype/build_32/freetype-2.1.10/include
 depcomp = $(SHELL) $(top_srcdir)/depcomp
 am__depfiles_maybe = depfiles
 @AMDEP_TRUE@DEP_FILES = ./$(DEPDIR)/xftcolor.Plo ./$(DEPDIR)/xftcore.Plo \
