--- src/mesa/main/texstate.c	Thu Jun 18 12:47:54 2009
+++ src/mesa/main/texstate.c	Thu Jun 18 12:48:54 2009
@@ -429,7 +429,7 @@
       }
       if (texObj->_Complete) {
          texUnit->_ReallyEnabled = textureBit;
-         texUnit->_Current = texObj;
+	 _mesa_reference_texobj(&texUnit->_Current, texObj);
          update_texture_compare_function(ctx, texObj);
       }
    }
@@ -485,7 +485,6 @@
       GLbitfield enableBits;
       GLuint tex;
 
-      texUnit->_Current = NULL;
       texUnit->_ReallyEnabled = 0;
       texUnit->_GenFlags = 0;
 
