--- src/mesa/drivers/dri/i965/intel_context.c	Tue Feb 19 07:30:18 2008
+++ src/mesa/drivers/dri/i965/intel_context.c	Tue Sep 16 10:52:22 2008
@@ -119,8 +119,15 @@
 	 chipset = "Intel(R) 965GM"; break;
          break;
       case PCI_CHIP_IGD_GM:
+      case PCI_CHIP_IGD_E_G:
 	 chipset = "Intel(R) Integrated Graphics Device";
          break;
+      case PCI_CHIP_Q45_G:
+	 chipset = "Intel(R) Q45/Q43";
+	 break;
+      case PCI_CHIP_G45_G:
+	 chipset = "Intel(R) G45/G43";
+	 break;
       default:
 	 chipset = "Unknown Intel Chipset"; break;
       }
--- src/mesa/drivers/dri/i965/intel_context.h	Tue Feb 19 07:30:18 2008
+++ src/mesa/drivers/dri/i965/intel_context.h	Tue Sep 16 10:53:36 2008
@@ -387,7 +387,10 @@
 #define PCI_CHIP_I946_GZ		0x2972
 #define PCI_CHIP_I965_GM                0x2A02
 
-#define PCI_CHIP_IGD_GM       0x2A42
+#define PCI_CHIP_IGD_GM       		0x2A42
+#define PCI_CHIP_IGD_E_G                0x2E02
+#define PCI_CHIP_Q45_G                  0x2E12
+#define PCI_CHIP_G45_G                  0x2E22
 
 
 /* ================================================================
--- src/mesa/drivers/dri/i965/brw_defines.h	Tue Feb 19 07:30:18 2008
+++ src/mesa/drivers/dri/i965/brw_defines.h	Tue Sep 16 10:57:17 2008
@@ -848,7 +848,9 @@
 #define R02_PRIM_END    0x1
 #define R02_PRIM_START  0x2
 
-#define BRW_IS_IGD(brw)     ((brw)->intel.intelScreen->deviceID == PCI_CHIP_IGD_GM)
+#define BRW_IS_IGD_GM(brw)  ((brw)->intel.intelScreen->deviceID == PCI_CHIP_IGD_GM)
+#define BRW_IS_IGD_E_G(brw) (((brw)->intel.intelScreen->deviceID == PCI_CHIP_IGD_E_G) || ((brw)->intel.intelScreen->deviceID == PCI_CHIP_G45_G) || ((brw)->intel.intelScreen->deviceID == PCI_CHIP_Q45_G))
+#define BRW_IS_IGD(brw)     (BRW_IS_IGD_GM(brw) || BRW_IS_IGD_E_G(brw))
 #define CMD_PIPELINE_SELECT(brw)       ((BRW_IS_IGD(brw)) ? CMD_PIPELINE_SELECT_IGD : CMD_PIPELINE_SELECT_965)
 #define CMD_VF_STATISTICS(brw)         ((BRW_IS_IGD(brw)) ? CMD_VF_STATISTICS_IGD : CMD_VF_STATISTICS_965)
 #define URB_SIZES(brw)                 ((BRW_IS_IGD(brw)) ? 384 : 256)  /* 512 bit unit */
