From c2b838f049aa5f46c43b6a49b66a20c11d6128ab Mon Sep 17 00:00:00 2001
From: Alan Coopersmith <alan.coopersmith@sun.com>
Date: Wed, 16 Jul 2008 12:47:27 -0700
Subject: [PATCH] Add Solaris to OS'es using PROT_EXEC mmap() to get executable heap space
---
 src/mesa/main/execmem.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/src/mesa/main/execmem.c b/src/mesa/main/execmem.c
index f95c318..b71846c 100644
--- a/src/mesa/main/execmem.c
+++ b/src/mesa/main/execmem.c
@@ -36,7 +36,7 @@ #include "glapi/glthread.h"
 
 
 
-#if defined(__linux__) || defined(__OpenBSD__) || defined(_NetBSD__)
+#if defined(__linux__) || defined(__OpenBSD__) || defined(_NetBSD__) || defined(__sun)
 
 /*
  * Allocate a large block of memory which can hold code then dole it out
-- 
1.4.1

