--- src/mesa/drivers/dri/Makefile.template	Tue Apr 19 07:55:49 2011
+++ src/mesa/drivers/dri/Makefile.template	Tue Apr 19 07:50:29 2011
@@ -6,6 +6,8 @@
 
 MESA_MODULES = $(TOP)/src/mesa/libmesa.a
 
+LIBGLSL_MOBULES = $(TOP)/src/glsl/libglsl.a
+
 COMMON_GALLIUM_SOURCES = \
         ../common/utils.c \
         ../common/vblank.c \
@@ -64,7 +66,7 @@
 $(LIBNAME): $(OBJECTS) $(MESA_MODULES) $(EXTRA_MODULES) Makefile \
 		$(TOP)/src/mesa/drivers/dri/Makefile.template $(TOP)/src/mesa/drivers/dri/common/dri_test.o
 	$(MKLIB) -o $@.tmp -noprefix -linker '$(CXX)' -ldflags '$(LDFLAGS)' \
-		$(OBJECTS) $(MESA_MODULES) $(EXTRA_MODULES) $(DRI_LIB_DEPS)
+		$(OBJECTS) $(MESA_MODULES) $(LIBGLSL_MOBULES) $(EXTRA_MODULES) $(DRI_LIB_DEPS)
 	$(CXX) $(CFLAGS) -o $@.test $(TOP)/src/mesa/drivers/dri/common/dri_test.o $@.tmp $(DRI_LIB_DEPS)
 	@rm -f $@.test
 	mv -f $@.tmp $@
