--- src/mesa/main/texenvprogram.c	Fri Sep 14 11:26:06 2007
+++ src/mesa/main/texenvprogram.c	Fri Sep 14 11:28:04 2007
@@ -34,10 +34,10 @@
 #include "shader/program_instruction.h"
 
 /**
- * According to Glean's texCombine test, no more than 21 instructions
- * are needed.  Allow a few extra just in case.
+ * The ARB specification for texture unit range is 0 to 31, so 31 instructions
+ * should be enough.  Allow a few extra just in case.
  */
-#define MAX_INSTRUCTIONS 24
+#define MAX_INSTRUCTIONS (3*31+3)
 
 #define DISASSEM (MESA_VERBOSE & VERBOSE_DISASSEM)
 
