diff -urp -x '*~' -x '*.orig' bin/mklib bin/mklib
--- bin/mklib	2008-10-03 14:18:14.760469000 -0700
+++ bin/mklib	2008-10-03 14:18:25.239726000 -0700
@@ -372,6 +372,12 @@ case $ARCH in
 		fi
 	    fi
 
+	    # If using Sun C++ compiler, need to tell it not to add runpaths
+	    # that are specific to the build machine
+	    if [ ${LINK} = "CC" ] ; then
+		OPTS="${OPTS} -norunpath"
+	    fi
+
 	    # Check if objects are SPARC v9
 	    # file says: ELF 64-bit MSB relocatable SPARCV9 Version 1
 	    set ${OBJECTS}
@@ -387,6 +393,9 @@ case $ARCH in
 
 	    # Solaris linker requires explicitly listing the Standard C & C++
 	    # libraries in the link path when building shared objects
+	    if [ ${LINK} = "CC" ] ; then
+		DEPS="${DEPS} -lCrun"
+	    fi
 	    DEPS="${DEPS} -lc"
 
 	    if [ $EXPORTS ] ; then
