--- src/mesa/drivers/dri/i915/intel_context.h	Fri Oct 24 13:15:00 2014
+++ src/mesa/drivers/dri/i915/intel_context.h	Fri Oct 24 13:16:08 2014
@@ -285,9 +285,13 @@
    driOptionCache optionCache;
 };
 
+#if defined(__sun)
+#include <stdlib.h>
+#define __progname getexecname()
+#else
 extern char *__progname;
+#endif
 
-
 #define SUBPIXEL_X 0.125
 #define SUBPIXEL_Y 0.125
 
@@ -302,7 +306,7 @@
  * than COPY_DWORDS would:
  * XXX Put this in src/mesa/main/imports.h ???
  */
-#if defined(i386) || defined(__i386__)
+#if (defined(i386) || defined(__i386__)) && !(defined(__SOLARIS__) || defined(sun))
 static INLINE void * __memcpy(void * to, const void * from, size_t n)
 {
    int d0, d1, d2;
--- src/mesa/drivers/dri/common/dri_util.h	Fri Oct 24 13:16:38 2014
+++ src/mesa/drivers/dri/common/dri_util.h	Fri Oct 24 13:17:23 2014
@@ -59,6 +59,10 @@
 #include "xmlconfig.h"
 #include <stdbool.h>
 
+#define u_int64_t uint64_t
+#define u_int32_t uint32_t
+#define u_int8_t uint8_t
+
 /**
  * Extensions.
  */
--- src/mesa/drivers/dri/common/mmio.h	Fri Oct 24 13:18:01 2014
+++ src/mesa/drivers/dri/common/mmio.h	Fri Oct 24 13:19:02 2014
@@ -34,6 +34,7 @@
 #define MMIO_H
 
 #include "main/glheader.h"
+#include "dri_util.h"
 
 #if defined( __powerpc__ )
 
--- src/mesa/drivers/dri/i965/brw_draw_upload.c	Fri Oct 24 13:19:36 2014
+++ src/mesa/drivers/dri/i965/brw_draw_upload.c	Fri Oct 24 13:20:28 2014
@@ -216,7 +216,24 @@
    BRW_SURFACEFORMAT_R8G8B8A8_SSCALED
 };
 
+#if defined (sun)
+/* Solaris does not have ffsll in libc */
+#define        NBITS_INT       (CHAR_BIT * sizeof (int))
+static int ffsll(long long i)
+{
+       int i1;
+       int ret;
 
+       i1 = i & UINT_MAX;
+       ret = ffs(i1);
+       if (ret == 0) {
+               i1 = (i & 0xffffffff00000000) >> NBITS_INT;
+               ret = ffs(i1) + NBITS_INT;
+       }
+       return ret;
+}
+#endif
+
 /**
  * Given vertex array type/size/format/normalized info, return
  * the appopriate hardware surface type.
--- src/mesa/drivers/dri/i915/intel_tris.c	Fri Oct 24 13:21:20 2014
+++ src/mesa/drivers/dri/i915/intel_tris.c	Fri Oct 24 13:21:38 2014
@@ -343,7 +343,7 @@
  *                    Emit primitives as inline vertices               *
  ***********************************************************************/
 
-#ifdef __i386__
+#if (defined(i386) || defined(__i386__)) && !(defined(__SOLARIS__) || defined(sun))
 #define COPY_DWORDS( j, vb, vertsize, v )			\
 do {								\
    int __tmp;							\
--- src/mesa/drivers/dri/i915/intel_batchbuffer.c	Fri Oct 24 13:22:38 2014
+++ src/mesa/drivers/dri/i915/intel_batchbuffer.c	Fri Oct 24 13:22:49 2014
@@ -56,7 +56,7 @@
    intel->batch.bo = drm_intel_bo_alloc(intel->bufmgr, "batchbuffer",
 					intel->maxBatchSize, 4096);
 
-   intel->batch.reserved_space = BATCH_RESERVED;
+   intel->batch.reserved_space = 4096;
    intel->batch.used = 0;
 }
 
--- configure.ac	Fri Oct 24 13:24:33 2014
+++ configure.ac	Fri Oct 24 13:26:10 2014
@@ -149,7 +149,7 @@
 _SAVE_CPPFLAGS="$CPPFLAGS"
 
 dnl Compiler macros
-DEFINES="-DUSE_EXTERNAL_DXTN_LIB=1"
+DEFINES="-DUSE_EXTERNAL_DXTN_LIB=0"
 AC_SUBST([DEFINES])
 case "$host_os" in
 linux*|*-gnu*|gnu*)
@@ -183,7 +183,7 @@
     # Enable -fvisibility=hidden if using a gcc that supports it
     save_CFLAGS="$CFLAGS"
     AC_MSG_CHECKING([whether $CC supports -fvisibility=hidden])
-    VISIBILITY_CFLAGS="-fvisibility=hidden"
+    VISIBILITY_CFLAGS=""
     CFLAGS="$CFLAGS $VISIBILITY_CFLAGS"
     AC_LINK_IFELSE([AC_LANG_PROGRAM()], AC_MSG_RESULT([yes]),
 		   [VISIBILITY_CFLAGS=""; AC_MSG_RESULT([no])]);
@@ -204,7 +204,7 @@
     # Enable -fvisibility=hidden if using a gcc that supports it
     save_CXXFLAGS="$CXXFLAGS"
     AC_MSG_CHECKING([whether $CXX supports -fvisibility=hidden])
-    VISIBILITY_CXXFLAGS="-fvisibility=hidden"
+    VISIBILITY_CXXFLAGS=""
     CXXFLAGS="$CXXFLAGS $VISIBILITY_CXXFLAGS"
     AC_LANG_PUSH([C++])
     AC_LINK_IFELSE([AC_LANG_PROGRAM()], AC_MSG_RESULT([yes]),
--- src/glsl/ir_constant_expression.cpp	Fri Oct 24 13:54:06 2014
+++ src/glsl/ir_constant_expression.cpp	Fri Oct 24 13:54:47 2014
@@ -44,7 +44,7 @@
 {
    return _fpclass(x) == _FPCLASS_NN || _fpclass(x) == _FPCLASS_PN;
 }
-#elif defined(__SUNPRO_CC)
+#elif (defined(__GNUC__) || (defined(__SUNPRO_C) && (__SUNPRO_C >= 0x590)) || (defined(__SUNPRO_CC) && (__SUNPRO_CC >= 0x5130)))
 #include <ieeefp.h>
 static int isnormal(double x)
 {
--- src/mesa/drivers/dri/common/dri_util.h	Tue Nov  4 09:09:50 2014
+++ src/mesa/drivers/dri/common/dri_util.h	Tue Nov  4 09:10:58 2014
@@ -54,7 +54,7 @@
 #define _DRI_UTIL_H_
 
 #include <GL/gl.h>
-#include <GL/internal/dri_interface.h>
+#include "../../../../../include/GL/internal/dri_interface.h"
 #include "main/mtypes.h"
 #include "xmlconfig.h"
 #include <stdbool.h>
--- src/mesa/drivers/dri/i915/intel_mipmap_tree.c	Tue Nov  4 09:14:35 2014
+++ src/mesa/drivers/dri/i915/intel_mipmap_tree.c	Tue Nov  4 09:16:46 2014
@@ -25,8 +25,8 @@
  * 
  **************************************************************************/
 
-#include <GL/gl.h>
-#include <GL/internal/dri_interface.h>
+#include "../../../../../include/GL/gl.h"
+#include "../../../../../include/GL/internal/dri_interface.h"
 
 #include "intel_batchbuffer.h"
 #include "intel_chipset.h"
--- src/mesa/drivers/dri/i915/intel_mipmap_tree.h	Tue Nov  4 09:14:42 2014
+++ src/mesa/drivers/dri/i915/intel_mipmap_tree.h	Tue Nov  4 09:15:04 2014
@@ -32,7 +32,7 @@
 
 #include "intel_screen.h"
 #include "intel_regions.h"
-#include "GL/internal/dri_interface.h"
+#include "../../../../../include/GL/internal/dri_interface.h"
 
 #ifdef __cplusplus
 extern "C" {
--- src/mesa/drivers/dri/common/utils.h	Tue Nov  4 12:34:00 2014
+++ src/mesa/drivers/dri/common/utils.h	Tue Nov  4 12:34:19 2014
@@ -29,7 +29,7 @@
 #define DRI_DEBUG_H
 
 #include <GL/gl.h>
-#include <GL/internal/dri_interface.h>
+#include "../../../../../include/GL/internal/dri_interface.h"
 #include "main/context.h"
 
 struct dri_debug_control {
--- src/mesa/swrast/s_span.h	Tue Nov  4 13:20:40 2014
+++ src/mesa/swrast/s_span.h	Tue Nov  4 13:20:57 2014
@@ -38,7 +38,7 @@
 struct gl_context;
 struct gl_renderbuffer;
 
-
+typedef GLint GLfixed;
 /**
  * \defgroup SpanFlags
  * Special bitflags to describe span data.
--- src/mesa/main/glheader.h	Tue Nov  4 13:25:36 2014
+++ src/mesa/main/glheader.h	Tue Nov  4 13:26:24 2014
@@ -35,7 +35,7 @@
 
 #define GL_GLEXT_PROTOTYPES
 #include "GL/gl.h"
-#include "GL/glext.h"
+#include "../../../include/GL/glext.h"
 
 
 #ifdef __cplusplus
--- src/mesa/drivers/dri/i965/intel_mipmap_tree.h	Tue Nov  4 13:48:12 2014
+++ src/mesa/drivers/dri/i965/intel_mipmap_tree.h	Tue Nov  4 13:48:39 2014
@@ -51,7 +51,7 @@
 #include "main/mtypes.h"
 #include "intel_bufmgr.h"
 #include "intel_resolve_map.h"
-#include <GL/internal/dri_interface.h>
+#include "../../../../../include/GL/internal/dri_interface.h"
 
 #ifdef __cplusplus
 extern "C" {
--- src/mesa/drivers/dri/i965/intel_mipmap_tree.c	Tue Nov  4 13:48:07 2014
+++ src/mesa/drivers/dri/i965/intel_mipmap_tree.c	Tue Nov  4 13:49:49 2014
@@ -26,7 +26,7 @@
  **************************************************************************/
 
 #include <GL/gl.h>
-#include <GL/internal/dri_interface.h>
+#include "../../../../../include/GL/internal/dri_interface.h"
 
 #include "intel_batchbuffer.h"
 #include "intel_chipset.h"
--- src/mesa/drivers/dri/i965/intel_image.h	Tue Nov  4 13:48:01 2014
+++ src/mesa/drivers/dri/i965/intel_image.h	Tue Nov  4 13:50:33 2014
@@ -43,7 +43,7 @@
 
 #include "main/mtypes.h"
 #include "intel_bufmgr.h"
-#include <GL/internal/dri_interface.h>
+#include "../../../../../include/GL/internal/dri_interface.h"
 
 #ifdef __cplusplus
 extern "C" {
--- src/mesa/program/program.h	Tue Nov  4 14:09:53 2014
+++ src/mesa/program/program.h	Tue Nov  4 14:10:42 2014
@@ -43,7 +43,9 @@
 #include "main/compiler.h"
 #include "main/mtypes.h"
 
+#define GL_COMPUTE_PROGRAM_NV             0x90FB
 
+
 #ifdef __cplusplus
 extern "C" {
 #endif
--- src/glx/dri_common.h	Wed Nov  5 10:21:35 2014
+++ src/glx/dri_common.h	Wed Nov  5 10:22:06 2014
@@ -36,7 +36,7 @@
 #ifndef _DRI_COMMON_H
 #define _DRI_COMMON_H
 
-#include <GL/internal/dri_interface.h>
+#include "../../include/GL/internal/dri_interface.h"
 #include <stdbool.h>
 #include "loader.h"
 
--- src/glx/indirect_vertex_array.h	Thu Dec  6 09:40:13 2012
+++ src/glx/indirect_vertex_array.h	Wed Nov  5 10:51:08 2014
@@ -26,6 +26,8 @@
 #ifndef INDIRECT_VERTEX_ARRAY_H
 #define INDIRECT_VERTEX_ARRAY_H
 
+#include <GL/glext.h>
+
 extern const GLuint __glXTypeSize_table[16];
 
 #define __glXTypeSize(e) ((((e) & ~0x0f) != 0x1400) \
--- src/glx/glx_pbuffer.c	Wed Nov  5 12:02:49 2014
+++ src/glx/glx_pbuffer.c	Wed Nov  5 12:19:52 2014
@@ -35,6 +35,7 @@
 #include <X11/extensions/Xext.h>
 #include <assert.h>
 #include <string.h>
+#include "../../include/GL/glxext.h"
 #include "glxextensions.h"
 
 #ifdef GLX_USE_APPLEGL
@@ -44,6 +45,8 @@
 
 #include "glx_error.h"
 
+#define GLX_BACK_BUFFER_AGE_EXT           0x20F4
+
 #define WARN_ONCE_GLX_1_3(a, b) {		\
 		static int warned=1;		\
 		if(warned) {			\
--- src/glx/glxclient.h	Wed Nov  5 14:38:02 2014
+++ src/glx/glxclient.h	Wed Nov  5 15:10:10 2014
@@ -42,8 +42,8 @@
 #include <X11/Xfuncproto.h>
 #include <X11/extensions/extutil.h>
 #define GLX_GLXEXT_PROTOTYPES
-#include <GL/glx.h>
-#include <GL/glxext.h>
+#include "../../include/GL/glx.h"
+#include "../../include/GL/glxext.h"
 #include <string.h>
 #include <stdlib.h>
 #include <stdio.h>
