--- setxkbmap.c.orig	Tue Apr 28 18:56:08 2009
+++ setxkbmap.c	Fri Jul 23 14:19:10 2010
@@ -392,9 +392,14 @@
             ok = setOptString(&i, argc, argv, COMPAT_NDX, FROM_CMD_LINE);
         else if (streq(argv[i], "-config"))
             ok = setOptString(&i, argc, argv, CONFIG_NDX, FROM_CMD_LINE);
-        else if (streq(argv[i], "-device"))
-            deviceSpec = atoi(argv[++i]); /* only allow device IDs, not names */
-        else if (streq(argv[i], "-display"))
+        else if (streq(argv[i], "-device")) {
+	    if( ++i < argc ) {
+                deviceSpec = atoi(argv[i]); /* only allow device IDs, not names */
+	    } else {
+                usage(argc, argv);
+                exit(0);
+	    }
+        } else if (streq(argv[i], "-display"))
             ok = setOptString(&i, argc, argv, DISPLAY_NDX, FROM_CMD_LINE);
         else if (streq(argv[i], "-geometry"))
             ok = setOptString(&i, argc, argv, GEOMETRY_NDX, FROM_CMD_LINE);
