From f480933cfcc683aba19b679a6d532828eeb2eaa9 Mon Sep 17 00:00:00 2001
From: Matt Turner <mattst88@gmail.com>
Date: Sat, 7 Nov 2009 00:34:00 -0500
Subject: [PATCH] Use usleep instead of xf86UDelay

Signed-off-by: Matt Turner <mattst88@gmail.com>
---
 src/vmodes.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/vmodes.c b/src/vmodes.c
index 5faf257..c0bd0df 100644
--- a/src/vmodes.c
+++ b/src/vmodes.c
@@ -395,14 +395,14 @@ verite_restore(ScrnInfoPtr pScreenInfo, RenditionRegPtr reg)
     if (pRendition->board.chip == V1000_DEVICE) {
 	/* fixme */
         set_PLL(iob, reg->pclkpll);
-	xf86UDelay(10000);
+	usleep(10000);
     } else { 
 	verite_out32(iob+PCLKPLL,reg->pclkpll);
 	/* 
 	 * Need to wait 200uS for PLL to stabilize --
 	 * let's play it safe with 500 
 	 */
-	xf86UDelay(10000);
+	usleep(10000);
 	/* wait until VBLANK */
 	while ((verite_in32(iob+CRTCSTATUS)&CRTCSTATUS_VERT_MASK) !=
 	       CRTCSTATUS_VERT_ACTIVE);
-- 
1.5.6.5

