--- src/delta/src/delta/lbxdelta.c	Sat Jul 30 14:07:25 2005
+++ src/delta/lbxdelta.c	Sun Jun 15 15:37:59 2008
@@ -55,6 +55,8 @@
 extern void Xfree(pointer /*ptr*/);
 #endif
 
+#define Xfree(ptr) free((ptr))
+
 /*
  * Allocate data structures needed for doing Delta compaction
  */

--- src/lbx_zlib/lbx_zlib.c	Wed Jul 13 00:23:56 2005
+++ src/lbx_zlib/lbx_zlib.c	Sun Jun 15 15:38:13 2008
@@ -79,6 +79,8 @@
 extern int LbxWhoAmI;
 #endif
 
+#define Xfree(ptr) free((ptr))
+
 struct ZlibInfo {
     struct compress_private compress_state;
     struct compress_private decompress_state;

--- src/lbx_zlib/lbx_zlib_io.c	Mon Nov  7 22:33:26 2005
+++ src/lbx_zlib/lbx_zlib_io.c	Sun Jun 15 15:38:13 2008
@@ -38,6 +38,7 @@
 
 #include <stddef.h>
 
+#define Xfree(ptr) free((ptr))
 
 /*
  * The following is taken from the xtrans code, almost as is,
@@ -69,6 +70,14 @@
 }
 #endif
 
+void *
+Xalloc (unsigned long amount)
+{
+    if (amount == 0)
+        amount = 1;
+    return malloc (amount);
+}
+
 int
 InitZlibBuffer(b, size)
     ZlibBufferPtr b;
