--- lib/server/CConfig.cpp	2006-03-21 21:40:27.000000000 -0800
+++ lib/server/CConfig.cpp	2009-07-14 12:12:16.156390983 -0700
@@ -1739,7 +1739,10 @@
 								index != config.m_nameToCanonicalName.end();
 								++index) {
 			if (index->first != index->second) {
-				aliases.insert(std::make_pair(index->second, index->first));
+			    std::pair<const std::string, std::string> elem =
+				std::make_pair<const std::string, std::string>(index->second, index->first);
+			    std::multimap<const std::string, std::string, CStringUtil::CaselessCmp>::value_type value(elem);
+			    aliases.insert(value);
 			}
 		}
 
