--- lib/platform/CXWindowsKeyState.cpp	2006-04-02 12:16:39.000000000 -0700
+++ lib/platform/CXWindowsKeyState.cpp	2009-07-14 11:48:19.585387330 -0700
@@ -468,7 +468,10 @@
 
 			// add key
 			keyMap.addKeyEntry(item);
-			m_keyCodeFromKey.insert(std::make_pair(item.m_id, keycode));
+			std::pair<const unsigned, unsigned char> elem = 
+			    std::make_pair<const unsigned, unsigned char>(item.m_id, keycode);
+			std::multimap<const unsigned, unsigned char>::value_type value(elem);
+			m_keyCodeFromKey.insert(value);
 
 			// add other ways to synthesize the key
 			if ((j & 1) != 0) {
@@ -713,10 +716,14 @@
 						keyMap.addKeyEntry(item);
 
 						if (group == 0) {
-							m_keyCodeFromKey.insert(
-									std::make_pair(lKeyID, keycode));
-							m_keyCodeFromKey.insert(
-									std::make_pair(uKeyID, keycode));
+						    std::pair<const unsigned, unsigned char> eleml = 
+							std::make_pair<const unsigned, unsigned char>(lKeyID, keycode);
+						    std::multimap<const unsigned, unsigned char>::value_type valuel(eleml);
+						    m_keyCodeFromKey.insert(valuel);
+						    std::pair<const unsigned, unsigned char> elemu =
+							std::make_pair<const unsigned, unsigned char>(uKeyID, keycode);
+						    std::multimap<const unsigned, unsigned char>::value_type valueu(elemu);
+						    m_keyCodeFromKey.insert(valueu);
 						}
 						continue;
 					}
@@ -726,7 +733,10 @@
 				item.m_id = CXWindowsUtil::mapKeySymToKeyID(keysym);
 				keyMap.addKeyEntry(item);
 				if (group == 0) {
-					m_keyCodeFromKey.insert(std::make_pair(item.m_id, keycode));
+				    std::pair<const unsigned, unsigned char> elem =
+					std::make_pair<const unsigned, unsigned char>(item.m_id, keycode);
+				    std::multimap<const unsigned, unsigned char>::value_type value(elem);
+				    m_keyCodeFromKey.insert(value);
 				}
 			}
 		}
