--- lib/platform/CXWindowsClipboard.cpp	2006-04-01 17:47:03.000000000 -0800
+++ lib/platform/CXWindowsClipboard.cpp	2009-07-14 11:30:48.296738984 -0700
@@ -25,6 +25,8 @@
 #include "CArch.h"
 #include "stdvector.h"
 #include <cstdio>
+#include <map>
+#include <string>
 #include <X11/Xatom.h>
 
 //
@@ -750,7 +752,10 @@
 		}
 
 		// save it
-		motifFormats.insert(std::make_pair(motifFormat->m_type, data));
+		std::pair<const unsigned long, std::string> elem =
+		    std::make_pair<const unsigned long, std::string>(static_cast<const unsigned long>(motifFormat->m_type), data);
+		std::map<const unsigned long, std::string>::value_type value(elem);
+		motifFormats.insert(value);
 	}
 	//const UInt32 numMotifFormats = motifFormats.size();
 
