--- lib/platform/CXWindowsScreen.cpp	2006-04-02 12:16:39.000000000 -0700
+++ lib/platform/CXWindowsScreen.cpp	2009-07-14 11:55:27.355880691 -0700
@@ -600,7 +600,10 @@
 						XGrabKey(m_display, code, modifiers2, m_root,
 									False, GrabModeAsync, GrabModeAsync);
 						if (!err) {
-							hotKeys.push_back(std::make_pair(code, modifiers2));
+						    const std::pair<int, unsigned> elem =
+							std::make_pair<int, unsigned>(code, modifiers2);
+						    const std::vector<const std::pair<int, unsigned> >::value_type value(elem);
+						    hotKeys.push_back(value);
 							m_hotKeyToIDMap[CHotKeyItem(code, modifiers2)] = id;
 						}
 					}
