--- src/i830_dri.h	Tue Jan  9 22:05:04 2007
+++ src/i830_dri.h	Wed Apr  4 10:16:29 2007
@@ -22,16 +22,20 @@
    drm_handle_t unused2; /* backbuffer */
 
    drmSize unused3; /* depthbufferSize */
+   drmSize pad0;
    drm_handle_t unused4; /* depthbuffer */
 
    drmSize unused5; /* rotatedSize /*/
+   drmSize pad1;
    drm_handle_t unused6; /* rotatedbuffer */
 
    drm_handle_t unused7; /* textures */
    int unused8; /* textureSize */
+   drmSize pad2;
 
    drm_handle_t unused9; /* agp_buffers */
    drmSize unused10; /* agp_buf_size */
+   drmSize pad3;
 
    int deviceID;
    int width;

--- src/i830_common.h	Fri Sep 29 15:47:07 2006
+++ src/i830_common.h	Wed Apr  4 10:15:54 2007
@@ -92,6 +92,7 @@
         int pf_current_page;	/* which buffer is being displayed? */
         int perf_boxes;	        /* performance boxes to be displayed */   
 	int width, height;      /* screen size in pixels */
+	int pad0;
 
 	drm_handle_t front_handle;
 	int front_offset;

--- src/i830_randr.c	Wed Dec 13 23:42:14 2006
+++ src/i830_randr.c	Thu Apr 26 09:22:38 2007
@@ -182,6 +182,11 @@
         /* Reset, just so we ensure the drivers SwitchMode is called */
         scrp->currentMode = NULL;
     }
+#if defined(__sun)
+    else {
+        currentMode = scrp->currentMode;
+    }
+#endif
     /*
      * We know that if the driver failed to SwitchMode to the rotated
      * version, then it should revert back to it's prior mode.
