--- src/Makefile.am	2006-12-07 01:43:21.000000000 -0800
+++ src/Makefile.am	2007-01-11 18:31:31.064210000 -0800
@@ -86,3 +86,9 @@
 EXTRA_DIST = \
 	packed_yuv_sf.g4a \
 	packed_yuv_wm.g4a
+
+
+# Add libraries needed by driver module
+i810_drv_la_LDFLAGS += \
+	-lddc -ldrm -lfb -li2c -lint10 -lramdac -lshadow -lshadowfb \
+	-lvbe -lvgahw -lxaa -ldri -lglx -lXfont -lm

--- src/xvmc/Makefile.am	2006-09-28 07:48:32.000000000 -0700
+++ src/xvmc/Makefile.am	2007-01-11 16:57:43.077171000 -0800
@@ -4,5 +4,7 @@
 			 I810XvMC.h
 
 libI810XvMC_la_CFLAGS = @XORG_CFLAGS@ @DRI_CFLAGS@ -I$(top_srcdir)/src -DTRUE=1 -DFALSE=0
-libI810XvMC_la_LDFLAGS = -version-number 1:0:0
-endif
\ No newline at end of file
+libI810XvMC_la_LDFLAGS = -version-number 1:0:0 \
+	-lXvMC -lXv -lX11 -ldrm
+endif
+
