Index: java/src/Makefile
===================================================================
--- java/src/Makefile	(revision 3916)
+++ java/src/Makefile	(working copy)
@@ -4,41 +4,56 @@
 
 CP = cp
 JC = javac
-JCFLAGS = -target 1.5 -classpath ../../../
+JCFLAGS = -target 1.5 -classpath .
 JAR = jar
 ARCHIVE = VncViewer.jar
-MANIFEST = MANIFEST.MF
-PAGES = index.vnc
+MANIFEST = com/tigervnc/vncviewer/MANIFEST.MF
+PAGES = com/tigervnc/vncviewer/index.vnc com/tigervnc/vncviewer/index.html
 INSTALL_DIR = /usr/local/vnc/classes
 
-CLASSES = VncViewer.class RfbProto.class AuthPanel.class VncCanvas.class \
-	  VncCanvas2.class \
-	  OptionsFrame.class ClipboardFrame.class ButtonPanel.class \
-	  DesCipher.class CapabilityInfo.class CapsContainer.class \
-	  RecordingFrame.class SessionRecorder.class \
-	  SocketFactory.class HTTPConnectSocketFactory.class \
-	  HTTPConnectSocket.class ReloginPanel.class \
-	  InStream.class MemInStream.class ZlibInStream.class
+SOURCES = \
+	com/tigervnc/decoder/CopyRectDecoder.java		\
+	com/tigervnc/decoder/CoRREDecoder.java			\
+	com/tigervnc/decoder/HextileDecoder.java		\
+	com/tigervnc/decoder/RawDecoder.java			\
+	com/tigervnc/decoder/RREDecoder.java			\
+	com/tigervnc/decoder/TightDecoder.java			\
+	com/tigervnc/decoder/ZlibDecoder.java			\
+	com/tigervnc/decoder/ZRLEDecoder.java			\
+	com/tigervnc/decoder/common/Repaintable.java		\
+	com/tigervnc/vncviewer/AuthPanel.java			\
+	com/tigervnc/vncviewer/ButtonPanel.java			\
+	com/tigervnc/vncviewer/CapabilityInfo.java		\
+	com/tigervnc/vncviewer/CapsContainer.java		\
+	com/tigervnc/vncviewer/ClipboardFrame.java		\
+	com/tigervnc/vncviewer/DesCipher.java			\
+	com/tigervnc/vncviewer/HTTPConnectSocket.java		\
+	com/tigervnc/vncviewer/HTTPConnectSocketFactory.java	\
+	com/tigervnc/vncviewer/InStream.java			\
+	com/tigervnc/vncviewer/MemInStream.java			\
+	com/tigervnc/vncviewer/OptionsFrame.java		\
+	com/tigervnc/vncviewer/RecordingFrame.java		\
+	com/tigervnc/vncviewer/RecordOutputStream.java		\
+	com/tigervnc/vncviewer/ReloginPanel.java		\
+	com/tigervnc/vncviewer/RfbInputStream.java		\
+	com/tigervnc/vncviewer/RfbProto.java			\
+	com/tigervnc/vncviewer/SessionRecorder.java		\
+	com/tigervnc/vncviewer/SocketFactory.java		\
+	com/tigervnc/vncviewer/VncCanvas.java			\
+	com/tigervnc/vncviewer/VncCanvas2.java			\
+	com/tigervnc/vncviewer/VncViewer.java			\
+	com/tigervnc/vncviewer/ZlibInStream.java
 
-SOURCES = VncViewer.java RfbProto.java AuthPanel.java VncCanvas.java \
-	  VncCanvas2.java \
-	  OptionsFrame.java ClipboardFrame.java ButtonPanel.java \
-	  DesCipher.java CapabilityInfo.java CapsContainer.java \
-	  RecordingFrame.java SessionRecorder.java \
-	  SocketFactory.java HTTPConnectSocketFactory.java \
-	  HTTPConnectSocket.java ReloginPanel.java \
-	  InStream.java MemInStream.java ZlibInStream.java
+CLASSES = $(SOURCES:.java=.class)
 
-all: $(CLASSES) $(ARCHIVE)
+all: $(ARCHIVE)
 
-$(CLASSES): $(SOURCES)
+$(ARCHIVE): $(SOURCES) $(MANIFEST)
 	$(JC) $(JCFLAGS) -O $(SOURCES)
-
-$(ARCHIVE): $(CLASSES) $(MANIFEST)
 	$(JAR) cfm $(ARCHIVE) $(MANIFEST) $(CLASSES)
 
-install: $(CLASSES) $(ARCHIVE)
-	$(CP) $(CLASSES) $(ARCHIVE) $(PAGES) $(INSTALL_DIR)
+install: $(ARCHIVE)
+	$(CP) $(ARCHIVE) $(PAGES) $(INSTALL_DIR)
 
 export:: $(CLASSES) $(ARCHIVE) $(PAGES)
 	@$(ExportJavaClasses)
Index: java/src/com/tigervnc/vncviewer/index.vnc
===================================================================
--- java/src/com/tigervnc/vncviewer/index.vnc	(revision 3916)
+++ java/src/com/tigervnc/vncviewer/index.vnc	(working copy)
@@ -1,3 +1,4 @@
+<HTML><HEAD>
 <!-- 
      index.vnc - default HTML page for TigerVNC Java viewer applet, to be
      used with Xvnc. On any file ending in .vnc, the HTTP server embedded in
@@ -6,20 +7,20 @@
      PARAMS. Use two dollar signs ($$) to get a dollar sign in the generated
      HTML page.
 
-     NOTE: the $PARAMS variable is not supported by the standard VNC, so
+     NOTE: the PARAMS variable is not supported by the standard VNC, so
      make sure you have TigerVNC on the server side, if you're using this
      variable.
 -->
-
-<HTML>
 <TITLE>
 $USER's $DESKTOP desktop ($DISPLAY)
 </TITLE>
-<APPLET CODE=VncViewer.class ARCHIVE=VncViewer.jar
-        WIDTH=$APPLETWIDTH HEIGHT=$APPLETHEIGHT>
-<param name=PORT value=$PORT>
+</HEAD><BODY>
+<APPLET CODE="com.tigervnc.vncviewer.VncViewer.class" ARCHIVE="VncViewer.jar"
+        WIDTH="$APPLETWIDTH" HEIGHT="$APPLETHEIGHT">
+<param name="PORT" value="$PORT">
 $PARAMS
 </APPLET>
 <BR>
 <A href="http://www.tigervnc.org/">TigerVNC site</A>
+</BODY>
 </HTML>
Index: java/src/com/tigervnc/vncviewer/MANIFEST.MF
===================================================================
--- java/src/com/tigervnc/vncviewer/MANIFEST.MF	(revision 3916)
+++ java/src/com/tigervnc/vncviewer/MANIFEST.MF	(working copy)
@@ -1,2 +1,2 @@
 Manifest-Version: 1.0
-Main-Class: VncViewer
+Main-Class: com.tigervnc.vncviewer.VncViewer
Index: java/src/com/tigervnc/vncviewer/index.html
===================================================================
--- java/src/com/tigervnc/vncviewer/index.html	(revision 3916)
+++ java/src/com/tigervnc/vncviewer/index.html	(working copy)
@@ -1,3 +1,4 @@
+<HTML><HEAD>
 <!-- 
      index.html - an example HTML page for TigerVNC Java viewer applet, to be
      used with a standalone Web server running on the same machine where the
@@ -15,15 +16,15 @@
        actual desktop size on the server (height should be increased to leave
        enough space for the button panel).
 -->
-
-<HTML>
 <TITLE>
 TigerVNC desktop
 </TITLE>
-<APPLET CODE="VncViewer.class" ARCHIVE="VncViewer.jar"
+</HEAD><BODY>
+<APPLET CODE="com.tigervnc.vncviewer.VncViewer.class" ARCHIVE="VncViewer.jar"
         WIDTH="800" HEIGHT="632">
 <PARAM NAME="PORT" VALUE="5901">
 </APPLET>
 <BR>
 <A href="http://www.tigervnc.org/">TigerVNC site</A>
+</BODY>
 </HTML>
