--- unix/xorg-x11-server-source/configure.ac~	2007-09-19 14:00:02.330977000 -0700
+++ unix/xorg-x11-server-source/configure.ac	2007-09-19 14:16:41.522248000 -0700
@@ -1121,7 +1121,7 @@
 xorg_bus_sparcpci=no
 xorg_bus_sparc=no
 
-if test "x$XORG" = xyes -o "x$XGL" = xyes; then
+if test "xyes" = xyes -o "x$XGL" = xyes; then
 	XORG_DDXINCS='-I$(top_srcdir)/hw/xfree86 -I$(top_srcdir)/hw/xfree86/include -I$(top_srcdir)/hw/xfree86/common'
 	XORG_OSINCS='-I$(top_srcdir)/hw/xfree86/os-support -I$(top_srcdir)/hw/xfree86/os-support/bus -I$(top_srcdir)/os'
 	XORG_INCS="$XORG_DDXINCS $XORG_OSINCS"


--- unix/xorg-x11-server-source/Xext/Makefile.am~	2007-09-19 14:18:08.721581000 -0700
+++ unix/xorg-x11-server-source/Xext/Makefile.am	2007-09-19 14:36:05.327716000 -0700
@@ -4,11 +4,7 @@
 #                    Xorg by default
 # libXextmodule.la:  includes those extensions that are built into a module
 #                    that Xorg loads
-if XORG
 noinst_LTLIBRARIES = libXext.la libXextbuiltin.la libXextmodule.la
-else
-noinst_LTLIBRARIES = libXext.la
-endif
 
 INCLUDES = -I$(top_srcdir)/hw/xfree86/dixmods/extmod
 
@@ -163,11 +159,9 @@
 
 libXext_la_SOURCES =		$(BUILTIN_SRCS) $(MODULE_SRCS)
 
-if XORG
 libXextbuiltin_la_SOURCES =	$(BUILTIN_SRCS)
 
 libXextmodule_la_SOURCES =	$(MODULE_SRCS)
-endif
 
 EXTRA_DIST = \
 	$(SERVERCONFIG_DATA) \

--- unix/xorg-x11-server-source/hw/Makefile.am~	2007-09-19 13:17:07.009236000 -0700
+++ unix/xorg-x11-server-source/hw/Makefile.am	2007-09-19 13:57:41.224962000 -0700
@@ -4,6 +4,8 @@
 
 if XORG
 XORG_SUBDIRS = xfree86
+else
+XORG_SUBDIRS = xfree86/os-support xfree86/common
 endif
 
 if XVFB

 
