--- unix/xc/programs/Xserver/vnc/vncExtInit.cc	2006-05-15 09:56:20.000000000 -0700
+++ unix/xc/programs/Xserver/vnc/vncExtInit.cc	2007-06-26 18:11:37.123722000 -0700
@@ -20,6 +20,7 @@
 
 extern "C" {
 #define class c_class
+#define public c_public    
 #define NEED_EVENTS
 #include "X.h"
 #include "Xproto.h"
@@ -34,6 +35,7 @@
 #include "vncExt.h"
 #undef class
 #undef xalloc
+#undef public
 }
 
 #include <rfb/Configuration.h>
