--- a/java/src/com/tigervnc/vncviewer/Makefile	Thu Jul 10 14:52:31 2014
+++ b/java/src/com/tigervnc/vncviewer/Makefile	Thu Jul 10 14:52:51 2014
@@ -3,8 +3,9 @@
 #
 
 CP = cp
-JC = javac
-JCFLAGS = -target 1.5 -classpath ../../../
+JDK_VERSION = 1.8.0
+JC = /usr/jdk/instances/jdk$(JDK_VERSION)/bin/javac
+JCFLAGS = -target 1.8 -classpath ../../../
 JAR = jar
 ARCHIVE = VncViewer.jar
 MANIFEST = MANIFEST.MF
