--- tigervnc-1.0.0/common/configure.ac~	Thu Aug 20 02:46:42 2009
+++ tigervnc-1.0.0/common/configure.ac	Mon Aug 31 07:14:23 2009
@@ -58,7 +58,9 @@
 	[AC_SEARCH_LIBS([inflateEnd], [z], [], [INCLUDED_ZLIB=yes])])
 
 AM_CONDITIONAL([INCLUDED_ZLIB], [ test "x$INCLUDED_ZLIB" = xyes ])
-AC_CONFIG_SUBDIRS([zlib])
+if test "x$INCLUDED_ZLIB" = xyes ; then
+  AC_CONFIG_SUBDIRS([zlib])
+fi
 
 dnl Check for libjpeg library
 INCLUDED_JPEG=yes
@@ -70,7 +72,9 @@
 	[])
 
 AM_CONDITIONAL([INCLUDED_JPEG], [ test "x$INCLUDED_JPEG" = xyes ])
-AC_CONFIG_SUBDIRS([jpeg])
+if test "x$INCLUDED_JPEG" = xyes ; then
+  AC_CONFIG_SUBDIRS([jpeg])
+fi
 
 AC_CHECK_FUNCS([vsnprintf snprintf strcasecmp strncasecmp])
 
