From 9719534fa3a184b848ef9e9c3a755703b8708fd2 Mon Sep 17 00:00:00 2001
From: Alan Coopersmith <alan.coopersmith@sun.com>
Date: Thu, 15 Jan 2009 07:29:44 -0800
Subject: [PATCH] Make --with-hal* configure options match their help output

---
 configure.ac |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/configure.ac b/configure.ac
index 5c5f639..118ae84 100644
--- a/configure.ac
+++ b/configure.ac
@@ -70,7 +70,7 @@ AC_ARG_WITH(xorg-module-dir,
 inputdir=${moduledir}/input
 AC_SUBST(inputdir)
 
-AC_ARG_WITH(hal_bin_dir,
+AC_ARG_WITH(hal-bin-dir,
 	    AC_HELP_STRING([--with-hal-bin-dir=DIR],
 			   [Directory where HAL binaries where installed [[default=$bindir]]]),
 	    [halbindir="$withval"],
@@ -78,7 +78,7 @@ AC_ARG_WITH(hal_bin_dir,
 HAL_BIN_DIR=${halbindir}
 AC_SUBST(HAL_BIN_DIR)
 
-AC_ARG_WITH(hal_callouts_dir,
+AC_ARG_WITH(hal-callouts-dir,
 	    AC_HELP_STRING([--with-hal-callouts-dir=DIR],
 			   [Directory where HAL expects its callout scripts to be located
 			    [[default=$libdir/hal]]]),
@@ -87,7 +87,7 @@ AC_ARG_WITH(hal_callouts_dir,
 HAL_CALLOUTS_DIR=${halcalloutsdir}
 AC_SUBST(HAL_CALLOUTS_DIR)
 
-AC_ARG_WITH(hal_fdi_dir,
+AC_ARG_WITH(hal-fdi-dir,
 	    AC_HELP_STRING([--with-hal-fdi-dir=DIR],
 			   [Directory where HAL expects its fdi files to be located
 			   [[default=$datadir/hal/fdi/policy/20thirdparty]]]),
-- 
1.5.6.5

