diff --git a/src/truetype/ttinterp.h b/src/truetype/ttinterp.h
index 6d0fc03..5c21a36 100644
--- a/src/truetype/ttinterp.h
+++ b/src/truetype/ttinterp.h
@@ -139,6 +139,8 @@ FT_BEGIN_HEADER
     FT_Size_Metrics    metrics;
     TT_Size_Metrics    tt_metrics; /* size metrics */
 
+#undef GS /* avoid conflict with <sys/regset.h> on Solaris x86 systems */
+#define GS FT_GS
     TT_GraphicsState   GS;         /* current graphics state */
 
     FT_Int             curRange;  /* current code range number   */
diff --git a/src/truetype/ttobjs.h b/src/truetype/ttobjs.h
index 47e4129..adf5920 100644
--- a/src/truetype/ttobjs.h
+++ b/src/truetype/ttobjs.h
@@ -312,6 +312,8 @@ FT_BEGIN_HEADER
 
     TT_CodeRangeTable  codeRangeTable;
 
+#undef GS /* avoid conflict with <sys/regset.h> on Solaris x86 systems */
+#define GS FT_GS
     TT_GraphicsState   GS;
 
     FT_ULong           cvt_size;      /* the scaled control value table */
