diff -urp freetype-2.4.1/src/truetype/ttinterp.h freetype-2.4.1/src/truetype/ttinterp.h
--- freetype-2.4.1/src/truetype/ttinterp.h	2010-05-20 05:04:47.000000000 -0700
+++ freetype-2.4.1/src/truetype/ttinterp.h	2010-08-05 17:23:58.073273648 -0700
@@ -138,6 +138,8 @@ FT_BEGIN_HEADER
     FT_Size_Metrics    metrics;
     TT_Size_Metrics    tt_metrics; /* size metrics */
 
+#undef GS /* avoid conflict with <sys/regset.h> on Solaris x86 systems */
+#define GS FT_GS
     TT_GraphicsState   GS;         /* current graphics state */
 
     FT_Int             curRange;  /* current code range number   */
diff -urp freetype-2.4.1/src/truetype/ttobjs.h freetype-2.4.1/src/truetype/ttobjs.h
--- freetype-2.4.1/src/truetype/ttobjs.h	2009-07-03 06:28:24.000000000 -0700
+++ freetype-2.4.1/src/truetype/ttobjs.h	2010-08-05 17:24:05.630291335 -0700
@@ -311,6 +311,8 @@ FT_BEGIN_HEADER
 
     TT_CodeRangeTable  codeRangeTable;
 
+#undef GS /* avoid conflict with <sys/regset.h> on Solaris x86 systems */
+#define GS FT_GS
     TT_GraphicsState   GS;
 
     FT_ULong           cvt_size;      /* the scaled control value table */
