--- src/truetype/ttgload.c.orig	Mon Mar  6 16:19:31 2006
+++ src/truetype/ttgload.c	Mon Mar 13 16:26:53 2006
@@ -916,7 +916,14 @@
         load->exec->pts              = *zone;
         load->exec->pts.n_points    += 4;
 
-        error = TT_Run_Context( load->exec, debug );
+#ifdef FT_CONFIG_ADD_FULL_HINTING
+	if ( load->load_flags & FT_LOAD_USE_FULL_HINTING )
+#endif /* FT_CONFIG_ADD_FULL_HINTING */
+            error = TT_Run_Context( load->exec, debug );
+#ifdef FT_CONFIG_ADD_FULL_HINTING
+	else
+	    error = TT_Err_Ok;
+#endif /* FT_CONFIG_ADD_FULL_HINTING */
         if ( error && load->exec->pedantic_hinting )
           goto Exit;
 
@@ -1635,7 +1642,14 @@
           if ( IS_HINTED( loader->load_flags ) && n_ins > 0 )
           {
             exec->is_composite     = TRUE;
-            error = TT_Run_Context( exec, ((TT_Size)loader->size)->debug );
+#ifdef FT_CONFIG_ADD_FULL_HINTING
+	    if ( loader->load_flags & FT_LOAD_USE_FULL_HINTING )
+#endif /* FT_CONFIG_ADD_FULL_HINTING */
+                error = TT_Run_Context( exec, ((TT_Size)loader->size)->debug );
+#ifdef FT_CONFIG_ADD_FULL_HINTING
+	    else
+	        error = TT_Err_Ok;
+#endif /* FT_CONFIG_ADD_FULL_HINTING */
             if ( error && exec->pedantic_hinting )
               goto Fail;
           }
--- include/freetype/freetype.h.orig	Fri Mar 10 15:53:18 2006
+++ include/freetype/freetype.h	Mon Mar 13 14:49:24 2006
@@ -2339,6 +2339,9 @@
 #define FT_LOAD_FORCE_AUTOHINT               0x20
 #define FT_LOAD_CROP_BITMAP                  0x40
 #define FT_LOAD_PEDANTIC                     0x80
+#ifdef FT_CONFIG_ADD_FULL_HINTING
+#define FT_LOAD_USE_FULL_HINTING	     0x100
+#endif
 #define FT_LOAD_IGNORE_GLOBAL_ADVANCE_WIDTH  0x200
 #define FT_LOAD_NO_RECURSE                   0x400
 #define FT_LOAD_IGNORE_TRANSFORM             0x800
--- include/freetype/config/ftoption.h.orig	Mon Mar 13 15:20:30 2006
+++ include/freetype/config/ftoption.h	Mon Mar 13 14:53:20 2006
@@ -436,7 +436,8 @@
   /*   Do not #undef this macro here, since the build system might         */
   /*   define it for certain configurations only.                          */
   /*                                                                       */
-/* #define TT_CONFIG_OPTION_BYTECODE_INTERPRETER */
+#define TT_CONFIG_OPTION_BYTECODE_INTERPRETER
+#define FT_CONFIG_ADD_FULL_HINTING
 
 
   /*************************************************************************/
