From 7a4d6f396597c4b96e1402dadfcd2c0e5b7b0579 Mon Sep 17 00:00:00 2001
From: Alan Coopersmith <alan.coopersmith@sun.com>
Date: Fri, 1 May 2009 14:47:59 -0700
Subject: [PATCH] Override HKTG & BSlash2 keycodes on Solaris

Signed-off-by: Alan Coopersmith <alan.coopersmith@sun.com>
---
 src/sun_kbdMap.c |    9 ++++++++-
 1 files changed, 8 insertions(+), 1 deletions(-)

diff --git a/src/sun_kbdMap.c b/src/sun_kbdMap.c
index 0b4b99c..50d7585 100644
--- a/src/sun_kbdMap.c
+++ b/src/sun_kbdMap.c
@@ -36,10 +36,17 @@
 
 /* Map the Solaris keycodes to the "XFree86" keycodes. */
 
-/* Additional Sun Japanese Keyboard Keys not defined in common/atKeynames.h */
+/* Additional Sun Japanese Keyboard Keys not defined in atKeynames.h */
 #define KEY_Kanji	0x82
 #define KEY_Execute	0x83
 
+/* Override atKeynames.h values with unique keycodes, so we can distinguish
+   KEY_F15 from KEY_HKTG & KEY_KP_DEC from KEY_BSlash2 */
+#undef KEY_HKTG
+#define KEY_HKTG         /* Hirugana/Katakana tog 0xC8  */  200 /* was 112 */
+#undef KEY_BSlash2
+#define KEY_BSlash2      /* \           _         0xCB  */  203 /* was 115 */
+
 static unsigned char sunmap[256] = {
 #if defined(i386) || defined(__i386) || defined(__i386__) || defined(__x86)
 	KEY_NOTUSED,		/*   0 */
-- 
1.5.6.5

