diff -urp -x '*~' -x '*.orig' src/sun_kbd.c src/sun_kbd.c
--- src/sun_kbd.c	2009-08-15 18:12:35.579394000 -0700
+++ src/sun_kbd.c	2009-08-15 18:12:36.475454000 -0700
@@ -63,6 +63,7 @@
 #include <sys/stropts.h>
 #include <sys/vuid_event.h>
 #include <sys/kbd.h>
+#include <sys/usb/clients/hid/hid.h>
 
 static int KbdOn(InputInfoPtr pInfo, int what);
 static Bool OpenKeyboard(InputInfoPtr pInfo);
@@ -180,6 +181,8 @@ KbdOn(InputInfoPtr pInfo, int what)
     sunKbdPrivPtr priv = (sunKbdPrivPtr) pKbd->private;
 
     int	ktrans, kdirect, i;
+    int io_get_direct = KIOCGDIRECT;
+    int io_set_direct = KIOCSDIRECT;
 
     if (priv->kbdActive) {
 	return Success;
@@ -198,9 +201,15 @@ KbdOn(InputInfoPtr pInfo, int what)
 		    "%s: cannot push module '%s' onto keyboard device: %s\n",
 		    pInfo->name, priv->strmod, strerror(errno));
 	}
+
+	if (strcmp(priv->strmod, "usbkbm") == 0) {
+	    io_get_direct = HIDIOCKMGDIRECT;
+	    io_set_direct = HIDIOCKMSDIRECT;
+	}
+
     }
 
-    SYSCALL(i = ioctl(pInfo->fd, KIOCGDIRECT, &kdirect));
+    SYSCALL(i = ioctl(pInfo->fd, io_get_direct, &kdirect));
     if (i < 0) {
 	xf86Msg(X_ERROR, 
 		"%s: Unable to determine keyboard direct setting: %s\n", 
@@ -211,7 +220,7 @@ KbdOn(InputInfoPtr pInfo, int what)
     priv->odirect = kdirect;
     kdirect = 1;
 
-    SYSCALL(i = ioctl(pInfo->fd, KIOCSDIRECT, &kdirect));
+    SYSCALL(i = ioctl(pInfo->fd, io_set_direct, &kdirect));
     if (i < 0) {
 	xf86Msg(X_ERROR, "%s: Failed turning keyboard direct mode on: %s\n",
 			pInfo->name, strerror(errno));
@@ -255,6 +264,7 @@ KbdOff(InputInfoPtr pInfo, int what)
     sunKbdPrivPtr priv = (sunKbdPrivPtr) pKbd->private;
 
     int i;
+    int io_set_direct, kdirect;
 
     if (priv->remove_timer) {
 	TimerFree(priv->remove_timer);
@@ -288,8 +298,16 @@ KbdOff(InputInfoPtr pInfo, int what)
 	priv->otranslation = -1;
     }
 
-    if (priv->odirect != -1) {
-        SYSCALL(i = ioctl(pInfo->fd, KIOCSDIRECT, &priv->odirect));
+    io_set_direct = KIOCSDIRECT;
+    kdirect = priv->odirect;
+
+    if ((priv->strmod != NULL) && (strcmp(priv->strmod, "usbkbm") == 0)) {
+	io_set_direct = HIDIOCKMSDIRECT;
+	kdirect = 0;
+    }
+
+    if (kdirect != -1) {
+        SYSCALL(i = ioctl(pInfo->fd, io_set_direct, &kdirect));
 	if (i < 0) {
 	    xf86Msg(X_ERROR,
 		    "%s: Unable to restore keyboard direct setting: %s\n",
