diff -urp -x '*~' -x '*.orig' src/i830_driver.c src/i830_driver.c
--- src/i830_driver.c	2008-08-25 18:25:53.000000000 -0700
+++ src/i830_driver.c	2009-01-10 20:01:46.246642000 -0800
@@ -1589,6 +1589,9 @@ I830PreInit(ScrnInfoPtr pScrn, int flags
 	       pI830->useEXA = FALSE;
 	   }
        }
+       if (DEVICE_ID(pI830->PciInfo) == PCI_CHIP_845_G) {
+	       pI830->useEXA = FALSE;
+       }
 #endif
        xf86DrvMsg(pScrn->scrnIndex, from, "Using %s for acceleration\n",
 		  pI830->useEXA ? "EXA" : "XAA");
