--- configure.ac	Mon Aug 25 18:28:31 2008
+++ configure.ac	Mon Sep 15 15:32:01 2008
@@ -37,7 +37,7 @@
 # Checks for programs.
 AC_DISABLE_STATIC
 AC_PROG_LIBTOOL
-DOLT
+#DOLT
 AC_PROG_CC
 AM_PROG_CC_C_O

@@ -84,7 +84,11 @@
 
 # Checks for pkg-config packages
 PKG_CHECK_MODULES(XORG, [xorg-server xproto fontsproto $REQUIRED_MODULES])
-sdkdir=$(pkg-config --variable=sdkdir xorg-server)
+AC_ARG_WITH(xorg-sdk-dir,
+	    AC_HELP_STRING([--with-xorg-sdk-dir=DIR],
+		           [Default xorg sdk include directory [[default is taken from pkgconfig]]]),
+	    [sdkdir="$withval"],
+	    [sdkdir=$(pkg-config --variable=sdkdir xorg-server)])
 
 # Checks for libraries.
 
