--- configure.ac	Wed Feb 10 11:08:23 2010
+++ configure.ac	Wed Feb 10 11:09:48 2010
@@ -97,7 +97,11 @@
                   HAVE_XEXTPROTO_71="no")
 AM_CONDITIONAL(HAVE_XEXTPROTO_71, [ test "$HAVE_XEXTPROTO_71" = "yes" ])
 
-sdkdir=`$PKG_CONFIG --variable=sdkdir xorg-server`
+AC_ARG_WITH(xorg-sdk-dir,
+	AC_HELP_STRING([--with-xorg-sdk-dir=DIR],
+	[Default xorg sdk include directory [[default is taken from pkgconfig]]]),
+	[sdkdir="$withval"],
+	[sdkdir=$(pkg-config --variable=sdkdir xorg-server)])
 
 save_CFLAGS="$CFLAGS"
 CFLAGS="$XORG_CFLAGS $DRI_CFLAGS $DRM_CFLAGS"
