--- src/Makefile.in	Fri Sep 14 12:39:05 2012
+++ src/Makefile.in	Fri Sep 14 12:39:15 2012
@@ -454,7 +454,7 @@
 intel_drv_la_LIBADD = legacy/liblegacy.la @PCIACCESS_LIBS@ \
 	$(am__append_2) $(am__append_4) $(am__append_7) \
 	$(am__append_10)
-NULL := #
+#NULL := #
 intel_drv_la_SOURCES = intel_list.h intel_options.h intel_options.c \
 	intel_module.c compat-api.h $(NULL) $(am__append_5) \
 	$(am__append_8) $(am__append_9) $(am__append_11)
--- src/legacy/Makefile.in	Fri Sep 14 12:40:07 2012
+++ src/legacy/Makefile.in	Fri Sep 14 12:40:17 2012
@@ -316,7 +316,7 @@
 top_srcdir = @top_srcdir@
 SUBDIRS = $(am__append_1)
 noinst_LTLIBRARIES = liblegacy.la
-NULL := #
+#NULL := #
 liblegacy_la_SOURCES = legacy.h $(am__append_2)
 liblegacy_la_LIBADD = $(am__append_3)
 EXTRA_DIST = README
--- src/Makefile.am	Fri Sep 14 12:40:49 2012
+++ src/Makefile.am	Fri Sep 14 12:43:18 2012
@@ -32,6 +32,9 @@
 intel_drv_la_LDFLAGS = -module -avoid-version
 intel_drv_ladir = @moduledir@/drivers
 intel_drv_la_LIBADD = legacy/liblegacy.la @PCIACCESS_LIBS@
+intel_drv_la_LDFLAGS += -R $(moduledir)
+intel_drv_la_LIBADD += $(XORG_EXTERNS_FLAG) -lfb -lshadowfb -lm \
+       -lvbe -lvgahw -lexa $(XORG_LIBS)
 
 if SNA
 SUBDIRS += sna
@@ -38,7 +41,7 @@
 intel_drv_la_LIBADD += sna/libsna.la
 endif
 
-NULL:=#
+#NULL:=#
 
 intel_drv_la_SOURCES = \
 	intel_list.h \
--- src/legacy/Makefile.am	Fri Sep 14 12:43:52 2012
+++ src/legacy/Makefile.am	Fri Sep 14 12:45:12 2012
@@ -1,11 +1,22 @@
 SUBDIRS =
 noinst_LTLIBRARIES = liblegacy.la
 
-NULL:=#
+#NULL:=#
 
-liblegacy_la_SOURCES = legacy.h
-liblegacy_la_LIBADD =
+liblegacy_la_SOURCES = legacy.h \
+                      i810/i810_accel.c \
+                      i810/i810_cursor.c \
+                      i810/i810_dga.c \
+                      i810/i810_dri.c \
+                      i810/i810_driver.c \
+                      i810/i810_hwmc.c \
+                      i810/i810_memory.c \
+                      i810/i810_video.c \
+                      i810/i810_wmark.c
 
+AM_CFLAGS = @XORG_CFLAGS@ @DRM_CFLAGS@ @DRI_CFLAGS@ -I$(top_srcdir)/src
+liblegacy_la_LIBADD = @DRI_LIBS@ @DRM_LIBS@ @XVMCLIB_LIBS@
+
 if ! KMS_ONLY
 SUBDIRS += i810
 liblegacy_la_SOURCES +=
