--- src/sna/Makefile.am	Wed Apr  1 11:42:31 2015
+++ src/sna/Makefile.am	Wed Apr  1 11:43:16 2015
@@ -35,6 +35,7 @@
 
 noinst_LTLIBRARIES = libsna.la
 libsna_la_LDFLAGS = -pthread
+libsna_la_LDFLAGS += -R $(moduledir)
 libsna_la_LIBADD = $(UDEV_LIBS) -lm $(DRM_LIBS) brw/libbrw.la fb/libfb.la ../../libobj/libcompat.la
 
 libsna_la_SOURCES = \
--- src/Makefile.am	Wed Apr  1 11:55:38 2015
+++ src/Makefile.am	Wed Apr  1 11:56:12 2015
@@ -53,7 +53,7 @@
 
 if UXA
 SUBDIRS += uxa
-intel_drv_la_LIBADD += uxa/libuxa.la
+intel_drv_la_LIBADD += uxa/libuxa.la ./sna/libsna.la ./sna/brw/libbrw.la ./sna/fb/libfb.la -lXfont
 endif
 
 #NULL:=#
--- src/sna/sna_display.c	Wed Apr  1 11:51:57 2015
+++ src/sna/sna_display.c	Wed Apr  1 12:00:46 2015
@@ -39,6 +39,7 @@
 #include <errno.h>
 #include <poll.h>
 #include <ctype.h>
+#include <alloca.h>
 
 #include "sna.h"
 #include "sna_reg.h"
@@ -78,6 +79,8 @@
 	uint32_t pad[20];
 };
 
+/*extern void *alloca(); */
+
 #define KNOWN_MODE_FLAGS ((1<<14)-1)
 
 #ifndef MONITOR_EDID_COMPLETE_RAWDATA
