diff --git a/src/Makefile.am b/src/Makefile.am
index c036ae7..91e8029 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -72,7 +72,7 @@ AM_CFLAGS = -I$(top_srcdir)/include \
             $(CWARNFLAGS)
 
 
-libXi_la_LDFLAGS = -version-number 6:1:0 -no-undefined
+libXi_la_LDFLAGS = -version-number 5:0:0 -no-undefined
 
 libXiincludedir = $(includedir)/X11/extensions
 libXiinclude_HEADERS = $(top_srcdir)/include/X11/extensions/XInput.h \
diff --git a/src/Makefile.in b/src/Makefile.in
index 4ca3ac7..b66f7b3 100644
--- a/src/Makefile.in
+++ b/src/Makefile.in
@@ -340,7 +340,7 @@ AM_CFLAGS = -I$(top_srcdir)/include \
             $(MALLOC_ZERO_CFLAGS) \
             $(CWARNFLAGS)
 
-libXi_la_LDFLAGS = -version-number 6:1:0 -no-undefined
+libXi_la_LDFLAGS = -version-number 5:0:0 -no-undefined
 libXiincludedir = $(includedir)/X11/extensions
 libXiinclude_HEADERS = $(top_srcdir)/include/X11/extensions/XInput.h \
 		       $(top_srcdir)/include/X11/extensions/XInput2.h
