diff --git a/src/Makefile.am b/src/Makefile.am
index 806265c..ad41371 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -73,7 +73,7 @@ AM_CFLAGS = $(XI_CFLAGS) \
             $(CWARNFLAGS)
 
 
-libXi_la_LDFLAGS = -version-number 6:1:0 -no-undefined
+libXi_la_LDFLAGS = -version-number 5:0:0 -no-undefined
 
 libXiincludedir = $(includedir)/X11/extensions
 libXiinclude_HEADERS = $(top_srcdir)/include/X11/extensions/XInput.h \
diff --git a/src/Makefile.in b/src/Makefile.in
index b3c9830..776ba0c 100644
--- a/src/Makefile.in
+++ b/src/Makefile.in
@@ -343,7 +343,7 @@ AM_CFLAGS = $(XI_CFLAGS) \
             $(MALLOC_ZERO_CFLAGS) \
             $(CWARNFLAGS)
 
-libXi_la_LDFLAGS = -version-number 6:1:0 -no-undefined
+libXi_la_LDFLAGS = -version-number 5:0:0 -no-undefined
 libXiincludedir = $(includedir)/X11/extensions
 libXiinclude_HEADERS = $(top_srcdir)/include/X11/extensions/XInput.h \
 		       $(top_srcdir)/include/X11/extensions/XInput2.h
