--- src/riva_include.h	Mon Feb 25 12:17:24 2013
+++ src/riva_include.h	Mon Feb 25 12:17:54 2013
@@ -22,9 +22,6 @@
 /* All drivers initialising the SW cursor need this */
 #include "mipointer.h"
 
-/* All drivers implementing backing store need this */
-#include "mibstore.h"
-
 #include "micmap.h"
 
 #include "xf86DDC.h"
--- src/nv_include.h	Mon Feb 25 12:17:27 2013
+++ src/nv_include.h	Mon Feb 25 12:18:34 2013
@@ -24,9 +24,6 @@
 /* All drivers initialising the SW cursor need this */
 #include "mipointer.h"
 
-/* All drivers implementing backing store need this */
-#include "mibstore.h"
-
 #include "micmap.h"
 
 #include "xf86DDC.h"
--- src/g80_driver.c	Mon Feb 25 12:17:20 2013
+++ src/g80_driver.c	Mon Feb 25 12:17:33 2013
@@ -34,7 +34,6 @@
 #include <xf86Resources.h>
 #endif
 #include <mipointer.h>
-#include <mibstore.h>
 #include <micmap.h>
 #include <xf86cmap.h>
 #include <fb.h>
--- src/nv_driver.c	Mon Feb 25 12:20:33 2013
+++ src/nv_driver.c	Mon Feb 25 12:20:39 2013
@@ -2550,7 +2550,6 @@
     if (!pNv->NoAccel)
 	NVAccelInit(pScreen);
     
-    miInitializeBackingStore(pScreen);
     xf86SetBackingStore(pScreen);
     xf86SetSilkenMouse(pScreen);
 
--- src/riva_driver.c	Mon Feb 25 12:21:18 2013
+++ src/riva_driver.c	Mon Feb 25 12:21:24 2013
@@ -1168,7 +1168,6 @@
     if (!pRiva->NoAccel)
 	RivaAccelInit(pScreen);
     
-    miInitializeBackingStore(pScreen);
     xf86SetBackingStore(pScreen);
     xf86SetSilkenMouse(pScreen);
 
--- src/g80_driver.c	Mon Feb 25 12:20:40 2013
+++ src/g80_driver.c	Mon Feb 25 12:22:18 2013
@@ -832,7 +832,6 @@
         }
     }
 
-    miInitializeBackingStore(pScreen);
     xf86SetBackingStore(pScreen);
     xf86SetSilkenMouse(pScreen);
 
