--- src/Makefile.am	Fri Apr 12 11:17:54 2013
+++ src/Makefile.am	Fri Apr 12 11:18:14 2013
@@ -1,5 +1,7 @@
 lib_LTLIBRARIES = libxcb.la
 
+man3dir = $(mandir)/man3xcb
+
 EXTSOURCES =	xproto.c \
 		bigreq.c \
 		xc_misc.c

--- src/c_client.py	Tue Apr 30 08:57:32 2013
+++ src/c_client.py	Tue Apr 30 11:11:09 2013
@@ -2270,18 +2270,18 @@
     func_name = self.c_request_name if not aux else self.c_aux_name
 
     def create_link(linkname):
-        name = 'man/%s.3' % linkname
+        name = 'man/%s.3xcb' % linkname
         if manpaths:
             sys.stdout.write(name)
         f = open(name, 'w')
-        f.write('.so man3/%s.3' % func_name)
+        f.write('.so man3xcb/%s.3xcb' % func_name)
         f.close()
 
     if manpaths:
-        sys.stdout.write('man/%s.3 ' % func_name)
+        sys.stdout.write('man/%s.3xcb ' % func_name)
     # Our CWD is src/, so this will end up in src/man/
-    f = open('man/%s.3' % func_name, 'w')
-    f.write('.TH %s 3  %s "XCB" "XCB Requests"\n' % (func_name, today))
+    f = open('man/%s.3xcb' % func_name, 'w')
+    f.write('.TH %s 3xcb  %s "XCB" "XCB Requests"\n' % (func_name, today))
     # Left-adjust instead of adjusting to both sides
     f.write('.ad l\n')
     f.write('.SH NAME\n')
@@ -2652,10 +2652,10 @@
 
 def _man_event(self, name):
     if manpaths:
-        sys.stdout.write('man/%s.3 ' % self.c_type)
+        sys.stdout.write('man/%s.3xcb ' % self.c_type)
     # Our CWD is src/, so this will end up in src/man/
-    f = open('man/%s.3' % self.c_type, 'w')
-    f.write('.TH %s 3  %s "XCB" "XCB Events"\n' % (self.c_type, today))
+    f = open('man/%s.3xcb' % self.c_type, 'w')
+    f.write('.TH %s 3xcb  %s "XCB" "XCB Events"\n' % (self.c_type, today))
     # Left-adjust instead of adjusting to both sides
     f.write('.ad l\n')
     f.write('.SH NAME\n')
--- src/Makefile.am	Tue Apr 30 12:35:08 2013
+++ src/Makefile.am	Tue Apr 30 12:35:35 2013
@@ -229,8 +229,8 @@
 nodist_xcbinclude_HEADERS = $(EXTHEADERS)
 noinst_HEADERS = xcbint.h
 
-STATIC_MANS = man/xcb-examples.3 man/xcb-requests.3
-BUILT_MANS = man/xcb_*.3
+STATIC_MANS = # handled in Solaris Makefile
+BUILT_MANS = # handled in Solaris Makefile
 man_MANS = $(STATIC_MANS) $(BUILT_MANS)
 EXTRA_DIST = $(STATIC_MANS)
 
--- src/man/xcb-examples.3	Tue Apr 30 12:32:55 2013
+++ src/man/xcb-examples.3	Tue Apr 30 12:36:07 2013
@@ -1,4 +1,4 @@
-.TH xcb-examples 3 2011-12-11 "XCB" "XCB examples"
+.TH xcb-examples 3xcb 2011-12-11 "XCB" "XCB examples"
 .ad l
 .SH NAME
 xcb-examples \- manpage examples
@@ -52,8 +52,8 @@
 No line is longer than 80 characters (including indention).
 
 .SH SEE ALSO
-.BR xcb_connect (3),
-.BR xcb_get_property (3),
-.BR xcb_flush (3)
+.BR xcb_connect (3xcb),
+.BR xcb_get_property (3xcb),
+.BR xcb_flush (3xcb)
 .SH AUTHOR
 Michael Stapelberg <michael+xcb at stapelberg dot de>
--- src/man/xcb-requests.3	Tue Apr 30 12:32:59 2013
+++ src/man/xcb-requests.3	Tue Apr 30 12:36:32 2013
@@ -1,4 +1,4 @@
-.TH xcb-requests 3 2011-12-11 "XCB" "XCB examples"
+.TH xcb-requests 3xcb 2011-12-11 "XCB" "XCB examples"
 .ad l
 .SH NAME
 xcb-requests \- about request manpages
@@ -156,10 +156,10 @@
 .fi
 
 .SH SEE ALSO
-.BR xcb_map_window (3),
-.BR xcb_intern_atom (3),
-.BR xcb_list_fonts (3),
-.BR xcb_poll_for_event (3),
-.BR xcb_request_check (3)
+.BR xcb_map_window (3xcb),
+.BR xcb_intern_atom (3xcb),
+.BR xcb_list_fonts (3xcb),
+.BR xcb_poll_for_event (3xcb),
+.BR xcb_request_check (3xcb)
 .SH AUTHOR
 Michael Stapelberg <michael+xcb at stapelberg dot de>
