--- a/src/Makefile.am	Wed Feb  6 13:47:17 2013
+++ b/src/Makefile.am	Wed Feb  6 13:55:46 2013
@@ -8,7 +8,7 @@
 libxcb_la_LIBADD = $(NEEDED_LIBS) $(XDMCP_LIBS)
 libxcb_la_SOURCES = \
 		xcb_conn.c xcb_out.c xcb_in.c xcb_ext.c xcb_xid.c \
-		xcb_list.c xcb_util.c xcb_auth.c c_client.py
+		xcb_list.c xcb_util.c xcb_auth.c xia.c c_client.py
 nodist_libxcb_la_SOURCES = xproto.c bigreq.c xc_misc.c
 
 # Explanation for -version-info:
@@ -24,6 +24,8 @@
 
 # FIXME: find a way to autogenerate this from the XML files.
 
+EXTSOURCES += xia.c 
+
 EXTSOURCES += composite.c
 if BUILD_COMPOSITE
 lib_LTLIBRARIES += libxcb-composite.la
@@ -234,6 +236,6 @@
 CLEANFILES = $(EXTSOURCES) $(EXTHEADERS) $(BUILT_MANS)
 
 $(EXTSOURCES): c_client.py
-	$(PYTHON) $(srcdir)/c_client.py -p $(XCBPROTO_XCBPYTHONDIR) $(XCBPROTO_XCBINCLUDEDIR)/$(@:.c=.xml)
+	$(PYTHON) $(srcdir)/c_client.py -p $(PROTODIR)/$(XCBPROTO_XCBPYTHONDIR) $(PROTODIR)/$(XCBPROTO_XCBINCLUDEDIR)/$(@:.c=.xml)
 
 $(man_MANS): $(EXTSOURCES)
