diff --git a/src/Makefile.am b/src/Makefile.am
index 950de5c..627995c 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -8,7 +8,7 @@ AM_CFLAGS = $(CWARNFLAGS) $(NEEDED_CFLAGS) $(XDMCP_CFLAGS)
 libxcb_la_LIBADD = $(NEEDED_LIBS) $(XDMCP_LIBS)
 libxcb_la_SOURCES = \
 		xcb_conn.c xcb_out.c xcb_in.c xcb_ext.c xcb_xid.c \
-		xcb_list.c xcb_util.c xcb_auth.c c_client.py
+		xcb_list.c xcb_util.c xcb_auth.c xia.c c_client.py
 nodist_libxcb_la_SOURCES = xproto.c bigreq.c xc_misc.c
 
 # Explanation for -version-info:
@@ -216,6 +216,7 @@ libxcb_xvmc_la_LIBADD = $(XCB_LIBS)
 nodist_libxcb_xvmc_la_SOURCES = xvmc.c xvmc.h
 endif
 
+EXTSOURCES += xia.c
 
 EXTHEADERS=$(EXTSOURCES:.c=.h)
 xcbinclude_HEADERS = xcb.h xcbext.h
@@ -229,4 +230,4 @@ BUILT_SOURCES = $(EXTSOURCES)
 CLEANFILES = $(EXTSOURCES) $(EXTHEADERS)
 
 $(EXTSOURCES): c_client.py
-	$(PYTHON) $(srcdir)/c_client.py -p $(XCBPROTO_XCBPYTHONDIR) $(XCBPROTO_XCBINCLUDEDIR)/$(@:.c=.xml)
+	$(PYTHON) $(srcdir)/c_client.py -p $(PROTODIR)/$(XCBPROTO_XCBPYTHONDIR) $(PROTODIR)/$(XCBPROTO_XCBINCLUDEDIR)/$(@:.c=.xml)
