diff --git a/doc/Makefile.am b/doc/Makefile.am
index dbe2e0b..fff55cc 100644
--- a/doc/Makefile.am
+++ b/doc/Makefile.am
@@ -31,8 +31,8 @@ install-data-local:
 	  $(mkinstalldirs) '$(DESTDIR)$(htmldir)'; \
 	fi
 	@for d in $(docdirs); do \
-	  echo "cp -pR $$d '$(DESTDIR)$(htmldir)/'"; \
-	  cp -pR $$d '$(DESTDIR)$(htmldir)/'; \
+	  echo "/usr/gnu/bin/cp -pR $$d '$(DESTDIR)$(htmldir)/'"; \
+	  /usr/gnu/bin/cp -pR $$d '$(DESTDIR)$(htmldir)/'; \
 	done
 
 uninstall-local:
