--- src/radeon_driver.c.orig	Sat Jul 28 19:48:26 2007
+++ src/radeon_driver.c	Sat Jul 28 22:59:42 2007
@@ -9874,7 +9874,7 @@
        info->CRT2pScrn->monitor = xalloc(sizeof(MonRec));
        if(info->CRT2pScrn->monitor) {
           DisplayModePtr tempm = NULL, currentm = NULL, newm = NULL;
-          memcpy(info->CRT2pScrn->monitor, pScrn->monitor, sizeof(MonRec));
+          memset(info->CRT2pScrn->monitor, 0, sizeof(MonRec));
           info->CRT2pScrn->monitor->DDC = NULL;
 	  info->CRT2pScrn->monitor->Modes = NULL;
 	  info->CRT2pScrn->monitor->id = "CRT2 Monitor";
