--- src/vesa.c	Mon Feb 25 12:28:58 2013
+++ src/vesa.c	Mon Feb 25 12:29:15 2013
@@ -49,9 +49,6 @@
 /* All drivers initialising the SW cursor need this */
 #include "mipointer.h"
 
-/* All drivers implementing backing store need this */
-#include "mibstore.h"
-
 /* Colormap handling */
 #include "micmap.h"
 #include "xf86cmap.h"
@@ -1081,7 +1078,6 @@
     VESADGAInit(pScrn, pScreen);
 
     xf86SetBlackWhitePixels(pScreen);
-    miInitializeBackingStore(pScreen);
     xf86SetBackingStore(pScreen);
 
     /* software cursor */
