diff -urp -x '*~' -x '*.orig' utils/rss-glx_install.pl utils/rss-glx_install.pl
--- utils/rss-glx_install.pl	2009-05-20 20:16:33.348436000 -0700
+++ utils/rss-glx_install.pl	2009-05-20 20:16:36.684360000 -0700
@@ -1,10 +1,10 @@
 #!/usr/bin/env perl
 use strict;
 
-if (`ps x | grep ' xscreensaver ' | grep -v grep`) {
-	print "XScreenSaver seems to be running. Exit it before running this script.\n";
-	exit;
-}
+#if (`ps x | grep ' xscreensaver ' | grep -v grep`) {
+#	print "XScreenSaver seems to be running. Exit it before running this script.\n";
+#	exit;
+#}
 
 my $screensavers = { };
 
@@ -31,15 +31,15 @@ $screensavers->{'lorenz'}	= '  GL:      
 $screensavers->{'drempels'}	= '  GL:               "Drempels"  drempels --root                             \\n\\';
 $screensavers->{'feedback'}	= '  GL:               "Feedback"  feedback --root                             \\n\\';
 
-open XSCREENSAVER, "$ENV{'HOME'}/.xscreensaver";
+open XSCREENSAVER, "$ARGV[0]";
 my @xscreensaver_config_file = <XSCREENSAVER>;
 close XSCREENSAVER;
 
-open XSCREENSAVER, ">$ENV{'HOME'}/.xscreensaver";
+open XSCREENSAVER, ">$ARGV[0]";
 
 my $programs_section_flag = 0;
 foreach my $line (@xscreensaver_config_file) {
-	if ($line =~ /^programs:/) {
+	if ($line =~ /^\*programs:/) {
 		$programs_section_flag = 1;
 	} elsif ($programs_section_flag) {
 		if ($line =~ /\\\s+/) {
