diff -urp -x '*~' -x '*.orig' utils/rss-glx_install.pl utils/rss-glx_install.pl
--- utils/rss-glx_install.pl	2009-04-15 17:30:13.075036000 -0700
+++ utils/rss-glx_install.pl	2009-04-15 17:30:15.087953000 -0700
@@ -1,10 +1,10 @@
 #!/usr/bin/env perl
 use strict;
 
-if (`ps x | grep ' xscreensaver ' | grep -v grep`) {
-	print "XScreenSaver seems to be running. Exit it before running this script.\n";
-	exit;
-}
+#if (`ps x | grep ' xscreensaver ' | grep -v grep`) {
+#	print "XScreenSaver seems to be running. Exit it before running this script.\n";
+#	exit;
+#}
 
 my $screensavers = { };
 
@@ -29,15 +29,15 @@ $screensavers->{'spirographx'} 	= '  GL:
 $screensavers->{'matrixview'} 	= '  GL:             "MatrixView"  matrixview --root -i /usr/share/pixmaps/faces \\n\\';
 $screensavers->{'lorenz'}	= '  GL:                 "Lorenz"  lorenz --root                               \\n\\';
 
-open XSCREENSAVER, "$ENV{'HOME'}/.xscreensaver";
+open XSCREENSAVER, "$ARGV[0]";
 my @xscreensaver_config_file = <XSCREENSAVER>;
 close XSCREENSAVER;
 
-open XSCREENSAVER, ">$ENV{'HOME'}/.xscreensaver";
+open XSCREENSAVER, ">$ARGV[0]";
 
 my $programs_section_flag = 0;
 foreach my $line (@xscreensaver_config_file) {
-	if ($line =~ /^programs:/) {
+	if ($line =~ /^\*programs:/) {
 		$programs_section_flag = 1;
 	} elsif ($programs_section_flag) {
 		if ($line =~ /\\\s+/) {
