--- save_xrdb.c Wed May  7 17:08:05 2008
+++ xrdb.c      Wed May  7 17:08:20 2008
@@ -129,7 +129,7 @@
 #endif
 int oper = OPLOAD;
 char *editFile = NULL;
-char *cpp_program = CPP;
+char *cpp_program = NULL;
 char *backup_suffix = BACKUP_SUFFIX;
 Bool dont_execute = False;
 String defines;

--- xrdb.c	Mon Jul 10 19:54:07 2006
+++ filemerge.out	Tue Jan 29 13:24:42 2008
@@ -910,7 +910,11 @@
 
     if (!filename &&
 #ifdef PATHETICCPP
+/*
+4035378
 	need_real_defines
+*/
+	(oper == OPLOAD || oper == OPMERGE || oper == OPOVERRIDE)
 #else
 	(oper == OPLOAD || oper == OPMERGE || oper == OPOVERRIDE) &&
 	(whichResources == RALL || whichResources == RSCREENS)

--- save_xrdb.c	Thu May  8 09:24:00 2008
+++ xrdb.c	Thu May  8 09:27:19 2008
@@ -142,6 +142,19 @@
 Buffer buffer;
 Entries newDB;
 
+#ifdef CPP
+    static char* cpp_locations[] = {
+	CPP, 
+	"/usr/lib/cpp", 
+	"/usr/ccs/lib/cpp"
+};
+#else
+    static char* cpp_locations[] = {
+	"/usr/lib/cpp", 
+	"/usr/ccs/lib/cpp"
+};
+#endif /* CPP */
+
 static void fatal(char *, ...);
 static void addstring ( String *arg, const char *s );
 static void FormatEntries ( Buffer *buffer, Entries *entries );
@@ -772,6 +785,15 @@
     int retainProp = 0;
     FILE *fp = NULL;
     Bool need_newline;
+    int  number_of_elements = (sizeof cpp_locations) / 
+			      (sizeof cpp_locations[0]);
+    int j;
+    for (j = 0; j < number_of_elements; j++) {
+      if (access(cpp_locations[j], X_OK) == 0) {
+          cpp_program = cpp_locations[j];
+          break;
+      }
+    }
 
     ProgramName = argv[0];

--- xrdb.c.old	Tue Aug  5 09:51:22 2008
+++ xrdb.c	Tue Aug  5 09:52:21 2008
@@ -489,6 +489,8 @@
 static int 
 Resolution(int pixels, int mm)
 {
+    if (mm == 0) 	/* An occasionally seen ddx bug, but pass it on to others */
+	return (0);
     return ((pixels * 100000 / mm) + 50) / 100;
 }
 
