--- src/vmware.c	Thu Oct 11 14:04:32 2012
+++ src/vmware.c	Thu Oct 11 14:06:59 2012
@@ -37,6 +37,7 @@
 #include "svga_modes.h"
 #include "vmware_bootstrap.h"
 #include "vmware_common.h"
+#include "compat-api.h"
 
 #ifndef HAVE_XORG_SERVER_1_5_0
 #include <xf86_ansic.h>
@@ -1089,9 +1090,9 @@
 }
 
 static Bool
-VMWARECloseScreen(int scrnIndex, ScreenPtr pScreen)
+VMWARECloseScreen(CLOSE_SCREEN_ARGS_DECL)
 {
-    ScrnInfoPtr pScrn = xf86Screens[scrnIndex];
+    ScrnInfoPtr pScrn = xf86ScreenToScrn(pScreen);
     VMWAREPtr pVMWARE = VMWAREPTR(pScrn);
     ScreenPtr save = &pVMWARE->ScrnFuncs;
 
@@ -1119,7 +1120,7 @@
     pScrn->DriverFunc = NULL;
 #endif
 
-    return (*pScreen->CloseScreen)(scrnIndex, pScreen);
+    return (*pScreen->CloseScreen)(CLOSE_SCREEN_ARGS);
 }
 
 static Bool
