From 8bb6a7c3eaefe9449a1ddf587f57af70635cd1bf Mon Sep 17 00:00:00 2001
From: Alan Coopersmith <alan.coopersmith@sun.com>
Date: Mon, 2 Nov 2009 16:34:14 -0800
Subject: [PATCH] Fix CFLAGS/CPPFLAGS so lint works properly

Signed-off-by: Alan Coopersmith <alan.coopersmith@sun.com>
---
 configure.ac    |    2 --
 src/Makefile.am |    4 +++-
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/configure.ac b/configure.ac
index 8c674a9..fa93255 100644
--- a/configure.ac
+++ b/configure.ac
@@ -36,7 +36,6 @@ XORG_DEFAULT_OPTIONS
 
 # Checks for pkg-config packages
 PKG_CHECK_MODULES(XEXT, [xproto >= 7.0.13] [x11 >= 1.1.99.1] [xextproto >= 7.0.99.2])
-XEXT_CFLAGS="$CWARNFLAGS $XEXT_CFLAGS"
 AC_SUBST(XEXT_CFLAGS)
 AC_SUBST(XEXT_LIBS)
 
@@ -53,7 +52,6 @@ XORG_PROG_RAWCPP
 dnl Allow checking code with lint, sparse, etc.
 XORG_WITH_LINT
 XORG_LINT_LIBRARY([Xext])
-LINT_FLAGS="${LINT_FLAGS} ${XEXT_CFLAGS}"
 
 AC_OUTPUT([Makefile
            man/Makefile
diff --git a/src/Makefile.am b/src/Makefile.am
index 6a7d379..b9b43f7 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -1,10 +1,12 @@
 lib_LTLIBRARIES=libXext.la
 
-AM_CFLAGS=\
+AM_CPPFLAGS=\
           -I$(top_srcdir)/include \
           -I$(top_srcdir)/include/X11/extensions\
           $(XEXT_CFLAGS) $(MALLOC_ZERO_CFLAGS)
 
+AM_CFLAGS=$(CWARNFLAGS)
+
 libXext_la_LDFLAGS = -version-number $(XEXT_SOREV) -no-undefined
 
 libXext_la_LIBADD = $(XEXT_LIBS)
-- 
1.5.6.5

