diff -urp -x '*~' -x '*.orig' man/Makefile.am man/Makefile.am
--- man/Makefile.am	2009-10-21 20:26:57.000000000 -0700
+++ man/Makefile.am	2009-12-02 08:59:46.287398680 -0800
@@ -3,6 +3,16 @@ libmandir = $(LIB_MAN_DIR)
 LIB_MAN_DIR_SUFFIX = $(LIB_MAN_DIR:@mandir@/man%=%)
 
 libman_PRE = \
+	XReadScreen.man \
+	XSolarisOvlCopyAreaAndPaintType.man \
+	XSolarisOvlCopyPaintType.man \
+	XSolarisOvlCreateWindow.man \
+	XSolarisOvlGetPaintType.man \
+	XSolarisOvlIsOverlayWindow.man \
+	XSolarisOvlSelectPair.man \
+	XSolarisOvlSelectPartner.man \
+	XSolarisOvlSetPaintType.man \
+	XSolarisOvlSetWindowTransparent.man \
 	DBE.man \
 	DPMSCapable.man \
 	DPMSDisable.man \
diff -urp -x '*~' -x '*.orig' src/Makefile.am src/Makefile.am
--- src/Makefile.am	2009-12-02 08:59:46.063768361 -0800
+++ src/Makefile.am	2009-12-02 08:59:46.287602685 -0800
@@ -66,3 +66,13 @@ lintlib_DATA = $(LINTLIB)
 $(LINTLIB): $(libXext_la_SOURCES)
 	$(LINT) -y -oXext -x $(ALL_LINT_FLAGS) $(libXext_la_SOURCES) 
 endif MAKE_LINT_LIB
+
+# Solaris specific additions:
+libXext_la_SOURCES += \
+	AllPlanes.c \
+	FBPM.c \
+	TransOvl.c \
+	XGrabWin.c \
+	XPanoramiX.c
+
+libXextinclude_HEADERS += $(top_srcdir)/include/X11/extensions/xinerama.h
