diff -urp -x '*~' -x '*.orig' Makefile.am Makefile.am
--- Makefile.am	2008-08-05 14:58:52.000000000 -0700
+++ Makefile.am	2009-07-30 18:11:28.517520000 -0700
@@ -53,6 +53,7 @@ PROGCPPDEFS = \
 	-DXINIT=@XINIT@
 
 CPP_FILES_FLAGS = \
+	-Dsun \
 	-DXINITDIR=$(XINITDIR) $(PROGCPPDEFS) -DLIBDIR=$(libdir) \
 	-DSHELL_CMD=$(SHELL_CMD) $(STARTX_COOKIE_FLAGS) \
 	-D__libexecdir__="$(libexecdir)" \
diff -urp -x '*~' -x '*.orig' Makefile.in Makefile.in
--- Makefile.in	2008-12-17 15:24:32.000000000 -0800
+++ Makefile.in	2009-07-30 18:11:28.518046000 -0700
@@ -307,6 +307,7 @@ PROGCPPDEFS = \
 	-DXINIT=@XINIT@
 
 CPP_FILES_FLAGS = \
+	-Dsun \
 	-DXINITDIR=$(XINITDIR) $(PROGCPPDEFS) -DLIBDIR=$(libdir) \
 	-DSHELL_CMD=$(SHELL_CMD) $(STARTX_COOKIE_FLAGS) \
 	-D__libexecdir__="$(libexecdir)" \
diff -urp -x '*~' -x '*.orig' xinit.c xinit.c
--- xinit.c	2008-11-22 10:52:00.000000000 -0800
+++ xinit.c	2009-07-30 18:11:28.522338000 -0700
@@ -199,6 +199,11 @@ static void set_environment ( void );
 static void Fatal(char *msg);
 static void Error ( char *fmt, ... );
 
+/* Bug 4284529, vivekp, this flag is set by the SIGUSR1
+   signal handler in xinit, it tells Xserver is ready to accept connection.
+*/
+volatile static int  received_SIGUSR1_from_xserver   = 0;
+
 #ifdef RETSIGTYPE /* autoconf AC_TYPE_SIGNAL */
 # define SIGVAL RETSIGTYPE
 #endif /* RETSIGTYPE */
@@ -221,6 +226,14 @@ sigAlarm(int sig)
 static SIGVAL
 sigUsr1(int sig)
 {
+        /* Bug 4284529, vivekp, set this flag if received SIGUSR1 from X server
+           so that we know X server is ready to accept connections.
+
+          SIGUSR1 is not received by xinit if Xsun is executed thru shell that
+           is in case of executing X server thru .xserverrc.
+        */
+       received_SIGUSR1_from_xserver   = 1;
+
 #if defined(SYSV) || defined(SVR4) || defined(linux) || defined(__UNIXOS2__) || defined(__APPLE__)
 	signal (sig, sigUsr1);
 #endif
@@ -567,6 +580,12 @@ startServer(char *server[])
 	sigaddset(&mask, SIGUSR1);
 	sigprocmask(SIG_BLOCK, &mask, &old);
 
+        /* Bug 4284529, vivekp, reset it, this flag gets set in
+           SIGUSR1 handler, sigUsr1().
+        */
+        received_SIGUSR1_from_xserver  = 0;
+
+
 	serverpid = fork(); 
 
 	switch(serverpid) {
@@ -654,6 +673,17 @@ startServer(char *server[])
 		if (!sigismember(&pendings, SIGUSR1))
 #endif /* __UNIXOS2__ */
 		sigsuspend(&old);
+
+               /* Bug 4284529, vivekp, check if SIGUSR1 is already received,
+                  no need to pause. SIGUSR1 is never received only in the case
+                  where Xserver is started thru $HOME/.xserverrc file.
+                  This fix corrects last few fixes from version 1.6, 1.3, 1.4
+                  of xinit.c. That is bug 4284529, 4245301 and 4226277.
+                */
+                if (!received_SIGUSR1_from_xserver)
+                   pause ();
+
+
 		alarm (0);
 		sigprocmask(SIG_SETMASK, &old, NULL);
 
@@ -820,7 +850,12 @@ shutdown(void)
 	if (serverpid < 0)
 		return;
 	errno = 0;
+#ifndef i386
 	if (killpg(serverpid, SIGTERM) < 0) {
+#else
+        if (kill(serverpid, SIGTERM) < 0) {
+#endif
+
 		if (errno == EPERM)
 			Fatal("Can't kill X server\r\n");
 		if (errno == ESRCH)
@@ -844,6 +879,13 @@ shutdown(void)
 		fprintf (stderr, "\r\n");
 		Fatal("Can't kill server\r\n");
 	}
+	else {
+		/* B U G : 4221128 */
+          	/* Restore keyboard mode. */
+       		fprintf (stderr, "\r\nRestoring keyboard mode\r\n");
+       		execlp ("kbd_mode", "kbd_mode", "-a");
+    	}
+
 	fprintf (stderr, "\r\n");
 	return;
 }
diff -urp -x '*~' -x '*.orig' xinitrc.cpp xinitrc.cpp
--- xinitrc.cpp	2008-04-08 17:25:11.000000000 -0700
+++ xinitrc.cpp	2009-07-30 18:11:28.522525000 -0700
@@ -97,7 +97,7 @@ fi
 #endif
 
 TWM &
-XCLOCK -geometry 50x50-1+1 &
+XCOMM XCLOCK -geometry 50x50-1+1 &
 XTERM -geometry 80x50+494+51 &
 XTERM -geometry 80x20+494-0 &
 exec XTERM -geometry 80x66+0+0 -name login
