diff --git a/Makefile.am b/Makefile.am
index 3867bea..8288450 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -49,6 +49,7 @@ PROGCPPDEFS = \
 	-DXINIT=@XINIT@
 
 CPP_FILES_FLAGS = \
+	-Dsun \
 	-DXINITDIR=$(XINITDIR) $(PROGCPPDEFS) -DLIBDIR=$(libdir) \
 	-DSHELL_CMD=$(SHELL_CMD) $(STARTX_COOKIE_FLAGS) \
 	-D__libexecdir__="$(libexecdir)" \
diff --git a/Makefile.in b/Makefile.in
index e6d2356..e53e951 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -356,6 +356,7 @@ PROGCPPDEFS = \
 	-DXINIT=@XINIT@
 
 CPP_FILES_FLAGS = \
+	-Dsun \
 	-DXINITDIR=$(XINITDIR) $(PROGCPPDEFS) -DLIBDIR=$(libdir) \
 	-DSHELL_CMD=$(SHELL_CMD) $(STARTX_COOKIE_FLAGS) \
 	-D__libexecdir__="$(libexecdir)" \
diff --git a/xinit.c b/xinit.c
index 42ff008..e8b7559 100644
--- a/xinit.c
+++ b/xinit.c
@@ -629,6 +629,26 @@ shutdown(void)
 
     if (processTimeout(3, "server to die"))
         Fatalx("X server refuses to die");
+#ifdef sun
+    else {
+      	/* Restore keyboard mode. */
+	serverpid = fork();
+	switch (serverpid) {
+	case 0:
+	    execlp ("kbd_mode", "kbd_mode", "-a");
+	    Fatal("Unable to run program \"%s\"", "kbd_mode");
+	    break;
+
+	case 1:
+	    Error("fork failed");
+	    break;
+
+	default:
+	    fprintf (stderr, "\r\nRestoring keyboard mode\r\n");
+	    processTimeout(1, "kbd_mode");
+	}
+    }
+#endif
 }
 
 static void
