--- xinitrc.cpp.orig	Wed Mar 16 10:11:59 2011
+++ xinitrc.cpp	Wed Mar 16 10:13:12 2011
@@ -39,6 +39,19 @@
     XMODMAP "$usermodmap"
 fi
 
+XCOMM
+XCOMM  If OpenSolaris branded desktop is available, run it
+XCOMM
+if [ -x /etc/gdm/Xsession ] && [ -x /usr/bin/dtstart ] ; then
+  exec /etc/gdm/Xsession /usr/bin/dtstart jds
+fi
+if [ -x /etc/X11/gdm/Xsession ] && [ -x /usr/bin/dtstart ] ; then
+  exec /etc/X11/gdm/Xsession /usr/bin/dtstart jds
+fi
+if [ -x /usr/dt/config/Xsession.jds ]; then
+  exec /usr/dt/config/Xsession.jds
+fi
+
 XCOMM start some nice programs
 
 #if defined(__SCO__) || defined(__UNIXWARE__)
