--- man/xinit.man	Mon Aug 27 12:02:48 2012
+++ man/xinit.man	Mon Aug 27 12:03:00 2012
@@ -121,7 +121,7 @@
 This will use the command \fI\.\/Xorg \-l \-c\fP to start the server and will
 append the arguments \fI\-e widgets\fP to the default \fIxterm\fP command.
 .TP 8
-.B "xinit /usr/ucb/rsh fasthost cpupig \-display ws:1 \-\^\-  :1 \-a 2 \-t 5"
+.B "xinit /usr/bin/rsh fasthost cpupig \-display ws:1 \-\^\-  :1 \-a 2 \-t 5"
 This will start a server named \fIX\fP on display 1 with the arguments
 \fI\-a 2 \-t 5\fP.  It will then start a remote shell on the machine
 \fBfasthost\fP in which it will run the command \fIcpupig\fP, telling it
