--- Makefile.am	Mon Sep 17 14:03:38 2012
+++ Makefile.am	Tue Jul  9 12:56:59 2013
@@ -131,7 +131,7 @@
 	src/libnurbs/nurbtess/sampledLine.cc			\
 	src/libnurbs/nurbtess/searchTree.cc
 libGLU_la_LIBADD = $(GL_LIBS) $(OSMESA_LIBS)
-libGLU_la_LDFLAGS = -version-number $(GLU_LIBRARY_VERSION) -no-undefined -export-symbols-regex '^glu'
+libGLU_la_LDFLAGS = -version-number $(GLU_LIBRARY_VERSION) -no-undefined -export-symbols-regex 'glu'
 
 EXTRA_DIST =							\
 	include/GL/glu.h					\
