diff -urp -x '*~' -x '*.orig' src/xftdraw.c src/xftdraw.c
--- src/xftdraw.c	2009-01-29 15:19:09.000000000 -0800
+++ src/xftdraw.c	2009-11-12 08:16:50.072683872 -0800
@@ -303,7 +303,15 @@ XftDrawSrcPicture (XftDraw *draw, _Xcons
 
     if (!info)
 	return 0;
-    
+
+    /*
+     * Sometimes DDX'en don't play fair.
+     *
+     */
+
+    if (!info->solidFormat)
+	return 0; 
+
     /*
      * Monochrome targets require special handling; the PictOp controls
      * the color, and the color must be opaque
