diff -urp -x '*~' -x '*.orig' src/xftfreetype.c src/xftfreetype.c
--- src/xftfreetype.c	2008-05-12 14:31:03.000000000 -0700
+++ src/xftfreetype.c	2009-07-30 17:06:15.114484000 -0700
@@ -561,7 +561,20 @@ XftFontInfoFill (Display *dpy, _Xconst F
     {
 	fi->load_flags |= FT_LOAD_NO_HINTING;
     }
-
+#ifdef FC_HINT_STYLE
+#ifdef FT_CONFIG_ADD_FULL_HINTING
+    /*
+     * So long as we are not in NO_HINTING, then check to see if the user has
+     * selected Full hinting mode. If they have, then set the new FT flag to
+     * use full hinting.
+     */
+    else
+    {
+	if ( hint_style == FC_HINT_FULL )
+	    fi->load_flags |= FT_LOAD_USE_FULL_HINTING;
+    }
+#endif /* FT_CONFIG_ADD_FULL_HINTING */
+#endif /* FC_HINT_STYLE */
     /* Figure out the load target, which modifies the hinting
      * behavior of FreeType based on the intended use of the glyphs.
      */
