--- xkbprint.man	Thu Jun 12 15:36:50 2008
+++ xkbprint.man	Thu Jun 12 15:40:08 2008
@@ -13,7 +13,7 @@
 .PP
 The
 .I xkbprint
-comman generates a printable or encapsulated PostScript description of the 
+command generates a printable or encapsulated PostScript description of the 
 XKB keyboard description specified by \fIsource\fP.  The \fIsource\fP can be 
 any compiled keymap (.xkm) file that includes a geometry description or an 
 X display specification.  If an \fIoutput_file\fP is specified, xkbprint 
