--- po/Makefile.in.in	Mon Sep 18 13:54:18 2006
+++ po/Makefile.in.in	Mon Dec 15 15:13:15 2008
@@ -84,6 +84,9 @@
 #       refering to $(srcdir) anyway.  So I just commented (my half-fixed
 #       version of) this rule out.
 #
+# The following line is needed to keep the intltool.m4 autoconf macros from
+# throwing an error message:
+# INTLTOOL_MAKEFILE
 
 GETTEXT_PACKAGE = @GETTEXT_PACKAGE@
 PACKAGE = @PACKAGE@
