diff -urp -x '*~' -x '*.orig' fontconfig-2.7.3/conf.d/30-metric-aliases.conf fontconfig-2.7.3/conf.d/30-metric-aliases.conf
--- fontconfig-2.7.3/conf.d/30-metric-aliases.conf	2009-06-28 13:48:47.000000000 -0400
+++ fontconfig-2.7.3/conf.d/30-metric-aliases.conf	2009-08-26 17:14:48.048084000 -0400
@@ -207,5 +207,13 @@
 	  </accept>
 	</alias>
 
+      <!-- StarOffice formerly included Arial Narrow - Solaris includes Arial Narrow MT instead. -->
+      <alias>
+          <family>Arial Narrow</family>
+          <accept>
+              <family>Arial Narrow MT</family>
+              <family>Nimbus Sans L Condensed</family>
+          </accept>
+      </alias>
 
 </fontconfig>
diff -urp -x '*~' -x '*.orig' fontconfig-2.7.3/conf.d/40-nonlatin.conf fontconfig-2.7.3/conf.d/40-nonlatin.conf
--- fontconfig-2.7.3/conf.d/40-nonlatin.conf	2009-06-28 13:48:47.000000000 -0400
+++ fontconfig-2.7.3/conf.d/40-nonlatin.conf	2009-08-26 17:18:21.871571000 -0400
@@ -18,6 +18,17 @@
 		<family>Zar</family>
 		<family>Titr</family>
 		<family>Jadid</family>
+                <family>HG-PMinchoL-Sun</family>
+                <family>HG-MinchoL-Sun</family>
+                <family>IPAPMincho</family>
+                <family>IPAMincho</family>
+                <family>Sazanami Mincho</family>
+                <family>Sazanami Gothic</family>
+                <family>FZSongTi</family>
+                <family>FZMingTi</family>
+                <family>KacstQurn</family>
+                <family>SunDotum</family>
+                <family>AR PL ShanHeiSun Uni</family>
 		<family>Kochi Mincho</family>
 		<family>AR PL SungtiL GB</family>
 		<family>AR PL Mingti2L Big5</family>
@@ -43,6 +54,15 @@
 		<family>Roya</family>
 		<family>Koodak</family>
 		<family>Terafik</family>
+                <family>HG-PGothicB-Sun</family>
+                <family>HG-GothicB-Sun</family>
+                <family>IPAPGothic</family>
+                <family>IPAGothic</family>
+                <family>Sazanami Gothic</family>
+                <family>FZSongTi</family>
+                <family>FZMingTi</family>
+                <family>KacstQurn</family>
+                <family>SunDotum</family>
 		<family>Kochi Gothic</family>
 		<family>AR PL KaitiM GB</family>
 		<family>AR PL KaitiM Big5</family>
@@ -61,6 +81,7 @@
   Monospace faces
  -->
  	<alias>
+                <family>SunDotumChe</family>
 		<family>NSimSun</family>
 		<family>ZYSong18030</family>
 		<family>FreeMono</family>
diff -urp -x '*~' -x '*.orig' fontconfig-2.7.3/conf.d/45-latin.conf fontconfig-2.7.3/conf.d/45-latin.conf
--- fontconfig-2.7.3/conf.d/45-latin.conf	2009-07-10 13:08:51.000000000 -0400
+++ fontconfig-2.7.3/conf.d/45-latin.conf	2009-08-26 17:20:58.512527000 -0400
@@ -10,6 +10,7 @@
   Serif faces
  -->
 	<alias>
+                <family>Lucida Bright</family>
 		<family>Bitstream Vera Serif</family>
 		<family>DejaVu Serif</family>
 		<family>Liberation Serif</family>
@@ -25,6 +26,7 @@
   Sans-serif faces
  -->
 	<alias>
+                <family>Lucida Sans Typewriter</family>
 		<family>Bitstream Vera Sans</family>
 		<family>DejaVu Sans</family>
 		<family>Liberation Sans</family>
@@ -41,6 +43,7 @@
   Monospace faces
  -->
  	<alias>
+                <family>Lucida Sans Typewriter</family>
 		<family>Bitstream Vera Sans Mono</family>
 		<family>DejaVu Sans Mono</family>
 		<family>Liberation Mono</family>
diff -urp -x '*~' -x '*.orig' fontconfig-2.7.3/conf.d/65-nonlatin.conf fontconfig-2.7.3/conf.d/65-nonlatin.conf
--- fontconfig-2.7.3/conf.d/65-nonlatin.conf	2009-06-28 13:48:47.000000000 -0400
+++ fontconfig-2.7.3/conf.d/65-nonlatin.conf	2009-08-26 17:29:28.727064000 -0400
@@ -4,6 +4,7 @@
 	<alias>
 		<family>serif</family>
 		<prefer>
+                        <family>EUDC</family>
 			<family>Artsounk</family> <!-- armenian -->
 			<family>BPG UTF8 M</family> <!-- georgian -->
 			<family>Kinnari</family> <!-- thai -->
@@ -34,10 +35,18 @@
 			<family>ZYSong18030</family> <!-- han (zh-cn,zh-tw) -->
 			<family>HanyiSong</family> <!-- han (zh-cn,zh-tw) -->
 			<family>MgOpen Canonica</family>
-			<family>Sazanami Mincho</family>
+                        <family>HG-PMinchoL-Sun</family>
+                        <family>HG-MinchoL-Sun</family>
 			<family>IPAMonaMincho</family>
+                        <family>IPAPMincho</family>
 			<family>IPAMincho</family>
-			<family>Kochi Mincho</family>
+                        <family>Sazanami Mincho</family>
+                        <family>Sazanami Gothic</family>
+                        <family>FZSongTi</family>
+                        <family>FZMingTi</family>
+                        <family>KacstQurn</family>
+                        <family>SunDotum</family>
+                        <family>AR PL ShanHeiSun Uni</family>
 			<family>AR PL SungtiL GB</family>
 			<family>AR PL Mingti2L Big5</family>
  			<family>AR PL Zenkai Uni</family>
@@ -69,6 +78,7 @@
 	<alias>
 		<family>sans-serif</family>
 		<prefer>
+                        <family>EUDC</family>
 			<family>Nachlieli</family> <!-- hebrew -->
 			<family>Lucida Sans Unicode</family>
 			<family>Yudit Unicode</family>
@@ -105,10 +115,16 @@
 			<family>AR PL New Sung</family> <!-- han (zh-cn,zh-tw) -->
 			<family>MgOpen Modata</family>
 			<family>VL Gothic</family>
+                        <family>HG-PGothicB-Sun</family>
+                        <family>HG-GothicB-Sun</family>
 			<family>IPAMonaGothic</family>
+                        <family>IPAPGothic</family>
 			<family>IPAGothic</family>
 			<family>Sazanami Gothic</family>
-			<family>Kochi Gothic</family>
+                        <family>FZSongTi</family>
+                        <family>FZMingTi</family>
+                        <family>KacstQurn</family>
+                        <family>SunDotum</family>
 			<family>AR PL KaitiM GB</family>
 			<family>AR PL KaitiM Big5</family>
  			<family>AR PL ShanHeiSun Uni</family>
@@ -143,12 +159,18 @@
 	<alias>
 		<family>monospace</family>
 		<prefer>
+                        <family>EUDC</family>
 			<family>Miriam Mono</family> <!-- hebrew -->
 			<family>VL Gothic</family>
+                        <family>HG-GothicB-Sun</family>
+                        <family>HG-MinchoL-Sun</family>
 			<family>IPAMonaGothic</family>
 			<family>IPAGothic</family>
 			<family>Sazanami Gothic</family>
-			<family>Kochi Gothic</family>
+                        <family>FZSongTi</family>
+                        <family>FZMingTi</family>
+                        <family>KacstQurn</family>
+                        <family>SunDotumChe</family>
 			<family>AR PL KaitiM GB</family>
 			<family>MS Gothic</family> <!-- han (ja) -->
 			<family>UmePlus Gothic</family> <!-- han (ja) -->
