diff -urp -x '*~' -x '*.orig' conf.d/30-metric-aliases.conf conf.d/30-metric-aliases.conf
--- conf.d/30-metric-aliases.conf	2007-11-05 15:38:03.000000000 -0800
+++ conf.d/30-metric-aliases.conf	2008-04-25 15:48:20.630200000 -0700
@@ -208,4 +208,14 @@
 	</alias>
 
 
+	<!-- StarOffice formerly included Arial Narrow - Solaris includes
+		Arial Narrow MT instead. -->
+	<alias>
+	  <family>Arial Narrow</family>
+	  <accept>
+	  <family>Arial Narrow MT</family>
+	  <family>Nimbus Sans L Condensed</family>
+	  </accept>
+	</alias>
+
 </fontconfig>
diff -ruN fontconfig-2.5.0/conf.d/40-nonlatin.conf fontconfig-2.5.0.new/conf.d/40-nonlatin.conf
diff -urpN -x '*~' -x '*.orig' conf.d/40-nonlatin.conf conf.d/40-nonlatin.conf
--- conf.d/40-nonlatin.conf	2007-11-05 15:38:03.000000000 -0800
+++ conf.d/40-nonlatin.conf	2008-05-14 09:57:19.180680000 +0900
@@ -10,6 +10,17 @@
   Serif faces
  -->
 	<alias>
+		<family>HG-PMinchoL-Sun</family>
+		<family>HG-MinchoL-Sun</family>
+		<family>IPAPMincho</family>
+		<family>IPAMincho</family>
+		<family>Sazanami Mincho</family>
+		<family>Sazanami Gothic</family>
+		<family>FZSongTi</family>
+		<family>FZMingTi</family>
+		<family>KacstQurn</family>
+		<family>SunDotum</family>
+		<family>AR PL ShanHeiSun Uni</family>
 		<family>Kochi Mincho</family>
 		<family>AR PL SungtiL GB</family>
 		<family>AR PL Mingti2L Big5</family>
@@ -26,6 +37,15 @@
   Sans-serif faces
  -->
 	<alias>
+		<family>HG-PGothicB-Sun</family>
+		<family>HG-GothicB-Sun</family>
+		<family>IPAPGothic</family>
+		<family>IPAGothic</family>
+		<family>Sazanami Gothic</family>
+		<family>FZSongTi</family>
+		<family>FZMingTi</family>
+		<family>KacstQurn</family>
+		<family>SunDotum</family>
 		<family>Kochi Gothic</family>
 		<family>AR PL KaitiM GB</family>
 		<family>AR PL KaitiM Big5</family>
@@ -43,6 +63,7 @@
   Monospace faces
  -->
  	<alias>
+		<family>SunDotumChe</family>
 		<family>NSimSun</family>
 		<family>ZYSong18030</family>
 		<family>FreeMono</family>
diff -urpN -x '*~' -x '*.orig' conf.d/45-latin.conf conf.d/45-latin.conf
--- conf.d/45-latin.conf	2007-11-05 15:38:03.000000000 -0800
+++ conf.d/45-latin.conf	2008-03-19 18:14:16.242946000 -0700
@@ -10,6 +10,7 @@
   Serif faces
  -->
 	<alias>
+		<family>Lucida Bright</family>
 		<family>Bitstream Vera Serif</family>
 		<family>DejaVu Serif</family>
 		<family>Liberation Serif</family>
@@ -25,6 +26,7 @@
   Sans-serif faces
  -->
 	<alias>
+		<family>Arial</family>
 		<family>Bitstream Vera Sans</family>
 		<family>DejaVu Sans</family>
 		<family>Liberation Sans</family>
@@ -41,6 +43,7 @@
   Monospace faces
  -->
  	<alias>
+		<family>Lucida Sans Typewriter</family>
 		<family>Bitstream Vera Sans Mono</family>
 		<family>DejaVu Sans Mono</family>
 		<family>Liberation Mono</family>
diff -urpN -x '*~' -x '*.orig' conf.d/49-sun-preuser.conf conf.d/49-sun-preuser.conf
--- conf.d/49-sun-preuser.conf	1969-12-31 16:00:00.000000000 -0800
+++ conf.d/49-sun-preuser.conf	2008-03-19 18:14:16.243311000 -0700
@@ -0,0 +1,79 @@
+<?xml version="1.0"?>
+<!DOCTYPE fontconfig SYSTEM "fonts.dtd">
+<fontconfig>
+	<!--
+	 Hinting for CJK fonts doesn't yet work as well as for Latin fonts.
+	 Some people may want to turn it off:
+	 -->
+
+	<match target="font">
+		<test name="lang" compare="contains">
+			<string>zh</string>
+			<string>ko</string>
+			<string>ja</string>
+		</test>
+		<edit name="hinting" mode="assign">
+			<bool>false</bool>
+		</edit>
+	</match>
+
+	<!--
+	 Chinese fonts are too light. We want to darken them up.
+	 darken_value is an int between 1 and 9 where 1 is a little
+	 darker and 9 is very dark. The default is 6.
+	 -->
+
+	<match target="font">
+	        <test name="lang" compare="contains">
+	                <string>zh</string>
+	        </test>
+	        <edit name="darken" mode="assign">
+	                <bool>true</bool>
+	        </edit>
+	        <edit name="darken_value" mode="assign">
+	                <int>6</int>
+	        </edit>
+	</match>
+
+	<!--
+	  To avoid using embedded bitmaps in TrueType fonts
+	  put the following rule in your personal ~/.fonts.conf file:
+	                                                                         
+	  <match target="pattern">
+	        <edit name="prefer_bitmap">
+	                <bool>false</bool>
+	        </edit>
+	  </match>
+	-->
+	<match target="pattern">
+		<edit name="prefer_bitmap">
+			<bool>true</bool>
+		</edit>
+	</match>
+	<match target="font">
+		<!-- check to see if the font is roman -->
+		<test name="slant">
+			<const>roman</const>
+		</test>
+		<!-- check to see if the pattern requested non-roman -->
+		<test target="pattern" name="slant" compare="not_eq">
+			<const>roman</const>
+		</test>
+		<edit name="prefer_bitmap">
+			<bool>false</bool>
+		</edit>
+	</match>
+	<!-- Latin fonts should not be used for ASCII characters when using
+	     Japanese monospace families -->
+	<match target="pattern">
+		<test name="family">
+			<string>monospace</string>
+		</test>
+		<test name="lang" compare="eq">
+			<string>ja-jp</string>
+		</test>
+		<edit name="disable_pango_script">
+			<bool>true</bool>
+		</edit>
+	</match>
+</fontconfig>
diff -urpN -x '*~' -x '*.orig' conf.d/60-latin.conf conf.d/60-latin.conf
--- conf.d/60-latin.conf	2006-09-09 20:30:00.000000000 -0700
+++ conf.d/60-latin.conf	2008-03-19 18:14:16.251367000 -0700
@@ -4,8 +4,9 @@
 	<alias>
 		<family>serif</family>
 		<prefer>
-			<family>Bitstream Vera Serif</family>
+			<family>Lucida Bright</family>
 			<family>DejaVu Serif</family>
+			<family>Bitstream Vera Serif</family>
 			<family>Times New Roman</family>
 			<family>Thorndale AMT</family>
 			<family>Luxi Serif</family>
@@ -16,8 +17,10 @@
 	<alias>
 		<family>sans-serif</family>
 		<prefer>
-			<family>Bitstream Vera Sans</family>
+			<family>Arial</family>
 			<family>DejaVu Sans</family>
+			<family>Bitstream Vera Sans</family>
+			<family>EUDC</family>
 			<family>Verdana</family>
 			<family>Arial</family>
 			<family>Albany AMT</family>
diff -urpN -x '*~' -x '*.orig' conf.d/65-nonlatin.conf conf.d/65-nonlatin.conf
--- conf.d/65-nonlatin.conf	2007-11-05 15:38:03.000000000 -0800
+++ conf.d/65-nonlatin.conf	2008-05-14 10:03:44.697829000 +0900
@@ -4,12 +4,21 @@
 	<alias>
 		<family>serif</family>
 		<prefer>
+			<family>EUDC</family>
 			<family>Frank Ruehl</family>
 			<family>MgOpen Canonica</family>
-			<family>Sazanami Mincho</family>
+			<family>HG-PMinchoL-Sun</family>
+			<family>HG-MinchoL-Sun</family>
 			<family>IPAMonaMincho</family>
+			<family>IPAPMincho</family>
 			<family>IPAMincho</family>
-			<family>Kochi Mincho</family>
+			<family>Sazanami Mincho</family>
+			<family>Sazanami Gothic</family>
+			<family>FZSongTi</family>
+			<family>FZMingTi</family>
+			<family>KacstQurn</family>
+			<family>SunDotum</family>
+			<family>AR PL ShanHeiSun Uni</family>
 			<family>AR PL SungtiL GB</family>
 			<family>AR PL Mingti2L Big5</family>
  			<family>AR PL Zenkai Uni</family>
@@ -33,13 +42,20 @@
 	<alias>
 		<family>sans-serif</family>
 		<prefer>
+			<family>EUDC</family>
 			<family>Nachlieli</family>
 			<family>MgOpen Modata</family>
 			<family>VL Gothic</family>
+			<family>HG-PGothicB-Sun</family>
+			<family>HG-GothicB-Sun</family>
 			<family>IPAMonaGothic</family>
+			<family>IPAPGothic</family>
 			<family>IPAGothic</family>
 			<family>Sazanami Gothic</family>
-			<family>Kochi Gothic</family>
+			<family>FZSongTi</family>
+			<family>FZMingTi</family>
+			<family>KacstQurn</family>
+			<family>SunDotum</family>
 			<family>AR PL KaitiM GB</family>
 			<family>AR PL KaitiM Big5</family>
  			<family>AR PL ShanHeiSun Uni</family>
@@ -66,12 +82,18 @@
 	<alias>
 		<family>monospace</family>
 		<prefer>
+			<family>EUDC</family>
 			<family>Miriam Mono</family>
 			<family>VL Gothic</family>
+			<family>HG-GothicB-Sun</family>
+			<family>HG-MinchoL-Sun</family>
 			<family>IPAMonaGothic</family>
 			<family>IPAGothic</family>
 			<family>Sazanami Gothic</family>
-			<family>Kochi Gothic</family>
+			<family>FZSongTi</family>
+			<family>FZMingTi</family>
+			<family>KacstQurn</family>
+			<family>SunDotumChe</family>
 			<family>AR PL KaitiM GB</family>
  			<family>AR PL ShanHeiSun Uni</family>
 			<family>AR PL SungtiL GB</family>
diff -urpN -x '*~' -x '*.orig' conf.d/Makefile.am conf.d/Makefile.am
--- conf.d/Makefile.am	2007-11-05 15:38:03.000000000 -0800
+++ conf.d/Makefile.am	2008-03-19 18:14:16.252419000 -0700
@@ -38,6 +38,7 @@ CONF_FILES = \
 	40-nonlatin.conf \
 	45-latin.conf \
 	49-sansserif.conf \
+	49-sun-preuser.conf \
 	50-user.conf \
 	51-local.conf \
 	60-latin.conf \
@@ -57,6 +58,7 @@ CONF_LINKS = \
 	40-nonlatin.conf \
 	45-latin.conf \
 	49-sansserif.conf \
+	49-sun-preuser.conf \
 	50-user.conf \
 	51-local.conf \
 	60-latin.conf \
