diff --git a/scanpci/scanpci.c b/scanpci/scanpci.c
index 1f5f8bd..480e152 100644
--- a/scanpci/scanpci.c
+++ b/scanpci/scanpci.c
@@ -151,11 +151,14 @@ print_pci_device( struct pci_device * dev, int verbose )
 	pci_device_probe( dev );
 	for ( i = 0 ; i < 6 ; i++ ) {
 	    if ( dev->regions[i].base_addr != 0 ) {
-		printf( "  BASE%u     0x%08"PRIxPTR" SIZE %zu  %s",
+		static const char *fmt8 = "  BASE%u     0x%08llx SIZE %d  %s";
+		static const char *fmt16 = "  BASE%u     0x%016llx SIZE %d  %s";
+		printf( dev->regions[i].is_64 ? fmt16 : fmt8,
 			i,
 			(intptr_t) dev->regions[i].base_addr,
 			(size_t) dev->regions[i].size,
-			(dev->regions[i].is_IO) ? "I/O" : "MEM" );
+			(dev->regions[i].is_IO) ? "I/O" : 
+			((dev->regions[i].is_64) ? "MEM64" : "MEM") );
 
 		if ( ! dev->regions[i].is_IO ) {
 		    if ( dev->regions[i].is_prefetchable ) {
diff --git a/src/solx_devfs.c b/src/solx_devfs.c
index 5f0db2d..6b8db2c 100644
--- a/src/solx_devfs.c
+++ b/src/solx_devfs.c
@@ -897,14 +897,10 @@ pci_device_solx_devfs_probe( struct pci_device * dev )
 	}
 
 
-	/*
-	 * We split the shift count 32 into two 16 to
-	 * avoid the complaining of the compiler
-	 */
-	dev->regions[ent].base_addr = reg->pci_phys_low +
-	    ((reg->pci_phys_mid << 16) << 16);
-	dev->regions[ent].size = reg->pci_size_low +
-	    ((reg->pci_size_hi << 16) << 16);
+	dev->regions[ent].base_addr = reg->pci_phys_low |
+	    ((uint64_t)reg->pci_phys_mid << 32);
+	dev->regions[ent].size = reg->pci_size_low |
+	    ((uint64_t)reg->pci_size_hi << 32);
 
 	switch (reg->pci_phys_hi & PCI_REG_ADDR_M) {
 	    case PCI_ADDR_IO:
