diff -urp -x '*~' -x '*.orig' src/scanpci.man src/scanpci.man
--- src/scanpci.man	2008-10-17 13:46:29.000000000 -0700
+++ src/scanpci.man	2008-10-17 14:37:48.524928000 -0700
@@ -35,10 +35,16 @@ about the configuration space settings f
 On most platforms,
 .I scanpci
 can only be run by the root user.
+On __distroname__ it may also be run using the 
+.I "Desktop Configuration"
+rights profile.  See \fBrbac\fR(5) and \fBuser_attr\fR(4).
 .SH OPTIONS
 .TP 8
 .B \-v
 Print the configuration space information for each device in a verbose
 format.  Without this option, only a brief description is printed for
 each device.
+.SH "SEE ALSO"
+.BR prtconf (1M), prtdiag (1M), smbios (1M),
+.BR user_attr (4), rbac (5).
 
