+++ a/src/ast.h	Fri Mar 15 10:45:43 2013
--- /net/overthere/export/home/x-s12-clone/open-src/driver/xf86-video-ast/ast.h	Fri Mar 15 13:49:13 2013
@@ -341,9 +341,19 @@
 #define VIS_SETIOREG                    (VIOC | 17)
 #define VIS_GETIOREG                    (VIOC | 18)
 
+#ifdef __sparc
+#define ASTIOC                  ('Y' << 8)
+#define AST_GET_INDEX          (ASTIOC | 14)
+#define AST_SET_INDEX          (ASTIOC | 15)
+#endif
+
 typedef struct vis_io_reg {
         uchar_t         offset;
         uchar_t         value;
+#ifdef __sparc
+        uchar_t         offset1;
+        uchar_t         value1;
+#endif
 } vis_io_reg_t;
 #endif
 extern struct pci_device *ASTGetPciInfo(ASTRecPtr);
