--- src/ast_2dtool.c	Wed Feb 20 11:23:06 2013
+++ src/ast_2dtool.c	Wed Feb 20 11:23:32 2013
@@ -27,7 +27,6 @@
 #include "xf86_OSproc.h"
 #include "xf86cmap.h"
 #include "compiler.h"
-#include "mibstore.h"
 #include "vgaHW.h"
 #include "mipointer.h"
 #include "micmap.h"
--- src/ast_accel.c	Wed Feb 20 11:22:41 2013
+++ src/ast_accel.c	Wed Feb 20 11:23:12 2013
@@ -27,7 +27,6 @@
 #include "xf86_OSproc.h"
 #include "xf86cmap.h"
 #include "compiler.h"
-#include "mibstore.h"
 #include "vgaHW.h"
 #include "mipointer.h"
 #include "micmap.h"
--- src/ast_cursor.c	Wed Feb 20 11:23:47 2013
+++ src/ast_cursor.c	Wed Feb 20 11:24:00 2013
@@ -27,7 +27,6 @@
 #include "xf86_OSproc.h"
 #include "xf86cmap.h"
 #include "compiler.h"
-#include "mibstore.h"
 #include "vgaHW.h"
 #include "mipointer.h"
 #include "micmap.h"
--- src/ast_mode.c	Wed Feb 20 11:23:00 2013
+++ src/ast_mode.c	Wed Feb 20 11:23:37 2013
@@ -27,7 +27,6 @@
 #include "xf86_OSproc.h"
 #include "xf86cmap.h"
 #include "compiler.h"
-#include "mibstore.h"
 #include "vgaHW.h"
 #include "mipointer.h"
 #include "micmap.h"
--- src/ast_vgatool.c	Wed Feb 20 11:23:50 2013
+++ src/ast_vgatool.c	Wed Feb 20 11:23:55 2013
@@ -27,7 +27,6 @@
 #include "xf86_OSproc.h"
 #include "xf86cmap.h"
 #include "compiler.h"
-#include "mibstore.h"
 #include "vgaHW.h"
 #include "mipointer.h"
 #include "micmap.h"
--- src/ast_tool.c	Wed Feb 20 11:22:51 2013
+++ src/ast_tool.c	Wed Feb 20 11:23:19 2013
@@ -27,7 +27,6 @@
 #include "xf86_OSproc.h"
 #include "xf86cmap.h"
 #include "compiler.h"
-#include "mibstore.h"
 #include "vgaHW.h"
 #include "mipointer.h"
 #include "micmap.h"
--- src/ast_driver.c	Wed Feb 20 11:22:46 2013
+++ src/ast_driver.c	Wed Feb 20 11:31:25 2013
@@ -31,7 +31,6 @@
 #endif
 #include "xf86cmap.h"
 #include "compiler.h"
-#include "mibstore.h"
 #include "vgaHW.h"
 #include "mipointer.h"
 #include "micmap.h"
@@ -939,7 +938,6 @@
    }
 #endif /* end of Accel_2D */
      
-   miInitializeBackingStore(pScreen);
    xf86SetBackingStore(pScreen);
    xf86SetSilkenMouse(pScreen);
 
