--- src/Cmap.h	Fri Jun  1 23:37:19 2012
+++ src/Cmap.h	Thu May  9 10:17:19 2013
@@ -1,3 +1,4 @@
+#include <X11/Xlib.h>
 
 #ifndef _CMAP_H_
 #define _CMAP_H_
--- src/Cr.h	Fri Jun  1 23:37:19 2012
+++ src/Cr.h	Thu May  9 10:17:25 2013
@@ -1,3 +1,4 @@
+#include <X11/Xlib.h>
 
 #ifndef _CR_H_
 #define _CR_H_
--- src/Key.h	Fri Jun  1 23:37:19 2012
+++ src/Key.h	Thu May  9 10:24:27 2013
@@ -1,3 +1,5 @@
+#include <X11/Xlib.h>
+#include <X11/Xresource.h>
 
 #ifndef _KEY_H_
 #define _KEY_H_
--- src/Xresinternal.h	Fri Jun  1 23:37:19 2012
+++ src/Xresinternal.h	Thu May  9 10:27:08 2013
@@ -2,6 +2,8 @@
 #ifndef _XRESINTERNAL_H_
 #define _XRESINTERNAL_H_
 
+#include <X11/Xlib.h>
+#include <X11/Xresource.h>
 #include <inttypes.h>
 
 /* type defines */
--- src/locking.h	Fri Jun  1 23:37:19 2012
+++ src/locking.h	Thu May  9 10:29:25 2013
@@ -36,6 +36,8 @@
 
 #define xmalloc(s) Xmalloc(s)
 #define xfree(s) Xfree(s)
+#include <X11/Xlib.h>
+#include <X11/Xlibint.h>
 #include <X11/Xthreads.h>
 
 struct _XCVList {
--- src/Xintatom.h	Fri Jun  1 23:37:19 2012
+++ src/Xintatom.h	Thu May  9 10:19:14 2013
@@ -1,7 +1,7 @@
-
 #ifndef _XINTATOM_H_
 #define _XINTATOM_H_ 1
 
+#include <X11/Xlib.h>
 #include <X11/Xfuncproto.h>
 
 /* IntAtom.c */
--- src/Xintconn.h	Fri Jun  1 23:37:19 2012
+++ src/Xintconn.h	Thu May  9 10:19:05 2013
@@ -1,8 +1,8 @@
-
 #ifndef _XINTCONN_H_
 #define _XINTCONN_H_ 1
 
 #include <X11/Xfuncproto.h>
+#include <X11/Xlib.h>
 
 _XFUNCPROTOBEGIN
 
--- src/Xprivate.h	Fri Jun  1 23:37:19 2012
+++ src/Xprivate.h	Thu May  9 10:19:26 2013
@@ -8,6 +8,8 @@
 #ifndef XPRIVATE_H
 #define XPRIVATE_H
 
+#include <X11/Xlib.h>
+
 extern _X_HIDDEN void _XIDHandler(Display *dpy);
 extern _X_HIDDEN void _XSeqSyncFunction(Display *dpy);
 extern _X_HIDDEN void _XSetPrivSyncFunction(Display *dpy);
