--- src/OpenDis.c	2010-12-01 22:37:45.756982154 -0800
+++ src/OpenDis.c	2010-12-01 22:37:46.942754311 -0800
@@ -39,6 +39,13 @@ in this Software without prior written a
 #include <X11/Xatom.h>
 #include <X11/Xresource.h>
 #include <stdio.h>
+#if defined(sun) || defined(__sun) || defined(__sun__)
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#endif
 #include "Xintconn.h"
 
 #ifdef XKB
@@ -841,9 +848,11 @@ fallback_success:
      */
         _conn_buf_size = conn_buf_size;
 
+#if !USE_XCB
         if (sme_conn)
            InitializeSmeConn(dpy, idisplay);
         else
+#endif /* USE_XCB */
            dpy->smeBuffer = NULL ;
 
 #endif /* SME */

--- src/Makefile.am	2010-12-01 22:37:45.741732419 -0800
+++ src/Makefile.am	2010-12-01 22:37:46.942993564 -0800
@@ -3,7 +3,7 @@ XKB_SUBDIRS = xkb
 endif
 SUBDIRS = util xcms xlibi18n $(XKB_SUBDIRS)
 
-lib_LTLIBRARIES=libX11.la
+lib_LTLIBRARIES=libX11.la libX11-xcb.la
 
 BUILT_SOURCES=ks_tables.h
 CLEANFILES=ks_tables.h ks_tables_h
@@ -353,7 +353,6 @@ libX11_la_SOURCES += \
                   xcb_io.c \
                   Xxcbint.h
 
-lib_LTLIBRARIES += libX11-xcb.la
 libX11_xcb_la_SOURCES = x11_xcb.c Xxcbint.h
 libX11_xcb_la_LDFLAGS = -version-info 1:0:0 -no-undefined
 libX11_xcb_la_LIBADD = libX11.la
