--- nls/en_US.UTF-8/XLC_LOCALE.pre	Thu Apr 17 21:51:32 2008
+++ nls/en_US.UTF-8/XLC_LOCALE.pre.new	Wed Apr 23 02:14:56 2008
@@ -34,7 +34,6 @@
 fs2	{
 	charset		ISO8859-5:GR
 	font {
-	    load_option delay_nocheck
 	    primary	ISO8859-5:GR
 	}
 }
@@ -42,7 +41,6 @@
 fs3	{
 	charset		ISO8859-7:GR
 	font {
-	    load_option delay_nocheck
 	    primary	ISO8859-7:GR
 	}
 }
@@ -50,7 +48,6 @@
 fs4	{
 	charset		ISO8859-2:GR
 	font {
-	    load_option delay_nocheck
 	    primary	ISO8859-2:GR
 	}
 }
@@ -66,7 +63,6 @@
 fs6	{
 	charset		ISO8859-9:GR
 	font {
-	    load_option delay_nocheck
 	    primary	ISO8859-9:GR
 	}
 }
