--- nls/locale.alias.pre	Sat May 17 05:28:12 2008
+++ nls/locale.alias.pre.new	Thu May 29 03:58:12 2008
@@ -837,6 +837,8 @@
 no_NO.ISO-8859-1:				no_NO.ISO8859-1
 no_NO.iso885915:				no_NO.ISO8859-15
 no_NO.ISO-8859-15:				no_NO.ISO8859-15
+no_NO.ISO8859-1@bokmal:			no_NO.ISO8859-1
+no_NO.ISO8859-1@nynorsk:			no_NO.ISO8859-1
 no_NO.utf8:					no_NO.UTF-8
 nr:							nr.ISO8859-2
 nr_ZA:						nr_ZA.ISO8859-1
@@ -946,6 +948,7 @@
 XCOMM YU was the old ISO code for Yugoslavia (now CS for Serbia&Montenegro).
 XCOMM These lines should be deleted in the future.
 sh:						sr_CS.ISO8859-2
+sh_BA.ISO8859-2@bosnia:				nr.ISO8859-2
 sh_HR.iso88592:					hr_HR.ISO8859-2
 sh_SP:						sr_CS.ISO8859-2
 sh_YU:						sr_CS.ISO8859-2
