diff -urp -x '*~' -x '*.orig' modules/lc/gen/lcGenConv.c modules/lc/gen/lcGenConv.c
--- modules/lc/gen/lcGenConv.c	2009-03-17 14:58:15.000000000 -0700
+++ modules/lc/gen/lcGenConv.c	2009-04-08 00:19:18.692831000 -0700
@@ -1786,9 +1786,6 @@ mbtocs(
         return( 0 );
     }
 
-    if (*from_left > *to_left)
-        *from_left = *to_left;
-
     while (*from_left && *to_left) {
 
 	ch = *inbufptr++;
