diff -urp -x '*~' -x '*.orig' modules/lc/gen/lcGenConv.c modules/lc/gen/lcGenConv.c
--- modules/lc/gen/lcGenConv.c	2010-09-03 22:52:38.000000000 -0700
+++ modules/lc/gen/lcGenConv.c	2010-11-21 18:47:57.083531216 -0800
@@ -1784,9 +1784,6 @@ mbtocs(
         return( 0 );
     }
 
-    if (*from_left > *to_left)
-        *from_left = *to_left;
-
     while (*from_left && *to_left) {
 
 	ch = *inbufptr++;
