diff -urp -x '*~' -x '*.orig' modules/om/generic/omGeneric.c modules/om/generic/omGeneric.c
--- modules/om/generic/omGeneric.c	2008-10-07 10:18:17.000000000 -0700
+++ modules/om/generic/omGeneric.c	2009-04-08 00:19:29.991807000 -0700
@@ -841,7 +841,7 @@ parse_omit_name(
 	return True;
 
     /* This may mot be needed anymore as XListFonts() takes care of this */
-    while (num_fields < 12) {
+/*    while (num_fields < 12) {
 	if ((last - buf) > (XLFD_MAX_LEN - 2))
 	    return -1;
 	*last = '*';
@@ -852,7 +852,7 @@ parse_omit_name(
 	if ((font_data->xlfd_name = get_font_name(oc, buf)) != NULL)
 	    return True;
     }
-
+ */
 
     return False;
 }
