--- nls/locale.dir.pre	Thu Feb 19 02:13:38 2009
+++ nls/locale.dir.pre	Tue Mar  3 03:09:43 2009
@@ -427,6 +427,9 @@
 zh_TW.UTF-8/XLC_LOCALE:			zh_TW.UTF-8
 en_US.UTF-8/XLC_LOCALE:			zu_ZA.UTF-8
 en_US.UTF-8/XLC_LOCALE:			kk_KZ.UTF-8
+en_US.UTF-8/XLC_LOCALE:			es_US.UTF-8
+en_US.UTF-8/XLC_LOCALE:			mt_MT.UTF-8
+en_US.UTF-8/XLC_LOCALE:			sh_BA.UTF-8
 #ifdef __uxp__
 ja.U90/XLC_LOCALE:			ja_JP.U90
 ja.S90/XLC_LOCALE:			ja_JP.S90
--- nls/compose.dir.pre	Thu Feb 19 02:13:38 2009
+++ nls/compose.dir.pre	Tue Mar  3 03:12:22 2009
@@ -392,3 +392,7 @@
 zh_HK.UTF-8/Compose:		zh_HK.UTF-8
 en_US.UTF-8/Compose:		zu_ZA.UTF-8
 en_US.UTF-8/Compose:		bs_BA.UTF-8
+en_US.UTF-8/Compose:		kk_KZ.UTF-8
+en_US.UTF-8/Compose:		es_US.UTF-8
+en_US.UTF-8/Compose:		mt_MT.UTF-8
+en_US.UTF-8/Compose:		sh_BA.UTF-8
