From e5892467ae3308c8651be76e06db322dcbc08522 Mon Sep 17 00:00:00 2001
From: Alan Coopersmith <alan.coopersmith@sun.com>
Date: Fri, 15 Feb 2008 17:27:53 -0800
Subject: [PATCH] Add support for building lint library with --enable-lint-library
---
 configure.ac    |    5 +++++
 src/Makefile.am |   13 +++++++++++++
 2 files changed, 18 insertions(+), 0 deletions(-)

diff --git a/configure.ac b/configure.ac
index 604d1bb..7ab9b7f 100644
--- a/configure.ac
+++ b/configure.ac
@@ -14,6 +14,10 @@ AM_MAINTAINER_MODE
 AM_CONFIG_HEADER([src/config.h])
 AC_CONFIG_HEADER([include/X11/XlibConf.h])
 
+# Require xorg-macros version 1.1.0 or newer for XORG_WITH_LINT macro
+m4_ifndef([XORG_MACROS_VERSION], [AC_FATAL([must install xorg-macros 1.1 or later before running autoconf/autogen])])
+XORG_MACROS_VERSION(1.1)
+
 # Checks for programs.
 AC_PROG_LIBTOOL
 AC_PROG_CC
@@ -331,6 +335,7 @@ fi
 
 dnl Allow checking code with lint, sparse, etc.
 XORG_WITH_LINT
+XORG_LINT_LIBRARY([X11])
 LINT_FLAGS="${LINT_FLAGS} ${X11_CFLAGS} ${XPROTO_CFLAGS}"
 
 if test "x$GCC" = "xyes"; then
diff --git a/src/Makefile.am b/src/Makefile.am
index 7c44f5b..564e03f 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -402,6 +402,19 @@ lint:	$(BUILT_SOURCES)
 	done
 endif LINT
 
+if MAKE_LINT_LIB
+lintlibdir = $(libdir)
+
+lintlib_DATA = $(LINTLIB)
+
+lintlib_src = $(libX11_la_SOURCES) xcms/*.c xkb/*.c
+
+CLEANFILES += $(lintlib_DATA)
+
+$(LINTLIB): $(libX11_la_SOURCES)
+	$(LINT) -y -oX11 -x $(ALL_LINT_FLAGS) $(lintlib_src)
+endif MAKE_LINT_LIB
+
 #
 # Building ks_tables.h requires the makekeys utility
 #
-- 
1.4.1

