diff --git a/configure.ac b/configure.ac
index d622697..6b8a6ba 100644
--- a/configure.ac
+++ b/configure.ac
@@ -376,7 +376,6 @@ fi
 dnl Allow checking code with lint, sparse, etc.
 XORG_WITH_LINT
 XORG_LINT_LIBRARY([X11])
-LINT_FLAGS="${LINT_FLAGS} ${X11_CFLAGS} ${XPROTO_CFLAGS}"
 
 if test "x$GCC" = "xyes"; then
 	GCC_WARNINGS="-Wall -Wpointer-arith -Wstrict-prototypes \
diff --git a/src/Makefile.am b/src/Makefile.am
index 51e1e06..2a59d6f 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -19,6 +19,7 @@ AM_CPPFLAGS= \
 	-D_BSD_SOURCE -DX11_t -DTRANS_CLIENT
 
 AM_CFLAGS= \
+	$(XPROTO_CFLAGS) \
 	$(X11_CFLAGS) \
 	$(BIGFONT_CFLAGS) \
 	$(XDMCP_CFLAGS) \
