--- a/src/mga_driver.c	Thu Oct 18 13:55:11 2012
+++ b/src/mga_driver.c	Thu Oct 18 13:57:51 2012
@@ -109,6 +109,9 @@
 static const OptionInfoRec *	MGAAvailableOptions(int chipid, int busid);
 static void	MGAIdentify(int flags);
 #ifdef XSERVER_LIBPCIACCESS
+#ifdef __sparc__
+static Bool     MGACompProbe(DriverPtr drv, int flags);
+#endif
 static Bool MGAPciProbe(DriverPtr drv, int entity_num,
     struct pci_device * dev, intptr_t match_data);
 #else
@@ -510,7 +513,11 @@
     MGA_DRIVER_NAME,
     MGAIdentify,
 #ifdef XSERVER_LIBPCIACCESS
+#ifdef __sparc__
+    MGACompProbe,
+#else
     NULL,
+#endif
 #else
     MGAProbe,
 #endif
@@ -4180,3 +4187,20 @@
     OUTREG(MGAREG_MACCESS, 1<<15);
     usleep(10);
 }
+
+#ifdef __sparc__
+static Bool
+MGACompProbe(DriverPtr drvp, int flags)
+{
+   static Bool PciInitDone = FALSE;
+
+    /* Create pci system for use by the driver */
+    if (!PciInitDone) {
+       xf86BusProbe();
+      PciInitDone = TRUE;
+    }
+
+    return xf86PciProbeDev(drvp);
+}
+#endif
+
