diff -urp -x '*~' -x '*.orig' Makefile.am Makefile.am
--- Makefile.am	2010-10-30 17:17:29.000000000 -0700
+++ Makefile.am	2011-01-28 20:22:00.477766877 -0800
@@ -19,24 +19,18 @@
 #  TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
 #  PERFORMANCE OF THIS SOFTWARE.
 
-bin_PROGRAMS = xmag
+bin_PROGRAMS = xmag_multivis
 
-AM_CFLAGS = $(XMAG_CFLAGS) $(CWARNFLAGS)
-xmag_LDADD = $(XMAG_LIBS) -lm
+AM_CFLAGS = $(XMAG_CFLAGS) $(CWARNFLAGS) -DSHAPE -DALLPLANES
+xmag_multivis_LDADD = $(XMAG_LIBS) -lXext -lm
 
-xmag_SOURCES =	\
-        CutPaste.c \
-        CutPaste.h \
-        RootWin.c \
-        RootWin.h \
-        RootWinP.h \
-        Scale.c \
-        Scale.h \
-        ScaleP.h \
-        xmag.c
+xmag_multivis_SOURCES = \
+        xmag_multivis.c \
+        multivis.c \
+        multivis.h
 
 appman_PRE = \
-        xmag.man
+        xmag_multivis.man
 
 # App default files  (*.ad)
 
