--- icccm/xcb_icccm.h	2009-08-24 04:47:36.000000000 -0400
+++ icccm/xcb_icccm.h	2009-12-07 00:07:19.709961000 -0500
@@ -521,7 +521,7 @@
  * @param win_gravity Window gravity value.
  */
 void xcb_size_hints_set_win_gravity(xcb_size_hints_t *hints,
-                                    uint32_t win_gravity);
+                                    xcb_gravity_t win_gravity);
 
 /**
  * @brief Deliver a ChangeProperty request to set a value to a given property.
